# Gc_GENRE_2022

## Abstract
Abstract
The ability of bacterial pathogens to metabolically adapt to the environmental conditions of their hosts is critical to both colonization and invasive disease. Infection with Neisseria gonorrhoeae (the gonococcus, Gc) is characterized by the influx of neutrophils (PMNs), which fail to clear the bacteria and make antimicrobial products that can exacerbate tissue damage. The inability of the human host to clear Gc infection is particularly concerning in light of the emergence of strains that are resistant to all clinically recommended antibiotics. Bacterial metabolism represents a promising target for the development of new therapeutics against Gc. Here, we generated a curated genome-scale metabolic network reconstruction (GENRE) of Gc strain FA1090. This GENRE links genetic information to metabolic phenotypes and predicts Gc biomass synthesis and energy consumption. We validated this model with published data and in new results reported here. Contextualization of this model using the transcriptional profile of Gc exposed to PMNs revealed substantial rearrangements of Gc central metabolism and induction of Gc nutrient acquisition strategies for alternate carbon source use. These features enhanced the growth of Gc in the presence of neutrophils. From these results we conclude that the metabolic interplay between Gc and PMNs helps define infection outcomes. The use of transcriptional profiling and metabolic modeling to reveal new mechanisms by which Gc persists in the presence of PMNs uncovers unique aspects of metabolism in this fastidious bacterium, which could be targeted to block infection and thereby reduce the burden of gonorrhea in the human population.

##Overview

project
|- README          		# description of content
|- LICENSE         		# the license for this project
|
|- Curate_Gc_Model/					# programmatic code and resources for curating iNGO_557(python)
| |- Annotation_resources/			# annotation information used for curating iNgo_557
| |- In_progress_curation/			# In progress model files for use during curation
| |- Notebooks/			# Jupyter notebooks containing code for curating iNGO_557 (python)
|
|- Figures/          		# programatic code for generating figures
| |- figure_1/			# manuscript figures
| |- figure_2/			# manuscript figures
| |- figure_3/			# manuscript figures
| |- figure_4/			# manuscript figures
| |- figure_5/			# manuscript figures	
|
|- Generate_Gc_Model/				# programmatic code and resources for generating iNGO_557 from |- Nmb_iTM560 (python) based on Norsigian 2020 protocol (PMID: 31863076)
| |- DraftModels/			# Draft models generated in progress
| |- FinalModels/			# final model generated by following Norsigian protocol
| |- Matrices/			# homology matrices
| |- Mendum_2011_annotations/			# annotations from Nmb_iTM560 from Mendum 2011
| |- Notebooks/			# Jupyter notebooks containing code for creating iNGO_557 (python)
| |- bbh/			# bidirectional best hit blast files
| |- genomes/			# genomes associated with this study
| |- nucl/			# nucleotide calls
| |- prots/			# protein calls
|
|- Memote/				# MEMOTE output for iNGO_557
|
|- Models/			# Final models created and used during this study
| |- AE004969.1.json   # model generated from Norsigian 2020 protocol
| |- Ngo_557.json   # final curated model for Gc strain FA 1090
| |- Ngo_557.sbml   # final curated model for Gc strain FA 1090
| |- Neisseria_gonorrhoeae_FA_1090.json   # Automated model for FA 1090 generated from CarveMe
| |- Neisseria_gonorrhoeae_FA_1090.json   # Automated model for FA 1090 generated from CarveMe (zip file)
| |- Nmb_iTM560.json   # Model for Neisseria meningitidis strain M58 from Mendum 2011
| |- Nmb_iTM560.sbml   # Model for Neisseria meningitidis strain M58 from Mendum 2011
| |- ecoli.json   # Model (iAF1260) for E.coli  K12  that Nmb_iTM560 was based on. 
| |- universal_model_cobrapy.json   # Universal model from BIGG
