{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1a64602c-5a60-4537-8040-f1b13d48a956",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception reporting mode: Verbose\n"
     ]
    }
   ],
   "source": [
    "# Import libraries - REQUIRES pip version 9.0.3\n",
    "import pandas\n",
    "import os\n",
    "from os.path import join\n",
    "import sys\n",
    "import scipy.stats\n",
    "import numpy\n",
    "import math\n",
    "import pickle\n",
    "import copy\n",
    "import time\n",
    "import warnings\n",
    "import gc\n",
    "\n",
    "# Using Cobrapy 0.13.0\n",
    "import cobra\n",
    "import cobra.test\n",
    "import cobra.flux_analysis.gapfilling\n",
    "from cobra.io import load_json_model, save_json_model, load_matlab_model, save_matlab_model, read_sbml_model, write_sbml_model\n",
    "from cobra.flux_analysis import  flux_variability_analysis\n",
    "from cobra.flux_analysis.reaction import assess_component\n",
    "from cobra.manipulation.delete import *\n",
    "from cobra.flux_analysis.parsimonious import add_pfba\n",
    "from cobra.medium import find_boundary_types\n",
    "from cobra.util import solver as sutil\n",
    "\n",
    "\n",
    "# Confidence levels in each annotation\n",
    "#import probanno\n",
    "\n",
    "# Estabish handler for logger\n",
    "import logging\n",
    "logging.basicConfig()\n",
    "logger = logging.getLogger('logger')\n",
    "\n",
    "# Verbose exception printing\n",
    "%xmode\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b6b987a-0029-493f-827b-08310836fa0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quicker way to read in models\n",
    "import pickle\n",
    "def read_model(fileName, obj='none'):\n",
    "    \n",
    "    fileType = fileName.split('.')[-1]\n",
    "    \n",
    "    if fileType == 'sbml' or fileType == 'xml':\n",
    "        model = cobra.io.read_sbml_model(fileName)\n",
    "    elif fileType == 'json':\n",
    "        model = cobra.io.load_json_model(fileName)\n",
    "    elif fileType == 'yaml':\n",
    "        model = cobra.io.load_yaml_model(fileName)\n",
    "    elif fileType == 'mat':\n",
    "        model = cobra.io.load_matlab_model(fileName)\n",
    "    elif fileType == 'pkl':\n",
    "        model = pickle.load(open(fileName, 'rb'))\n",
    "    else:\n",
    "        raise TypeError('Unrecognized file extension')\n",
    "    \n",
    "    if obj != 'none': model.objective = obj\n",
    "    for rxn in model.boundary: rxn.bounds = (-1000., 1000.)\n",
    "        \n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5c3f2373-ccd8-4c89-9e20-18544bc712b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/UNIPROT/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n",
      "WARNING:cobra.io.sbml:https://identifiers.org/PUBMLST_Locus/ does not conform to 'http(s)://identifiers.org/collection/id' or'http(s)://identifiers.org/COLLECTION:id\n"
     ]
    }
   ],
   "source": [
    "model= read_model(\"C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/annotatedGCmodel2.sbml\")\n",
    "NM_model = read_model('C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/archive/Nmb_iTM560.json')\n",
    "carveme_model= cobra.io.read_sbml_model('C:/Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/Neisseria_gonorrhoeae_FA_1090.xml.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2ece2aa3-48b1-4b91-844d-c38c214a1ae2",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_json_model(carveme_model, \"C:/Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/Neisseria_gonorrhoeae_FA_1090.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3be4d018-c1ce-456b-a023-5223297c3d9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ngo_558 - Genes: 560\n",
      "Nmb_iTM560 - Genes: 562\n",
      "CARVEME FA1090 - Genes: 510\n"
     ]
    }
   ],
   "source": [
    "#For Figure1A - description of models\n",
    "#Print Model Gene number\n",
    "#Subtract 2 from each for final value to remove Spontaneous and orphan identifiers\n",
    "print(\"Ngo_558 - Genes:\", len(model.genes))\n",
    "print(\"Nmb_iTM560 - Genes:\",len(NM_model.genes))\n",
    "print(\"CARVEME FA1090 - Genes:\",len(carveme_model.genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dfc29d4d-60bc-47db-93d9-85211da8feba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ngo_558-reactions : 1424\n",
      "Nmb_iTM560-reactions : 1519\n",
      "CARVEME FA1090 -reactions: 1327\n"
     ]
    }
   ],
   "source": [
    "#Print Model Reaction number\n",
    "print(\"Ngo_558-reactions :\", len(model.reactions))\n",
    "print(\"Nmb_iTM560-reactions :\",len(NM_model.reactions))\n",
    "print(\"CARVEME FA1090 -reactions:\",len(carveme_model.reactions))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3f28bda8-bef0-4c91-9f8c-eb1d58d559ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ngo_558- metabolites : 1265\n",
      "Nmb_iTM560- metabolites  : 1297\n",
      "CARVEME FA1090- metabolites  : 991\n"
     ]
    }
   ],
   "source": [
    "#Print Model Metabolites number\n",
    "print(\"Ngo_558- metabolites :\", len(model.metabolites))\n",
    "print(\"Nmb_iTM560- metabolites  :\",len(NM_model.metabolites))\n",
    "print(\"CARVEME FA1090- metabolites  :\",len(carveme_model.metabolites))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ef1075af-b0ce-4452-bb2b-aa5e63a7434c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ngo_558 Exchange reactions: 264\n",
      "Nmb_itm560 Exchange reactions: 262\n",
      "carveme_FA1090 Exchange reactions: 105\n"
     ]
    }
   ],
   "source": [
    "#Print Model GPR - Exchange Reactions gene number\n",
    "model_exchanges =[]\n",
    "for r in model.reactions:\n",
    "    if 'EX_' in r.id:\n",
    "        model_exchanges.append(r.id)\n",
    "print(\"Ngo_558 Exchange reactions:\", len(model_exchanges))\n",
    "\n",
    "model_exchanges =[]\n",
    "for r in NM_model.reactions:\n",
    "    if 'EX_' in r.id:\n",
    "        model_exchanges.append(r.id)\n",
    "print(\"Nmb_itm560 Exchange reactions:\", len(model_exchanges))\n",
    "\n",
    "model_exchanges =[]\n",
    "for r in carveme_model.reactions:\n",
    "    if 'EX_' in r.id:\n",
    "        model_exchanges.append(r.id)\n",
    "print(\"carveme_FA1090 Exchange reactions:\", len(model_exchanges))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2b8d86c7-8101-48cd-8e23-e5ba8e5e1c0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ngo_558 Exchange reactions: 264\n",
      "Ngo_558 Transport reactions: 474\n",
      "Ngo_558 purely metabolic reactions: 686\n",
      "\n",
      "\n",
      "Nmb_itm560 Exchange reactions: 262\n",
      "Nmb_itm560 Transport reactions: 555\n",
      "Nmb_itm560 purely metabolic reactions: 702\n",
      "\n",
      "\n",
      "carveme_FA1090 Exchange reactions: 105\n",
      "carveme_FA1090 Transport reactions: 279\n",
      "carveme_FA1090 purely metabolic reactions: 943\n"
     ]
    }
   ],
   "source": [
    "#For GPR Rules -- run memote to identify single gene, multigene, and none\n",
    "#And to define metabolic vs transport reactions\n",
    "model_exchanges =[]\n",
    "model_transports =[]\n",
    "model_metabolic =[]\n",
    "for rxn in model.reactions:\n",
    "    substrates = list(rxn.metabolites)\n",
    "    compartments = set([x.compartment for x in substrates])\n",
    "\n",
    "    if 'EX_' in rxn.id:\n",
    "        model_exchanges.append(rxn.id) # exchange\n",
    "    elif len(compartments) > 1:\n",
    "        model_transports.append(rxn.id) # transport\n",
    "    else:\n",
    "        model_metabolic.append(rxn.id) # metabolic\n",
    "\n",
    "print(\"Ngo_558 Exchange reactions:\", len(model_exchanges))\n",
    "print(\"Ngo_558 Transport reactions:\",len(model_transports))\n",
    "print(\"Ngo_558 purely metabolic reactions:\",len(model_metabolic))\n",
    "\n",
    "#Set compartments for Nmb_iTM560 so reaction types can be counted\n",
    "for metabolite in NM_model.metabolites:\n",
    "    if '_c_c' in metabolite.id:\n",
    "        metabolite.compartment= 'cytosol'\n",
    "for metabolite in NM_model.metabolites:\n",
    "    if '_p_c' in metabolite.id:\n",
    "        metabolite.compartment= 'periplasm'\n",
    "for metabolite in NM_model.metabolites:\n",
    "    if '_e_c' in metabolite.id:\n",
    "        metabolite.compartment= 'extracellular'\n",
    "        \n",
    "model_exchanges =[]\n",
    "model_transports =[]\n",
    "model_metabolic =[]\n",
    "for rxn in NM_model.reactions:\n",
    "    substrates = list(rxn.metabolites)\n",
    "    compartments = set([x.compartment for x in substrates])\n",
    "\n",
    "    if 'EX_' in rxn.id:\n",
    "        model_exchanges.append(rxn.id) # exchange\n",
    "    elif len(compartments) > 1:\n",
    "        model_transports.append(rxn.id) # transport\n",
    "    else:\n",
    "        model_metabolic.append(rxn.id) # metabolic\n",
    "print(\"\\n\")\n",
    "print(\"Nmb_itm560 Exchange reactions:\", len(model_exchanges))\n",
    "print(\"Nmb_itm560 Transport reactions:\",len(model_transports))\n",
    "print(\"Nmb_itm560 purely metabolic reactions:\",len(model_metabolic))\n",
    "\n",
    "model_exchanges =[]\n",
    "model_transports =[]\n",
    "model_metabolic =[]\n",
    "for rxn in carveme_model.reactions:\n",
    "    substrates = list(rxn.metabolites)\n",
    "    compartments = set([x.compartment for x in substrates])\n",
    "\n",
    "    if 'EX_' in rxn.id:\n",
    "        model_exchanges.append(rxn.id) # exchange\n",
    "    elif len(compartments) > 1:\n",
    "        model_transports.append(rxn.id) # transport\n",
    "    else:\n",
    "        model_metabolic.append(rxn.id) # metabolic\n",
    "print(\"\\n\")\n",
    "print(\"carveme_FA1090 Exchange reactions:\", len(model_exchanges))\n",
    "print(\"carveme_FA1090 Transport reactions:\",len(model_transports))\n",
    "print(\"carveme_FA1090 purely metabolic reactions:\",len(model_metabolic))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0486270f-2a25-44ff-b069-63b4260a2f96",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-11-87f40f18d62c>:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  kegg_annotations['ids'] = kegg_annotations['ids'] .str.replace('_', '')\n"
     ]
    }
   ],
   "source": [
    "#From https://www.biostars.org/p/492324/\n",
    "\n",
    "import pandas as pd\n",
    "from collections import * \n",
    "\n",
    "database = list()\n",
    "for _, v in pd.read_json(\"C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/Figures/Fig_1/ngo00001.json\").iterrows():\n",
    "    d = v[\"children\"]\n",
    "    cat_1 = d[\"name\"]\n",
    "    for child_1 in d[\"children\"]:\n",
    "        cat_2 = child_1[\"name\"] # Module?\n",
    "        for child_2 in child_1[\"children\"]:\n",
    "            cat_3 = child_2[\"name\"]\n",
    "            if \"children\" in child_2:\n",
    "                for child_3 in child_2[\"children\"]:\n",
    "                    cat_4 = child_3[\"name\"]\n",
    "                    fields = [cat_1, cat_2, cat_3, cat_4]\n",
    "                    database.append(fields)\n",
    "df_kegg = pd.DataFrame(database, columns=[\"Level_A\", \"Level_B\", \"Level_C\", \"Level_D\"])\n",
    "\n",
    "kos = list()\n",
    "for tag in df_kegg[\"Level_D\"]:\n",
    "        full_tag=tag.split()\n",
    "        geneid=full_tag[0]\n",
    "        kos.append(geneid) \n",
    "\n",
    "df_kegg[\"ids\"]= kos\n",
    "\n",
    "#Include only kegg metabolism related annotations\n",
    "kegg_annotations=df_kegg[df_kegg[\"Level_A\"].str.contains(\"09100 Metabolism\")==True]\n",
    "\n",
    "#Match id format between kegg and model\n",
    "kegg_annotations['ids'] = kegg_annotations['ids'] .str.replace('_', '')\n",
    "\n",
    "#Filter kegg annotations by genes found in model\n",
    "model_genes= list()\n",
    "for gene in model.genes:\n",
    "    model_genes.append(gene.id)\n",
    "model_kegg_orthologies=kegg_annotations[kegg_annotations.ids.isin(model_genes)]\n",
    "\n",
    "\n",
    "model_kegg_orthologies.to_csv('KEGG_orthologies_Ngo.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e38f88d4-d9c3-4cdb-9e19-d482282166be",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-13-dec624026df9>:35: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  kegg_annotations['ids'] = kegg_annotations['ids'] .str.replace('_', '')\n"
     ]
    }
   ],
   "source": [
    "#From https://www.biostars.org/p/492324/\n",
    "#Download Nme database of orthologies from KEGG\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "from collections import * \n",
    "\n",
    "database = list()\n",
    "for _, v in pd.read_json(\"C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/Figures/Fig_1/nme00001.json\").iterrows():\n",
    "    d = v[\"children\"]\n",
    "    cat_1 = d[\"name\"]\n",
    "    for child_1 in d[\"children\"]:\n",
    "        cat_2 = child_1[\"name\"] # Module?\n",
    "        for child_2 in child_1[\"children\"]:\n",
    "            cat_3 = child_2[\"name\"]\n",
    "            if \"children\" in child_2:\n",
    "                for child_3 in child_2[\"children\"]:\n",
    "                    cat_4 = child_3[\"name\"]\n",
    "                    fields = [cat_1, cat_2, cat_3, cat_4]\n",
    "                    database.append(fields)\n",
    "df_kegg = pd.DataFrame(database, columns=[\"Level_A\", \"Level_B\", \"Level_C\", \"Level_D\"])\n",
    "\n",
    "kos = list()\n",
    "for tag in df_kegg[\"Level_D\"]:\n",
    "        full_tag=tag.split()\n",
    "        geneid=full_tag[0]\n",
    "        kos.append(geneid) \n",
    "\n",
    "df_kegg[\"ids\"]= kos\n",
    "\n",
    "#Include only kegg metabolism related annotations\n",
    "kegg_annotations=df_kegg[df_kegg[\"Level_A\"].str.contains(\"09100 Metabolism\")==True]\n",
    "\n",
    "#Match id format between kegg and model\n",
    "kegg_annotations['ids'] = kegg_annotations['ids'] .str.replace('_', '')\n",
    "\n",
    "#Filter kegg annotations by genes found in model\n",
    "NM_model_genes= list()\n",
    "for gene in NM_model.genes:\n",
    "    NM_model_genes.append(gene.id)\n",
    "NM_model_kegg_orthologies=kegg_annotations[kegg_annotations.ids.isin(NM_model_genes)]\n",
    "\n",
    "\n",
    "NM_model_kegg_orthologies.to_csv('KEGG_orthologies_Nme.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "865eeadd-2382-4ac7-928f-d136cf75da10",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-14-081d2ed1840f>:17: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.\n",
      "  annotation_table['ProteinID'] = annotation_table['ProteinID'].str.replace('.', '_')\n",
      "<ipython-input-14-081d2ed1840f>:53: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  kegg_annotations['ids'] = kegg_annotations['ids'] .str.replace('_', '')\n"
     ]
    }
   ],
   "source": [
    "#Orthologies for Carveme\n",
    "#From https://www.biostars.org/p/492324/\n",
    "#Download Nme database of orthologies from KEGG\n",
    "\n",
    "#Get list of genes from Carveme model for FA1090 in NGO format\n",
    "carveme_model_genes= list()\n",
    "for gene in carveme_model.genes:\n",
    "    carveme_model_genes.append(gene.id)\n",
    "carveme_model_genes\n",
    "\n",
    "#Use look_back table of annotations from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5399841/\n",
    "annotation_table = pd.read_csv ('C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/annotation_resources/look_back_table.csv', usecols= ['Locus Tag ID','Protein.ID'])\n",
    "annotation_table=annotation_table.rename(columns={'Protein.ID': 'ProteinID', 'Locus Tag ID': \"NGO_ID\"})\n",
    "annotation_table\n",
    "                        \n",
    "#Match id format between kegg and model\n",
    "annotation_table['ProteinID'] = annotation_table['ProteinID'].str.replace('.', '_')\n",
    "\n",
    "annotatedgenes= annotation_table[annotation_table.ProteinID.isin(carveme_model_genes)]\n",
    "converted_model_genes= list()\n",
    "for tag in annotatedgenes['NGO_ID']:\n",
    "        converted_model_genes.append(tag) \n",
    "\n",
    "\n",
    "#Get KEGG orthologies for carveme genes\n",
    "database = list()\n",
    "for _, v in pd.read_json(\"C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/Figures/Fig_1/ngo00001.json\").iterrows():\n",
    "    d = v[\"children\"]\n",
    "    cat_1 = d[\"name\"]\n",
    "    for child_1 in d[\"children\"]:\n",
    "        cat_2 = child_1[\"name\"] # Module?\n",
    "        for child_2 in child_1[\"children\"]:\n",
    "            cat_3 = child_2[\"name\"]\n",
    "            if \"children\" in child_2:\n",
    "                for child_3 in child_2[\"children\"]:\n",
    "                    cat_4 = child_3[\"name\"]\n",
    "                    fields = [cat_1, cat_2, cat_3, cat_4]\n",
    "                    database.append(fields)\n",
    "df_kegg = pd.DataFrame(database, columns=[\"Level_A\", \"Level_B\", \"Level_C\", \"Level_D\"])\n",
    "\n",
    "kos = list()\n",
    "for tag in df_kegg[\"Level_D\"]:\n",
    "        full_tag=tag.split()\n",
    "        geneid=full_tag[0]\n",
    "        kos.append(geneid) \n",
    "\n",
    "df_kegg[\"ids\"]= kos\n",
    "\n",
    "#Include only kegg metabolism related annotations\n",
    "kegg_annotations=df_kegg[df_kegg[\"Level_A\"].str.contains(\"09100 Metabolism\")==True]\n",
    "\n",
    "#Match id format between kegg and model\n",
    "kegg_annotations['ids'] = kegg_annotations['ids'] .str.replace('_', '')\n",
    "\n",
    "#Filter kegg annotations by genes found in model\n",
    "\n",
    "carveme_model_kegg_orthologies=kegg_annotations[kegg_annotations.ids.isin(converted_model_genes)]\n",
    "\n",
    "\n",
    "carveme_model_kegg_orthologies.to_csv('KEGG_orthologies_carveme_FA1090.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "805d702b-54b7-47b5-9127-6fecdaaaf384",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ngo_558:\n",
      "09105 Amino acid metabolism                          115\n",
      "09108 Metabolism of cofactors and vitamins            96\n",
      "09101 Carbohydrate metabolism                         91\n",
      "09102 Energy metabolism                               59\n",
      "09104 Nucleotide metabolism                           45\n",
      "09107 Glycan biosynthesis and metabolism              43\n",
      "09103 Lipid metabolism                                27\n",
      "09110 Biosynthesis of other secondary metabolites     26\n",
      "09106 Metabolism of other amino acids                 26\n",
      "09109 Metabolism of terpenoids and polyketides        16\n",
      "09111 Xenobiotics biodegradation and metabolism        2\n",
      "Name: Level_B, dtype: int64\n",
      "\n",
      "\n",
      "Nmb_iTM560:\n",
      "09105 Amino acid metabolism                          114\n",
      "09101 Carbohydrate metabolism                         98\n",
      "09108 Metabolism of cofactors and vitamins            94\n",
      "09102 Energy metabolism                               72\n",
      "09107 Glycan biosynthesis and metabolism              50\n",
      "09104 Nucleotide metabolism                           48\n",
      "09110 Biosynthesis of other secondary metabolites     31\n",
      "09103 Lipid metabolism                                30\n",
      "09106 Metabolism of other amino acids                 30\n",
      "09109 Metabolism of terpenoids and polyketides        18\n",
      "09111 Xenobiotics biodegradation and metabolism        3\n",
      "Name: Level_B, dtype: int64\n",
      "\n",
      "\n",
      "carveme_FA1090:\n",
      "09105 Amino acid metabolism                          110\n",
      "09101 Carbohydrate metabolism                         90\n",
      "09108 Metabolism of cofactors and vitamins            83\n",
      "09104 Nucleotide metabolism                           43\n",
      "09107 Glycan biosynthesis and metabolism              41\n",
      "09102 Energy metabolism                               29\n",
      "09103 Lipid metabolism                                27\n",
      "09106 Metabolism of other amino acids                 24\n",
      "09110 Biosynthesis of other secondary metabolites     23\n",
      "09109 Metabolism of terpenoids and polyketides        16\n",
      "09111 Xenobiotics biodegradation and metabolism        2\n",
      "Name: Level_B, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#Trim out sublevels of metabolism categories not being used for all three models\n",
    "trimmed_model_kegg_orthologies=model_kegg_orthologies.drop([\"Level_C\", \"Level_D\"], axis=1)\n",
    "trimmed_NM_model_kegg_orthologies=NM_model_kegg_orthologies.drop([\"Level_C\", \"Level_D\"], axis=1)\n",
    "trimmed_carveme_model_kegg_orthologies=carveme_model_kegg_orthologies.drop([\"Level_C\", \"Level_D\"], axis=1)\n",
    "\n",
    "#drop duplicate genes that are listed twice within a single kegg category\n",
    "pruned_and_trimmed_model_kegg_orthologies=trimmed_model_kegg_orthologies.drop_duplicates()\n",
    "pruned_and_trimmed_NM_model_kegg_orthologies=trimmed_NM_model_kegg_orthologies.drop_duplicates()\n",
    "pruned_and_trimmed_carveme_model_kegg_orthologies=trimmed_carveme_model_kegg_orthologies.drop_duplicates()\n",
    "\n",
    "#Report counts of unique genes present in each category (genes may appear in more than one category)\n",
    "print(\"Ngo_558:\")\n",
    "print(pruned_and_trimmed_model_kegg_orthologies['Level_B'].value_counts())\n",
    "print(\"\\n\")\n",
    "print(\"Nmb_iTM560:\")\n",
    "print(pruned_and_trimmed_NM_model_kegg_orthologies['Level_B'].value_counts())\n",
    "print(\"\\n\")\n",
    "print(\"carveme_FA1090:\")\n",
    "print(pruned_and_trimmed_carveme_model_kegg_orthologies['Level_B'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a9b36c04-ae7d-41e9-afa5-470b89452965",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ngo_558:\n",
      "total genes in model: 560\n",
      "unique genes with kegg ids: 399\n",
      "unannotated genes: 161\n",
      "\n",
      "\n",
      "Nmb_iTM560:\n",
      "total genes in model: 562\n",
      "unique genes with kegg ids: 413\n",
      "unannotated genes: 149\n",
      "\n",
      "\n",
      "carveme_FA1090:\n",
      "total genes in model: 510\n",
      "unique genes with kegg ids: 347\n",
      "unannotated genes: 163\n"
     ]
    }
   ],
   "source": [
    "#Number of Genes with unidentified metabolic functions in KEGG\n",
    "print(\"Ngo_558:\")\n",
    "#Total Number of genes in the model\n",
    "print(\"total genes in model:\", len(model_genes))\n",
    "#total number of unique genes with Kegg ides\n",
    "print(\"unique genes with kegg ids:\", model_kegg_orthologies.ids.nunique())\n",
    "#Number of genes in the model without metabolism linked Kegg IDs\n",
    "print(\"unannotated genes:\",len(model_genes) - model_kegg_orthologies.ids.nunique())\n",
    "\n",
    "print(\"\\n\")\n",
    "print(\"Nmb_iTM560:\")\n",
    "#Total Number of genes in the model\n",
    "print(\"total genes in model:\", len(NM_model_genes))\n",
    "#total number of unique genes with Kegg ides\n",
    "print(\"unique genes with kegg ids:\", NM_model_kegg_orthologies.ids.nunique())\n",
    "#Number of genes in the model without metabolism linked Kegg IDs\n",
    "print(\"unannotated genes:\",len(NM_model_genes) - NM_model_kegg_orthologies.ids.nunique())\n",
    "\n",
    "print(\"\\n\")\n",
    "print(\"carveme_FA1090:\")\n",
    "#Total Number of genes in the model\n",
    "print(\"total genes in model:\", len(carveme_model_genes))\n",
    "#total number of unique genes with Kegg ides\n",
    "print(\"unique genes with kegg ids:\", carveme_model_kegg_orthologies.ids.nunique())\n",
    "#Number of genes in the model without metabolism linked Kegg IDs\n",
    "print(\"unannotated genes:\",len(carveme_model_genes) - carveme_model_kegg_orthologies.ids.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c5864e1c-6add-4e34-8750-f9fe607090d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Possible Alternative?\n",
    "#https://widdowquinn.github.io/2018-03-06-ibioic/02-sequence_databases/09-KEGG_programming.html\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3bc2391-6524-4e41-b281-0bc2c7c14ba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions\n",
    "NM_model.reactions\n",
    "carvememodel.reactions"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
