{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "49c4429d-c3cb-4354-9ae1-36987fe42c19",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception reporting mode: Verbose\n"
     ]
    }
   ],
   "source": [
    "# Import libraries - REQUIRES pip version 9.0.3\n",
    "import pandas\n",
    "import pandas as pd\n",
    "import os\n",
    "from os.path import join\n",
    "import sys\n",
    "import scipy.stats\n",
    "import numpy\n",
    "import math\n",
    "import pickle\n",
    "import copy\n",
    "import time\n",
    "import warnings\n",
    "import gc\n",
    "\n",
    "# Using Cobrapy 0.13.0\n",
    "import cobra\n",
    "import cobra.test\n",
    "import cobra.flux_analysis.gapfilling\n",
    "from cobra import Reaction, Metabolite\n",
    "from cobra.io import write_sbml_model\n",
    "from cobra.flux_analysis import  flux_variability_analysis\n",
    "from cobra.flux_analysis.reaction import assess_component\n",
    "from cobra.manipulation.delete import *\n",
    "from cobra.flux_analysis.parsimonious import add_pfba\n",
    "from cobra.medium import find_boundary_types\n",
    "from cobra.util import solver as sutil\n",
    "\n",
    "\n",
    "# Confidence levels in each annotation\n",
    "#import probanno\n",
    "\n",
    "# Estabish handler for logger\n",
    "import logging\n",
    "logging.basicConfig()\n",
    "logger = logging.getLogger('logger')\n",
    "\n",
    "# Verbose exception printing\n",
    "%xmode\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "595573f3-3072-4a1e-9e73-727978105f43",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Define functions\n",
    "\n",
    "# Identify potentially gapfilled reactions\n",
    "def findGapfilledRxn(model, exclude=[]):\n",
    "    gapfilled = []\n",
    "    transport = findTransports(model)\n",
    "    if not type(exclude) is list:\n",
    "        exclude = [exclude]\n",
    "        \n",
    "    for index in model.reactions:\n",
    "        if len(list(index.genes)) == 0:\n",
    "            if not index in model.boundary:\n",
    "                if not index.id in exclude or not index.id in transport:\n",
    "                    gapfilled.append(index.id)\n",
    "    \n",
    "    if len(gapfilled) > 0:\n",
    "        print(str(len(gapfilled)) + ' metabolic reactions not associated with genes')\n",
    "    \n",
    "    return gapfilled\n",
    "\n",
    "# Check for missing transport and exchange reactions\n",
    "def missingRxns(model, extracellular='e'):\n",
    "\n",
    "    transporters = set(findTransports(model))\n",
    "    exchanges = set([x.id for x in model.exchanges])\n",
    "    \n",
    "    missing_exchanges = []\n",
    "    missing_transports = []\n",
    "    \n",
    "    for metabolite in model.metabolites:\n",
    "        if not metabolite.compartment == extracellular:\n",
    "            continue\n",
    "\n",
    "        curr_rxns = set([x.id for x in list(metabolite.reactions)])\n",
    "        \n",
    "        if not bool(curr_rxns & transporters):\n",
    "            missing_transports.append(metabolite.id)\n",
    "        if not bool(curr_rxns & exchanges):\n",
    "            missing_exchanges.append(metabolite.id)\n",
    "    \n",
    "    if len(missing_transports) != 0:\n",
    "        print(str(len(missing_transports)) + ' extracellular metabolites are missing transport reactions')\n",
    "    if len(missing_exchanges) != 0:\n",
    "        print(str(len(missing_exchanges)) + ' extracellular metabolites are missing exchange reactions')\n",
    "        \n",
    "    return missing_transports, missing_exchanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5a3962c7-590a-4628-a5c7-6ab1f9a8800e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "nlog2 = math.log(2)\n",
    "\n",
    "def doubling_time(model):\n",
    "    with model as m:\n",
    "        ov = m.slim_optimize()\n",
    "        if ov < 1e-6:\n",
    "            growth = 'No growth'\n",
    "        else:  \n",
    "            growth = nlog2*60/ov\n",
    "            growth = str(round(growth, 3)) + ' minutes'\n",
    "        \n",
    "    return growth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9bbfe627-557c-440e-98ac-912eea43759c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Checks which cytosolic metabolites are generated for free (bacteria only)\n",
    "def checkFreeMass(raw_model, cytosol='cytosol'):\n",
    "    \n",
    "    with raw_model as model:\n",
    "        \n",
    "        # Close all exchanges\n",
    "        for index in model.boundary:\n",
    "            model.reactions.get_by_id(index.id).lower_bound = 0.\n",
    "        \n",
    "        # Identify all metabolites that are produced within the network\n",
    "        demand_metabolites = [x.reactants[0].id for x in model.demands if len(x.reactants) > 0]\n",
    "        demand_metabolites += [x.products[0].id for x in model.demands if len(x.products) > 0]\n",
    "\n",
    "        free = []\n",
    "        for index in model.metabolites: \n",
    "            if index.id in demand_metabolites:\n",
    "                continue\n",
    "            elif not index.compartment in cytosol:\n",
    "                continue\n",
    "            else:\n",
    "                demand = model.add_boundary(index, type='demand')\n",
    "                model.objective = demand\n",
    "                obj_val = model.slim_optimize(error_value=0.)\n",
    "                if obj_val > 1e-8:\n",
    "                    free.append(index.id)\n",
    "                model.remove_reactions([demand])\n",
    "    \n",
    "    if len(free) > 0:\n",
    "        print(str(len(free)) + ' metabolites are generated for free')\n",
    "\n",
    "    return(free)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c970a7f-4a39-49c9-8f99-cdb4d45acac3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b90b89ed-169a-41fa-af86-5d13c33f7810",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Check for mass and charge balance in reactions\n",
    "def checkBalance(raw_model, exclude=[]):\n",
    "    \n",
    "    with raw_model as model:\n",
    "        imbalanced = []\n",
    "        mass_imbal = 0\n",
    "        charge_imbal = 0\n",
    "        elem_set = set()\n",
    "        for metabolite in model.metabolites:\n",
    "            try:\n",
    "                elem_set |= set(metabolite.elements.keys())\n",
    "            except:\n",
    "                pass\n",
    "        \n",
    "        if len(elem_set) == 0:\n",
    "            imbalanced = model.reactions\n",
    "            mass_imbal = len(model.reactions)\n",
    "            charge_imbal = len(model.reactions)\n",
    "            print('No elemental data associated with metabolites!')\n",
    "        \n",
    "        else:\n",
    "            if not type(exclude) is list: \n",
    "                exclude = [exclude]\n",
    "            for index in model.reactions:\n",
    "                if index in model.boundary or index.id in exclude:\n",
    "                    continue\n",
    "\n",
    "                else:\n",
    "                    try:\n",
    "                        test = index.check_mass_balance()\n",
    "                    except ValueError:\n",
    "                        continue\n",
    "\n",
    "                    if len(list(test)) > 0:\n",
    "                        imbalanced.append(index.id)\n",
    "\n",
    "                        if 'charge' in test.keys():\n",
    "                            charge_imbal += 1\n",
    "                        if len(set(test.keys()).intersection(elem_set)) > 0:\n",
    "                            mass_imbal += 1\n",
    "\n",
    "    if mass_imbal != 0:\n",
    "        print(str(mass_imbal) + ' reactions are mass imbalanced')\n",
    "    if charge_imbal != 0:\n",
    "        print(str(charge_imbal) + ' reactions are charge imbalanced')\n",
    "    return(imbalanced)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "430b2e13-9fb8-49e9-b3ee-b5302537c21d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Checks the quality of models by a couple metrics and returns problems\n",
    "def checkQuality(model, exclude=[], cytosol='c'):\n",
    "    \n",
    "    start_time = time.time()\n",
    "    \n",
    "    if model.name != None:\n",
    "        model_name = model.name\n",
    "    else:\n",
    "        model_name = 'model'\n",
    "    \n",
    "    gaps = findGapfilledRxn(model, exclude)\n",
    "    freemass = checkFreeMass(model, cytosol)\n",
    "    balance = checkBalance(model, exclude)\n",
    "    trans, exch = missingRxns(model)\n",
    "    \n",
    "    test = gaps + freemass + balance\n",
    "    if len(test) == 0:\n",
    "        print('No inconsistencies detected')\n",
    "    \n",
    "    duration = int(round(time.time() - start_time))\n",
    "    print('Took ' + str(duration) + ' seconds to analyze ' + model_name) \n",
    "\n",
    "    return gaps, freemass, balance, trans, exch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c571510a-488c-4878-8006-5db55ed16fea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Trace back through reactions immediately adjacent to a given reaction to identify blocked precursor synthesis\n",
    "def checkPrecursors(model, reaction):\n",
    "    \n",
    "    if isinstance(reaction, str) == True:\n",
    "        reaction = model.reactions.get_by_id(reaction)\n",
    "    \n",
    "    model.objective = reaction\n",
    "    obj_val = max(model.optimize(objective_sense='maximize').objective_value, abs(model.optimize(objective_sense='minimize').objective_value))\n",
    "\n",
    "    if obj_val > 0.001:\n",
    "        print('Able to produce all precursors for this reaction.')\n",
    "        return None\n",
    "    \n",
    "    else:\n",
    "        reactants = reaction.reactants\n",
    "        check = 0\n",
    "        \n",
    "        for reactant in reactants:\n",
    "            sub_reactions = list(reactant.reactions)\n",
    "            \n",
    "            for sub_reaction in sub_reactions:\n",
    "                model.objective = sub_reaction\n",
    "                obj_val = max(model.optimize(objective_sense='maximize').objective_value, abs(model.optimize(objective_sense='minimize').objective_value))\n",
    "\n",
    "                if obj_val < 0.001 and reactant in sub_reaction.products:                    \n",
    "                    print('Cannot acquire ' + str(reactant.id) + ' via ' + str(sub_reaction.id))\n",
    "                elif obj_val < 0.001 and check < 1 and reactant in sub_reaction.reactants:\n",
    "                    print(str(reactant.id) + ' not produced in any reactions.')\n",
    "                    check += 1\n",
    "\n",
    "#------------------------------------------------------------------------------------#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2d1fcf26-bcd5-4bf5-a9a3-009bdb0747da",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Function to calculate doubling time from objective value\n",
    "def doubling(model):\n",
    "    \n",
    "    with model as m:\n",
    "        ov = m.slim_optimize()\n",
    "        if ov < 1e-6:\n",
    "            growth = 'No growth'\n",
    "        else:\n",
    "            growth = (1 / float(m.slim_optimize())) * 3600\n",
    "        \n",
    "            if growth < 60.0:\n",
    "                growth = str(round(growth, 3)) + ' minutes'\n",
    "            else:\n",
    "                growth = growth / 60\n",
    "                growth = str(round(growth, 3)) + ' hours'\n",
    "            \n",
    "    return growth\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "56c88f0e-4f64-41c5-93ff-d5e1b0d8d1c0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Checks for availability of reactants of a given reaction\n",
    "def availability(model, target_rxn):\n",
    "\n",
    "    precursors = model.reactions.get_by_id(target_rxn).reactants\n",
    "    \n",
    "    total = 0\n",
    "    unsuccessful = set()\n",
    "    limited = set()\n",
    "    for precursor in precursors:\n",
    "        precursor_rxn = list(model.metabolites.get_by_id(precursor.id).reactions)\n",
    "        \n",
    "        for rxn in precursor_rxn:\n",
    "            if rxn.id == target_rxn:\n",
    "                continue\n",
    "            elif precursor in model.reactions.get_by_id(rxn.id).reactants:\n",
    "                \n",
    "                model.objective = rxn\n",
    "                obj_val = model.slim_optimize()\n",
    "                if obj_val < 1e-8:\n",
    "                    unsuccessful |= set([rxn.id])\n",
    "                    limited |= set([precursor.id])\n",
    "    \n",
    "    print('Failed reactions: ' + str(len(unsuccessful)))\n",
    "    print('Limiting reactants: ' + str(len(limited)))\n",
    "    \n",
    "    return unsuccessful, limited\n",
    "\n",
    "# Removes all metabolites in a list of metabolite ids and all reactions associated with them\n",
    "def removeAll(model, metabolite_list):\n",
    "    \n",
    "    new_model = copy.deepcopy(model)\n",
    "    \n",
    "    for metabolite in metabolite_list:\n",
    "        try:\n",
    "            metabolite = new_model.metabolites.get_by_id(metabolite)\n",
    "            new_model.remove_reactions(metabolite.reactions)\n",
    "            new_model.remove_metabolites([metabolite])\n",
    "        except KeyError:\n",
    "            print(metabolite + ' not found')\n",
    "            continue\n",
    "        \n",
    "    return new_model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "072d32d2-42e2-48bf-824f-45fc07e14d80",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Identify transport reactions (for any number compartments)\n",
    "def findTransports(model):\n",
    "    transporters = []\n",
    "    compartments = set(list(model.compartments))\n",
    "    if len(compartments) == 1:\n",
    "        raise Exception('Model only has one compartment!')\n",
    "    \n",
    "    for reaction in model.reactions:\n",
    "        \n",
    "        reactant_compartments = set([x.compartment for x in reaction.reactants])\n",
    "        product_compartments = set([x.compartment for x in reaction.products])\n",
    "        reactant_baseID = set([x.id.split('_')[0] for x in reaction.reactants])\n",
    "        product_baseID = set([x.id.split('_')[0] for x in reaction.products])\n",
    "        \n",
    "        if reactant_compartments == product_compartments and reactant_baseID != product_baseID:\n",
    "            continue\n",
    "        elif bool(compartments & reactant_compartments) == True and bool(compartments & product_compartments) == True:\n",
    "            transporters.append(reaction.id)\n",
    "        \n",
    "    return(transporters)\n",
    "\n",
    "# Removes a given percentage of reactions from a model, ignoring objective\n",
    "def generate_gaps(model, percentage=0.2, prune=False, ignore=[]):\n",
    "    number_to_remove = int(round(len(model.reactions) * percentage))\n",
    "    rxn_ids = [x.id for x in model.reactions]\n",
    "    random.shuffle(rxn_ids)\n",
    "    rxns_to_remove = rxn_ids[-number_to_remove:]\n",
    "        \n",
    "    for rxn in ignore:\n",
    "        try:\n",
    "            rxns_to_remove.remove(rxn)\n",
    "        except ValueError:\n",
    "            continue\n",
    "    \n",
    "    truncated_model = copy.deepcopy(model)\n",
    "    truncated_model.remove_reactions(rxns_to_remove)\n",
    "    \n",
    "    if prune == True:\n",
    "        unused_cpds = prune_unused_metabolites(truncated_model)\n",
    "    \n",
    "    print('Reactions removed: ' + str(len(rxns_to_remove)))\n",
    "    print('New objective value: ' + str(truncated_model.slim_optimize()))\n",
    "    return truncated_model, rxns_to_remove\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e99f2b0e-960d-4d5d-8dc4-cf27d621b8eb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Calculates the sum of fluxes for a given model\n",
    "def sum_of_fluxes(model):\n",
    "    \n",
    "    with model as m:\n",
    "        solution = m.optimize()\n",
    "        flux_sum = sum(list(solution.fluxes))\n",
    "        \n",
    "    return flux_sum\n",
    "\n",
    "\n",
    "# reports the reaction ids that are in 1 model but not the other\n",
    "def rxn_diff(model_1, model_2):\n",
    "    diff = set([str(x.id) for x in model_1.reactions]).difference([str(y.id) for y in model_2.reactions])\n",
    "    return list(diff)\n",
    "\n",
    "# Check for missing transport and exchange reactions\n",
    "def missingRxn(model, extracellular='extracellular'):\n",
    "\n",
    "    transporters = set(findTransports(model))\n",
    "    exchanges = set([x.id for x in model.exchanges])\n",
    "    \n",
    "    missing_exchanges = []\n",
    "    missing_transports = []\n",
    "    \n",
    "    for metabolite in model.metabolites:\n",
    "        if not metabolite.compartment == extracellular:\n",
    "            continue\n",
    "\n",
    "        curr_rxns = set([x.id for x in list(metabolite.reactions)])\n",
    "        \n",
    "        if not bool(curr_rxns & transporters):\n",
    "            missing_transports.append(metabolite.id)\n",
    "        if not bool(curr_rxns & exchanges):\n",
    "            missing_exchanges.append(metabolite.id)\n",
    "    \n",
    "    if len(missing_transports) != 0:\n",
    "        print(str(len(missing_transports)) + ' extracellular metabolites are missing transport reactions')\n",
    "    if len(missing_exchanges) != 0:\n",
    "        print(str(len(missing_exchanges)) + ' extracellular metabolites are missing exchange reactions')\n",
    "        \n",
    "    return missing_transports, missing_exchanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e6c6d365-ad81-4970-8b54-9b6a36a24587",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Quicker way to read in models\n",
    "import pickle\n",
    "def read_model(fileName, obj='none'):\n",
    "    \n",
    "    fileType = fileName.split('.')[-1]\n",
    "    \n",
    "    if fileType == 'sbml' or fileType == 'xml':\n",
    "        model = cobra.io.read_sbml_model(fileName)\n",
    "    elif fileType == 'json':\n",
    "        model = cobra.io.load_json_model(fileName)\n",
    "    elif fileType == 'yaml':\n",
    "        model = cobra.io.load_yaml_model(fileName)\n",
    "    elif fileType == 'mat':\n",
    "        model = cobra.io.load_matlab_model(fileName)\n",
    "    elif fileType == 'pkl':\n",
    "        model = pickle.load(open(fileName, 'rb'))\n",
    "    else:\n",
    "        raise TypeError('Unrecognized file extension')\n",
    "    \n",
    "    if obj != 'none': model.objective = obj\n",
    "    for rxn in model.boundary: rxn.bounds = (-1000., 1000.)\n",
    "        \n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8d93ab32-fc79-4bed-875d-a4db2ed22a45",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set medias\n",
    "rpmi_req=set(['EX_apoACP_c_', 'EX_trdrd_c_','EX_cobalt2_e_','EX_cu2_e_','EX_fe3_e_','EX_mn2_e_','EX_mobd_e_','EX_zn2_e_','EX_ca2_e_','EX_no3_e_','EX_h2o_e_','EX_k_e_','EX_cl_e_','EX_mg2_e_','EX_so4_e_','EX_na1_e_','EX_pi_e_','EX_h_e_','EX_arg_L_e_','EX_asn_L_e_','EX_asp_L_e_','EX_cys_L_e_','EX_glu_L_e_','EX_gly_e_','EX_his_L_e_','EX_ile_L_e_','EX_leu_L_e_','EX_lys_L_e_','EX_met_L_e_','EX_phe_L_e_','EX_pro_L_e_','EX_ser_L_e_','EX_thr_L_e_','EX_trp_L_e_','EX_tyr_L_e_','EX_val_L_e_','EX_pnto_R_e_','EX_chol_e_','EX_inost_e_','EX_glc_D_e_','EX_gthrd_e_','EX_co2_e_'])\n",
    "mdm_req=set(['EX_apoACP_c_', 'EX_trdrd_c_','EX_co2_e_', 'EX_cobalt2_e_', 'EX_cu2_e_' , 'EX_h_e_', 'EX_h2o_e_','EX_mn2_e_','EX_mobd_e_','EX_tungs_e_', 'EX_zn2_e_','EX_na1_e_','EX_cl_e_','EX_k_e_','EX_so4_e_','EX_nh4_e_','EX_mg2_e_','EX_pi_e_','EX_ca2_e_','EX_fe3_e_','EX_no3_e_','EX_asp_L_e_','EX_glu_L_e_','EX_arg_L_e_','EX_gly_e_','EX_ser_L_e_','EX_leu_L_e_','EX_ile_L_e_','EX_val_L_e_','EX_tyr_L_e_','EX_cys_L_e_','EX_pro_L_e_','EX_trp_L_e_','EX_thr_L_e_','EX_phe_L_e_','EX_asn_L_e_','EX_gln_L_e_','EX_his_L_e_','EX_met_L_e_','EX_ala_L_e_','EX_lys_L_e_','EX_gthrd_e_','EX_thm_e_','EX_pnto_R_e_','EX_glc_D_e_','EX_hxan_e_','EX_ura_e_'])\n",
    "\n",
    "def rpmi(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "\n",
    "def complete(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=-20.\n",
    "\n",
    "def mdm(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -20.\n",
    "        elif reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "        elif reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "\n",
    "def rpmi_molarity(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -20. \n",
    "        if 'EX_glc_D_e_' in reaction.id:\n",
    "            reaction.lower_bound = -11.1012\n",
    "        if 'EX_ser_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.2855\n",
    "        if 'EX_asp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1503\n",
    "        if 'EX_asn_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.3784\n",
    "        if 'EX_glu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1359\n",
    "        if 'EX_gly_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1332\n",
    "        if 'EX_val_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1707\n",
    "        if 'EX_leu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.3812\n",
    "        if 'EX_pro_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1737\n",
    "        if 'EX_phe_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.0908\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.219\n",
    "        if 'EX_ca2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.424\n",
    "        if 'EX_no3_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.847\n",
    "        if 'EX_k_e_' in reaction.id:\n",
    "            reaction.lower_bound = -5.3655\n",
    "        if 'EX_cl_e_' in reaction.id:\n",
    "            reaction.lower_bound = -108.6993\n",
    "        if 'EX_mg2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.4056\n",
    "        if 'EX_so4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.4056\n",
    "        if 'EX_na1_e_' in reaction.id:\n",
    "            reaction.lower_bound = -137.9818\n",
    "        if 'EX_pi_e_' in reaction.id:\n",
    "            reaction.lower_bound = -6.354\n",
    "        if 'EX_arg_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.1481\n",
    "        if 'EX_cys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -10.416\n",
    "        if 'EX_his_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.0967\n",
    "        if 'EX_ile_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.3812\n",
    "        if 'EX_met_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1005\n",
    "        if 'EX_thr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1679\n",
    "        if 'EX_trp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.0245\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.219\n",
    "        if 'EX_tyr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1104\n",
    "        if 'EX_pnto_R_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.001\n",
    "        if 'EX_chol_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.0215\n",
    "        if 'EX_inost_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1943\n",
    "        if 'EX_gthrd_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.0033\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -20."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b5e8ec30-61c6-4927-8136-efa58e30a42d",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Read in  model\n",
    "import cobra\n",
    "model=read_model('C:/Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/AE004969.1.json')\n",
    "\n",
    "# Read in  other models to assist with annotation\n",
    "NM_model = read_model('C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/Nmb_iTM560.json')\n",
    "\n",
    "#Ecoli model NM model is based on - to be used for metabolite charges\n",
    "ecoli = read_model('C:/Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/curated_models/ecoli.json')\n",
    "\n",
    "#Universal model for adding reactions\n",
    "universalmodel= read_model('C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/universal_model_cobrapy.json')\n",
    "\n",
    "#Downloaded carvememodel for Neisseria gonorrhoeae FA1090 from PMID: 30192979 to compare annotations and add Gc specific reactions\n",
    "carveme_model= cobra.io.read_sbml_model('C:/Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/Neisseria_gonorrhoeae_FA_1090.xml.gz')\n",
    "carvememodel=carveme_model\n",
    "cobra.io.save_json_model(carveme_model, \"C:/Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/Neisseria_gonorrhoeae_FA_1090.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "87ee6c25-73f3-4925-934b-d21f16bec6a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "rpmi_req=set(['EX_apoACP_c_', 'EX_trdrd_c_','EX_cobalt2_e_','EX_cu2_e_','EX_fe3_e_','EX_mn2_e_','EX_mobd_e_','EX_zn2_e_','EX_ca2_e_','EX_no3_e_','EX_h2o_e_','EX_k_e_','EX_cl_e_','EX_mg2_e_','EX_so4_e_','EX_na1_e_','EX_pi_e_','EX_h_e_','EX_arg_L_e_','EX_asn_L_e_','EX_asp_L_e_','EX_cys_L_e_','EX_glu_L_e_','EX_gly_e_','EX_his_L_e_','EX_ile_L_e_','EX_leu_L_e_','EX_lys_L_e_','EX_met_L_e_','EX_phe_L_e_','EX_pro_L_e_','EX_ser_L_e_','EX_thr_L_e_','EX_trp_L_e_','EX_tyr_L_e_','EX_val_L_e_','EX_pnto_R_e_','EX_chol_e_','EX_inost_e_','EX_glc_D_e_','EX_gthrd_e_','EX_co2_e_'])\n",
    "\n",
    "def rpmi_with_lactate_only(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -0.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "            \n",
    "def rpmi(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -20.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "95818e08-3e1c-426a-8874-75101a21f092",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 56.068</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12DGR120tipp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR140tipp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR141tipp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR160tipp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR161tipp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XTSNtex</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLtex</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLUtex</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN2abcpp</th>\n",
       "      <td>0.177062</td>\n",
       "      <td>1.734723e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zn2tex</th>\n",
       "      <td>0.177062</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1484 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 56.068 at 0x19cd7a60040>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def complete(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=-1000.\n",
    "            \n",
    "complete(model)\n",
    "model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b73ae5b0-4464-4515-b9aa-098e08d507f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NGO0057', 'NGO0059', 'NGO0063', 'NGO0084', 'NGO0085', 'NGO0114', 'NGO0176', 'NGO0177', 'NGO0186', 'NGO0188', 'NGO0189', 'NGO0199', 'NGO0231', 'NGO0294', 'NGO0351', 'NGO0382', 'NGO0387', 'NGO0388', 'NGO0541', 'NGO0567', 'NGO0571', 'NGO0604', 'NGO0608', 'NGO0632', 'NGO0633', 'NGO0652', 'NGO0653', 'NGO0659', 'NGO0676', 'NGO0718', 'NGO0763', 'NGO0829', 'NGO0830', 'NGO0842', 'NGO0858', 'NGO0886', 'NGO0902', 'NGO0904', 'NGO0905', 'NGO0906', 'NGO0934', 'NGO0941', 'NGO0974', 'NGO1032', 'NGO1033', 'NGO1047', 'NGO1052', 'NGO1055', 'NGO1154', 'NGO1204', 'NGO1207', 'NGO1226', 'NGO1227', 'NGO1243', 'NGO1298', 'NGO1323', 'NGO1325', 'NGO1334', 'NGO1364', 'NGO1365', 'NGO1381', 'NGO1404', 'NGO1432', 'NGO1433', 'NGO1434', 'NGO1454', 'NGO1530', 'NGO1559', 'NGO1577', 'NGO1684', 'NGO1686', 'NGO1692', 'NGO1722', 'NGO1732', 'NGO1733', 'NGO1753', 'NGO1769', 'NGO1858', 'NGO1866', 'NGO1922', 'NGO1930', 'NGO1963', 'NGO1977', 'NGO1990', 'NGO2011', 'NGO2012', 'NGO2013', 'NGO2014', 'NGO2015', 'NGO2027', 'NGO2033', 'NGO2040', 'NGO2055', 'NGO2100', 'NGO2113', 'NGO2135', 'NGO2171', 'NGO2172']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-17-4a5b6254db06>:18: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.\n",
      "  annotation_table['ProteinID'] = annotation_table['ProteinID'].str.replace('.', '_')\n"
     ]
    }
   ],
   "source": [
    "#Identify genes/reactions to add to the model from carveme\n",
    "model_genes  = []\n",
    "for gene in model.genes:\n",
    "    model_genes.append(gene.id)\n",
    "\n",
    "#Get list of genes from Carveme model for FA1090 in NGO format\n",
    "carveme_model_genes= list()\n",
    "for gene in carveme_model.genes:\n",
    "    carveme_model_genes.append(gene.id)\n",
    "carveme_model_genes\n",
    "\n",
    "#Use look_back table of annotations from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5399841/\n",
    "annotation_table = pd.read_csv ('C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/annotation_resources/look_back_table.csv', usecols= ['Locus Tag ID','Protein.ID'])\n",
    "annotation_table=annotation_table.rename(columns={'Protein.ID': 'ProteinID', 'Locus Tag ID': \"NGO_ID\"})\n",
    "annotation_table\n",
    "                        \n",
    "#Match id format between kegg and model\n",
    "annotation_table['ProteinID'] = annotation_table['ProteinID'].str.replace('.', '_')\n",
    "\n",
    "annotatedgenes= annotation_table[annotation_table.ProteinID.isin(carveme_model_genes)]\n",
    "converted_carveme_model_genes= list()\n",
    "for tag in annotatedgenes['NGO_ID']:\n",
    "        converted_carveme_model_genes.append(tag)\n",
    "        \n",
    "unique_genes_carveme = set(converted_carveme_model_genes) - set(model_genes) # correct elements, but not yet in sorted order\n",
    "print(sorted(unique_genes_carveme)) # sort and print\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "012c0bba-df82-4c2f-ae0d-982919a3b0bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add NGOIDs to the carveme model as an annotation\n",
    "for gene in annotatedgenes['ProteinID']:\n",
    "    if gene in carveme_model.genes:\n",
    "        row=annotatedgenes.index[annotatedgenes['ProteinID']==gene]\n",
    "        id=annotatedgenes['NGO_ID']\n",
    "        annotation =id.loc[row].item()\n",
    "        carveme_model.genes.get_by_id(gene).annotation['NGO_ID']= annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7e4162b5-728d-4a88-9ac6-31ab5d9d21af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>carvemegenes</th>\n",
       "      <th>NGO_IDs</th>\n",
       "      <th>carvemereactions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YP_208072_1</td>\n",
       "      <td>{'NGO_ID': 'NGO0974'}</td>\n",
       "      <td>[2AGPEAT181: 2agpe181_c + atp_c + ocdcea_c --&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YP_208834_1</td>\n",
       "      <td>{'NGO_ID': 'NGO1804'}</td>\n",
       "      <td>[3HAD160: 3hpalmACP_c --&gt; h2o_c + tpalm2eACP_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YP_209169_1</td>\n",
       "      <td>{'NGO_ID': 'NGO2163'}</td>\n",
       "      <td>[3OAR180: 3ooctdACP_c + h_c + nadph_c &lt;=&gt; 3hoc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YP_208796_1</td>\n",
       "      <td>{'NGO_ID': 'NGO1763'}</td>\n",
       "      <td>[3OAS60: butACP_c + h_c + malACP_c --&gt; 3ohexAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YP_209174_1</td>\n",
       "      <td>{'NGO_ID': 'NGO2168'}</td>\n",
       "      <td>[3OAS160: h_c + malACP_c + myrsACP_c --&gt; 3opal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>YP_207518_1</td>\n",
       "      <td>{'NGO_ID': 'NGO0362'}</td>\n",
       "      <td>[UPPDC1: 4.0 h_c + uppg3_c --&gt; 4.0 co2_c + cpp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>YP_207509_1</td>\n",
       "      <td>{'NGO_ID': 'NGO0353'}</td>\n",
       "      <td>[UPPRT: prpp_c + ura_c --&gt; ppi_c + ump_c]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>507</th>\n",
       "      <td>YP_207732_1</td>\n",
       "      <td>{'NGO_ID': 'NGO0589'}</td>\n",
       "      <td>[URAt: ura_e &lt;=&gt; ura_c, URAt2: h_e + ura_e --&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>YP_208502_1</td>\n",
       "      <td>{'NGO_ID': 'NGO1448'}</td>\n",
       "      <td>[USHD: h2o_c + u23ga_c --&gt; 2.0 h_c + lipidX_c ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509</th>\n",
       "      <td>YP_207255_1</td>\n",
       "      <td>{'NGO_ID': 'NGO0082'}</td>\n",
       "      <td>[VPAMTr: 3mob_c + ala__L_c --&gt; pyr_c + val__L_c]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>510 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    carvemegenes                NGO_IDs  \\\n",
       "0    YP_208072_1  {'NGO_ID': 'NGO0974'}   \n",
       "1    YP_208834_1  {'NGO_ID': 'NGO1804'}   \n",
       "2    YP_209169_1  {'NGO_ID': 'NGO2163'}   \n",
       "3    YP_208796_1  {'NGO_ID': 'NGO1763'}   \n",
       "4    YP_209174_1  {'NGO_ID': 'NGO2168'}   \n",
       "..           ...                    ...   \n",
       "505  YP_207518_1  {'NGO_ID': 'NGO0362'}   \n",
       "506  YP_207509_1  {'NGO_ID': 'NGO0353'}   \n",
       "507  YP_207732_1  {'NGO_ID': 'NGO0589'}   \n",
       "508  YP_208502_1  {'NGO_ID': 'NGO1448'}   \n",
       "509  YP_207255_1  {'NGO_ID': 'NGO0082'}   \n",
       "\n",
       "                                      carvemereactions  \n",
       "0    [2AGPEAT181: 2agpe181_c + atp_c + ocdcea_c -->...  \n",
       "1    [3HAD160: 3hpalmACP_c --> h2o_c + tpalm2eACP_c...  \n",
       "2    [3OAR180: 3ooctdACP_c + h_c + nadph_c <=> 3hoc...  \n",
       "3    [3OAS60: butACP_c + h_c + malACP_c --> 3ohexAC...  \n",
       "4    [3OAS160: h_c + malACP_c + myrsACP_c --> 3opal...  \n",
       "..                                                 ...  \n",
       "505  [UPPDC1: 4.0 h_c + uppg3_c --> 4.0 co2_c + cpp...  \n",
       "506          [UPPRT: prpp_c + ura_c --> ppi_c + ump_c]  \n",
       "507  [URAt: ura_e <=> ura_c, URAt2: h_e + ura_e -->...  \n",
       "508  [USHD: h2o_c + u23ga_c --> 2.0 h_c + lipidX_c ...  \n",
       "509   [VPAMTr: 3mob_c + ala__L_c --> pyr_c + val__L_c]  \n",
       "\n",
       "[510 rows x 3 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "carvemegenes = []\n",
    "carvemengoids=[]\n",
    "carvemereactions = []\n",
    "\n",
    "\n",
    "for gene in carveme_model.genes:\n",
    "    carvemegenes.append(gene.id)\n",
    "    carvemengoids.append(gene.annotation)\n",
    "    sets=list(gene.reactions)\n",
    "    rxnlist=[]\n",
    "    for x in sets:\n",
    "        rxnlist.append(x)\n",
    "    carvemereactions.append(rxnlist)\n",
    "\n",
    "dict = {'carvemegenes': carvemegenes, 'NGO_IDs':carvemengoids,'carvemereactions':carvemereactions}\n",
    "carvemegenes_reactions_DF = pandas.DataFrame(dict)\n",
    "carvemegenes_reactions_DF.to_csv('carveme_genes_reactions.csv')\n",
    "carvemegenes_reactions_DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0a01f816-49a9-4542-b975-b996170ecd39",
   "metadata": {},
   "outputs": [],
   "source": [
    "carveme_model.metabolites.ala__L_c.id ='ala_L_c'\n",
    "carveme_model.metabolites.glu__L_c.id ='glu_L_c'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "06da6b2d-9302-4029-b213-a8bc941b90af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ALATA_L</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>L-alanine transaminase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd8c58ee0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>akg_c + ala_L_c <=> glu_L_c + pyr_c</p>\n",
       "                    <p style='text-align:right'>M_akg_c + M_ala_L_c <=> M_glu_L_c + M_pyr_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1047</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ALATA_L at 0x19cd8c58ee0>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALATA_L =carvememodel.reactions.get_by_id('ALATA_L')\n",
    "carvememodel.reactions.get_by_id('ALATA_L').gene_reaction_rule='NGO1047'\n",
    "model.add_reactions([ALATA_L])\n",
    "model.reactions.get_by_id('ALATA_L')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "943cac6e-5eff-4760-8764-bc9e0cc36eb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>AP4AH</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Ap4A hydrolase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd6d73f40</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>ap4a_c + h2o_c --> 2.0 adp_c + 2.0 h_c</p>\n",
       "                    <p style='text-align:right'>P1,P4-Bis(5'-adenosyl) tetraphosphate + M_h2o_c --> 2.0 M_adp_c + 2.0 M_h_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0231</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction AP4AH at 0x19cd6d73f40>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AP4AH =carvememodel.reactions.get_by_id('AP4AH')\n",
    "carvememodel.reactions.get_by_id('AP4AH').gene_reaction_rule='NGO0231'\n",
    "model.add_reactions([AP4AH])\n",
    "model.reactions.get_by_id('AP4AH')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d02078a8-ba82-4ad6-82f5-a61410932f0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NGO0872\n",
      "NGO0872 and NGO1204\n"
     ]
    }
   ],
   "source": [
    "#ANS already exists in annotated Gc model, the model GPR should require NGO9872 and NGO1204\n",
    "print (model.reactions.get_by_id('ANS').gene_reaction_rule)\n",
    "model.reactions.get_by_id('ANS').gene_reaction_rule='NGO0872 and NGO1204'\n",
    "print (model.reactions.get_by_id('ANS').gene_reaction_rule)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6e2500b5-f019-47f2-ac5d-3aa296f1b0c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>BTDD_RR</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R R butanediol dehydrogenase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd6f9fd00</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>btd_RR_c + nad_c <=> actn__R_c + h_c + nadh_c</p>\n",
       "                    <p style='text-align:right'>R R 2 3 Butanediol C4H10O2 + M_nad_c <=> R Acetoin C4H8O2 + M_h_c + M_nadh_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0186</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction BTDD_RR at 0x19cd6f9fd00>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BTDD_RR =carvememodel.reactions.get_by_id('BTDD_RR')\n",
    "carvememodel.reactions.get_by_id('BTDD_RR').gene_reaction_rule='NGO0186'\n",
    "model.add_reactions([BTDD_RR])\n",
    "model.reactions.get_by_id('BTDD_RR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "483952ee-5b3d-49b5-9de5-7866e9eae446",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>CYTOM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Cytosine 5 methyltransferase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd75eaf40</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>amet_c + csn_c <=> 5mcsn_c + ahcys_c + h_c</p>\n",
       "                    <p style='text-align:right'>M_amet_c + M_csn_c <=> 5-Methylcytosine + M_ahcys_c + M_h_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0676</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction CYTOM at 0x19cd75eaf40>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CYTOM =carvememodel.reactions.get_by_id('CYTOM')\n",
    "carvememodel.reactions.get_by_id('CYTOM').gene_reaction_rule='NGO0676'\n",
    "model.add_reactions([CYTOM])\n",
    "model.reactions.get_by_id('CYTOM')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "83096db6-9909-49cf-8766-21977986debf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DNTPPA</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_DNTPPA</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd15c4b50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c</p>\n",
       "                    <p style='text-align:right'>M_ahdt_c + M_h2o_c --> M_dhpmp_c + M_h_c + M_ppi_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0224</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DNTPPA at 0x19cd15c4b50>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#The gene currently indicated for this reaction is NGO1506 - however KEGG annotates this gene as a 8-oxo-dGTP diphosphatase (MutT), KEGG reaction \t R09832\n",
    "#CARVEME indicates this reactions is produced by NGO1334. However this gene is annotated as a putative RNA pyrophosphohydrolase. The function isn't clear. \n",
    "# Replacing with the KEGG annotated gene for this reaction NGO0224\n",
    "\n",
    "DNTPPA =carvememodel.reactions.get_by_id('DNTPPA')\n",
    "model.reactions.get_by_id('DNTPPA').gene_reaction_rule='NGO0224'\n",
    "model.reactions.get_by_id('DNTPPA')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "20388902-2684-4781-8da8-4585b660567a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DNMPPA</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_DNMPPA</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd15c45e0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>dhpmp_c + h2o_c --> dhnpt_c + pi_c</p>\n",
       "                    <p style='text-align:right'>M_dhpmp_c + M_h2o_c --> M_dhnpt_c + M_pi_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1334</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DNMPPA at 0x19cd15c45e0>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#DNMPPA is not annotated in KEGG, but is predicted to use an enzyme with E.C. number 3.6.1-, and is listed as an orphan gene in the current model. \n",
    "#I am tentatively de-orphaning the gene by adding the annotation from CARVEME for NGO1334\n",
    "\n",
    "DNMPPA =carvememodel.reactions.get_by_id('DNMPPA')\n",
    "model.reactions.get_by_id('DNMPPA').gene_reaction_rule='NGO1334'\n",
    "model.reactions.get_by_id('DNMPPA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d0e01a89-6219-44de-8e05-3727c50aace3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>FADRx</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>FAD reductase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd886be20</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>fad_c + h_c + nadh_c --> fadh2_c + nad_c</p>\n",
       "                    <p style='text-align:right'>M_fad_c + M_h_c + M_nadh_c --> M_fadh2_c + M_nad_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0059</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction FADRx at 0x19cd886be20>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME indicates FADRx uses NGO0059 or NGO0659. However NGO0659 has different annotations in KEGG that do not match this function. Removed NGO0659. \n",
    "FADRx =carvememodel.reactions.get_by_id('FADRx')\n",
    "carvememodel.reactions.get_by_id('FADRx').gene_reaction_rule='NGO0059'\n",
    "model.add_reactions([FADRx])\n",
    "model.reactions.get_by_id('FADRx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "fdd8de44-f8c7-47c3-9475-6dce142a1bef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>HPA3MO</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>4 hydroxyphenylacetate 3 monooxygenase nadh</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd7181df0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>4hphac_c + h_c + nadh_c + o2_c --> 34dhpha_c + h2o_c + nad_c</p>\n",
       "                    <p style='text-align:right'>4-Hydroxyphenylacetate + M_h_c + M_nadh_c + M_o2_c --> 3-4-Dihydroxyphenylacetate + M_h2o_c + M_nad_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0059</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction HPA3MO at 0x19cd7181df0>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME and KEGG indicate presence of this reaction\n",
    "HPA3MO =carvememodel.reactions.get_by_id('HPA3MO')\n",
    "carvememodel.reactions.get_by_id('HPA3MO').gene_reaction_rule='NGO0059'\n",
    "model.add_reactions([HPA3MO])\n",
    "model.reactions.get_by_id('HPA3MO')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1bcac71b-08c9-45f6-8b94-461db3681cd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GLYCL</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_GLYCL</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd16bcd30</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>gly_c + nad_c + thf_c <=> co2_c + mlthf_c + nadh_c + nh4_c</p>\n",
       "                    <p style='text-align:right'>M_gly_c + M_nad_c + M_thf_c <=> M_co2_c + M_mlthf_c + M_nadh_c + M_nh4_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0562 and NGO0915 and NGO0925 and NGO1325 and NGO1404 and NGO1406</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-999999.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GLYCL at 0x19cd16bcd30>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Model has NGO1406 annotated for GLYCL. CARVEME has NGO1325 annotated. \n",
    "#The Glycine cleavage system is a functional system composed of 4 different components *H, P, T, and L. https://www.biocyc.org/GCF_000006845/NEW-IMAGE?type=PATHWAY&object=GLYCLEAV-PWY\n",
    "#All 4 components should be required for function of this system. As such, I am listing ALL 6 genes, indicated in KEGG, for this reaction. \n",
    "\n",
    "GLYCL =carvememodel.reactions.get_by_id('GLYCL')\n",
    "model.reactions.get_by_id('GLYCL').gene_reaction_rule='NGO0562 and NGO0915 and NGO0925 and NGO1325 and NGO1404 and NGO1406'\n",
    "model.reactions.get_by_id('GLYCL')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "638d74b9-48cc-4650-b51b-5493e4b9f79b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GLYOX</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Hydroxyacylglutathione hydrolase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd6eb5b20</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h2o_c + lgt__S_c --> gthrd_c + h_c + lac__D_c</p>\n",
       "                    <p style='text-align:right'>H2O + (R)-S-Lactoylglutathione --> Reduced glutathione + H+ + D-Lactate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YP_208259_1 or YP_209053_1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GLYOX at 0x19cd6eb5b20>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME indicated NGO2033\n",
    "#KEGG indicates NGO1187 AND NGO2033 participate in this reaction https://www.genome.jp/entry/ngo:NGO_1187+ngo:NGO_2033\n",
    "\n",
    "\n",
    "GLYOX =carvememodel.reactions.get_by_id('GLYOX')\n",
    "carvememodel.reactions.get_by_id('GLYOX')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "bd931187-eaa4-45bd-af9d-349b52873405",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GLYOX</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_GLYOX</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd16c9490</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h2o_c + lgt_S_c <=> gthrd_c + h_c + lac_D_c</p>\n",
       "                    <p style='text-align:right'>M_h2o_c + M_lgt_S_c <=> M_gthrd_c + M_h_c + M_lac_D_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1187</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-999999.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GLYOX at 0x19cd16c9490>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('HPA3MO').gene_reaction_rule='NGO1187 and NGO2033'\n",
    "model.reactions.get_by_id('GLYOX')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "84e0b48f-e4ca-44d7-b4e3-a2702f54378c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GTPCI</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_GTPCI</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd16d4940</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>gtp_c + h2o_c --> ahdt_c + for_c + h_c</p>\n",
       "                    <p style='text-align:right'>M_gtp_c + M_h2o_c --> M_ahdt_c + M_for_c + M_h_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0387</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GTPCI at 0x19cd16d4940>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME indicates NGO0387 as performing this reaction, and so does KEGG. https://www.genome.jp/entry/ngo:NGO_0387\n",
    "GTPCI =carvememodel.reactions.get_by_id('GTPCI')\n",
    "model.reactions.get_by_id('GTPCI').gene_reaction_rule='NGO0387'\n",
    "model.reactions.get_by_id('GTPCI')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a7014173-e6d6-4228-8d8f-fc8afe1e6b82",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:cobra.core.model:Ignoring reaction 'G3PAT120' since it already exists.\n",
      "WARNING:cobra.core.model:Ignoring reaction 'G3PAT180' since it already exists.\n",
      "WARNING:cobra.core.model:Ignoring reaction 'G3PAT181' since it already exists.\n",
      "WARNING:cobra.core.model:Ignoring reaction 'G3PAT140' since it already exists.\n",
      "WARNING:cobra.core.model:Ignoring reaction 'G3PAT160' since it already exists.\n"
     ]
    }
   ],
   "source": [
    "#Adding reactions according to PMCID: PMC4697154/PMID: 26567338 which are conducted by plsX (Ngo2151)\n",
    "x = ['ACPPAT161' , 'G3PAT120' , 'ACPPAT181' , 'G3PAT180' , 'ACPPAT120' , 'G3PAT181' , 'ACPPAT141' , 'G3PAT140' , 'ACPPAT160' , 'G3PAT161' , 'G3PAT160' , 'ACPPAT140' , 'ACPPAT180']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO2171'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "402507da-306c-40da-bbf7-b3b31ce3c724",
   "metadata": {},
   "outputs": [],
   "source": [
    "#providing gene for orphan reactions according to PMCID: PMC4697154/PMID: 26567338 which are conducted by plsX (Ngo2151)\n",
    "x = ['G3PAT120' , 'G3PAT180' , 'G3PAT181' , 'G3PAT140' , 'G3PAT161' , 'G3PAT160' ]\n",
    "for r in x:\n",
    "    reaction =model.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO2171'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c0a63216-dd74-4d74-911a-7177bb5e88a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Adding reactions according to PMCID: PMC4697154/PMID: 26567338 which are conducted by plsY (Ngo2151)\n",
    "x = ['APG3PAT181' , 'APG3PAT160' , 'APG3PAT180' , 'APG3PAT120' , 'APG3PAT141' , 'APG3PAT161' , 'APG3PAT140']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0858'\n",
    "    model.add_reactions([reaction])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "458492ae-c934-4ef1-aac6-737e5b13a6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Adding peptidase reaction\n",
    "x = ['AMPTASEPG']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO1227'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "03324b10-47cb-4ed7-950f-100de086fe6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cardiolipin has been in cultures of Gc PMCID: PMC235958/ PMID: 810478\n",
    "x = ['CLPNS181pp' , 'CLPNS141pp' , 'CLPNS140pp' , 'CLPNS160pp' , 'CLPNS120pp' , 'CLPNS161pp' , 'CLPNS180pp']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0902'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b9cc3884-c9d9-4a03-9778-80afac12e73b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CYPOX: h2o2_c + 2.0 h_c --> 2.0 h2o_c\n",
      "{}\n"
     ]
    }
   ],
   "source": [
    "#reduction of H2O2 to water facilitated by cytochrome c peroxidase, with Laz as an electron donor. Laz is encoded by NGO0994\n",
    "reaction = cobra.Reaction('CYPOX')\n",
    "reaction.name = 'Cytochrome C Peroxidase'\n",
    "reaction.lower_bound = 0.  \n",
    "reaction.upper_bound = 1000.  \n",
    "reaction.gene_reaction_rule='NGO1769 and NGO0994'\n",
    "\n",
    "\n",
    "\n",
    "h_c=model.metabolites.get_by_id('h_c')\n",
    "h2o2= model.metabolites.get_by_id('h2o2_c')\n",
    "h2o_c=model.metabolites.get_by_id('h2o_c')\n",
    "\n",
    "reaction.add_metabolites({\n",
    "    h2o2: -1.0,\n",
    "    h_c: -2.0,\n",
    "    h2o_c: 2.0,\n",
    "})\n",
    "\n",
    "model.add_reactions([reaction])\n",
    "print(model.reactions.CYPOX)\n",
    "print(model.reactions.CYPOX.check_mass_balance())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "06484aa7-b0d7-4ee8-bed5-b274738022a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ['FLVR', 'FLVRx']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0388'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "79a86004-0a87-4d63-966b-5c0a9ab846fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This reaction is indicated as a bidirectional reaction in Bigg. However Kegg indicates it is \n",
    "#unidirection with NGO1977 performing the forward reaction and NGO1406 the reverse reaction (https://www.kegg.jp/pathway/ngo00670+NGO_1977)\n",
    "x = ['FOMETRi']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO1977 and NGO1406'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "9bb3facf-bdeb-4dc8-b336-9fa37f7d929b",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.GLYCLTDx.gene_reaction_rule='NGO2043 or NGO1243'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "083509b7-a646-4b0f-81eb-f255dddff745",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ['GRXR']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0114 or NGO1381 or NGO0351'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "7d18ac12-7ebb-4485-9de3-ec263fdc9ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Carveme did not correctly annotate NGO1684, (its Not GTPCI) but it does perform a reaction just downstream. \n",
    "#Involved in folate biosynthesis. Corrected and added the full pathway.\n",
    "#To include NGO0387 (GTPCI), NGO0131, NGO00720, NGO0129, NGO1684 (https://www.genome.jp/pathway/ngo00790+NGO_1684)\n",
    "x = ['PTHPS']\n",
    "for r in x:\n",
    "    reaction =universalmodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0131'\n",
    "    model.add_reactions([reaction])\n",
    "x = ['CPH4S']\n",
    "for r in x:\n",
    "    reaction =universalmodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0131'\n",
    "    model.add_reactions([reaction])\n",
    "x = ['CDGS']\n",
    "for r in x:\n",
    "    reaction =universalmodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO00720'\n",
    "    model.add_reactions([reaction])\n",
    "x = ['CDGR']\n",
    "for r in x:\n",
    "    reaction =universalmodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO1684'\n",
    "    model.add_reactions([reaction])\n",
    "x = ['CCGS']\n",
    "for r in x:\n",
    "    reaction =universalmodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0129'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "1644ac28-4779-4daa-8b9b-868787c19c6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.Kt2pp.gene_reaction_rule='NGO1774 or NGO1154'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "7ffa49c7-e8ac-4224-9e23-dafe2e1f740c",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ['MLDEP2pp']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0571 or NGO1686'\n",
    "    model.add_reactions([reaction])\n",
    "#ltgD (0626) and ltgC (2048) ltgB (1033) and ltgE(0608) are all murein lytic transcglycosylases. PMCID: PMC3412582/PMID: 22432703\n",
    "#Adding all of them as potential genes \n",
    "x = ['MLTGY3pp']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0608 or NGO1033 or NGO0626 or NGO2048'\n",
    "    model.add_reactions([reaction])\n",
    "\n",
    "x = ['MLTGY1pp' ,'MLTGY4pp' , 'MLTGY2pp']\n",
    "for r in x:\n",
    "    reaction =model.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0608 or NGO1033 or NGO0626 or NGO2048'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "eb4794c8-f72b-4b97-b740-97ea4a768dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Deorphan PPPGO\n",
    "model.reactions.PPPGO.gene_reaction_rule= 'NGO1692'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d5c19908-4d75-4144-a64b-32c9d21004bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add additional thioredoxin proteins\n",
    "model.reactions.THIORDXi.gene_reaction_rule=  'NGO0328 or NGO1923 or NGO0652 or NGO0057'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "747cb964-d0c2-4d40-8367-9ce5b2904d28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>PGLYCP</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_PGLYCP</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd182f910</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>2pglyc_c + h2o_c --> glyclt_c + pi_c</p>\n",
       "                    <p style='text-align:right'>M_2pglyc_c + M_h2o_c --> M_glyclt_c + M_pi_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0073 or NGO1052</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction PGLYCP at 0x19cd182f910>"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME indicates NGO1052\n",
    "#KEGG indicates NGO0073 and NGO1052\n",
    "#NGO0203 has a different EC number - EC 3.1.3.105 . I am replacing with the KEGG/CARVEME gene annotations\n",
    "\n",
    "PGLYCP =carvememodel.reactions.get_by_id('PGLYCP')\n",
    "\n",
    "model.reactions.get_by_id('PGLYCP').gene_reaction_rule='NGO0073 or NGO1052'\n",
    "model.reactions.get_by_id('PGLYCP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "73d8c37a-6c45-4c63-a77e-693c4f6409ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>SPMS</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_SPMS</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd1918700</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>ametam_c + ptrc_c --> 5mta_c + h_c + spmd_c</p>\n",
       "                    <p style='text-align:right'>M_ametam_c + M_ptrc_c --> M_5mta_c + M_h_c + M_spmd_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0435 or NGO1752</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction SPMS at 0x19cd1918700>"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME indicates NGO0435 or NGO1752. Model has mislabeled NGO number with an underscore. Replacing.\n",
    "SPMS =carvememodel.reactions.get_by_id('SPMS')\n",
    "model.reactions.get_by_id('SPMS').gene_reaction_rule='NGO0435 or NGO1752'\n",
    "model.reactions.get_by_id('SPMS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "bda925b0-d6e5-423d-9534-51a4c3f08082",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Model contains ethanol production - however Gc does not produce ethanol. Removing this reaction.\n",
    "ALCD2x =model.reactions.get_by_id('ALCD2x')\n",
    "model.remove_reactions([ALCD2x])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "0a9834e8-fc0f-4910-9c92-dd01e8ad9e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Similarly, the mening model had added acetylaldehyde production, without genetic evidence, to support the production of ethanol observed in mening. \n",
    "#Because Gc does not produce ethanol or acetylaldehyde, removing these associated reaction.\n",
    "ALDD2y =model.reactions.get_by_id('ALDD2y')\n",
    "model.remove_reactions([ALDD2y])\n",
    "\n",
    "ACALDtpp =model.reactions.get_by_id('ACALDtpp')\n",
    "model.remove_reactions([ACALDtpp])\n",
    "\n",
    "EX_acald_e_ =model.reactions.get_by_id('EX_acald_e_')\n",
    "model.remove_reactions([EX_acald_e_])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6e1d2180-1a74-434f-b94a-48b2f5425ff5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.6353484052933085</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.2122</td>\n",
       "      <td>6</td>\n",
       "      <td>0.34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.5239</td>\n",
       "      <td>4</td>\n",
       "      <td>0.55%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>21.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.00301</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.00301</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.002006</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.002006</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>15.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.009313</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc_D_e</td>\n",
       "      <td>EX_glc_D_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>31.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.4511</td>\n",
       "      <td>5</td>\n",
       "      <td>0.59%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>6.983</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.06425</td>\n",
       "      <td>6</td>\n",
       "      <td>0.10%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.1128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.3003</td>\n",
       "      <td>6</td>\n",
       "      <td>0.47%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.223</td>\n",
       "      <td>6</td>\n",
       "      <td>0.35%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.005016</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.002006</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.002006</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.1265</td>\n",
       "      <td>9</td>\n",
       "      <td>0.30%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.4651</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.1566</td>\n",
       "      <td>5</td>\n",
       "      <td>0.21%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>15.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.002508</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>11.38</td>\n",
       "      <td>4</td>\n",
       "      <td>12.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.04381</td>\n",
       "      <td>11</td>\n",
       "      <td>0.13%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.09088</td>\n",
       "      <td>9</td>\n",
       "      <td>0.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.2905</td>\n",
       "      <td>5</td>\n",
       "      <td>0.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.002008</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-0.0001994</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0002846</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-1.264E-06</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.6353</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0002834</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.01106</td>\n",
       "      <td>10</td>\n",
       "      <td>0.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.004916</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-55.13</td>\n",
       "      <td>2</td>\n",
       "      <td>31.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-77.56</td>\n",
       "      <td>1</td>\n",
       "      <td>21.94%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.000425</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-25.51</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>-11.01</td>\n",
       "      <td>6</td>\n",
       "      <td>18.69%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>-19.88</td>\n",
       "      <td>5</td>\n",
       "      <td>28.13%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-37.78</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cd175f070>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpmi(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "5d62aae3-d34a-44d4-946e-f4e43ba7d53b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CARVEME revealed that NGO0904-NGO0906 is not present in the Gc model - which was recently discovered as an Gc\n",
    "#lactate dehydrogenase. Additionally other aspects of lactate metabolism are corrected here. \n",
    "\n",
    "#PMCID: PMC4215069\n",
    "#PMCID: PMC7530546"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "00fe0604-2c52-4c32-83f1-9b6f7998984d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>lac_D_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>M_lac_D_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cc8951940</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C3H5O3</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>Cytosol</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 3 reaction(s)</strong></td><td>\n",
       "                    LDH_D, GLYOX, DLACt2pp</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite lac_D_c at 0x19cc8951940>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.get_by_id('lac_D_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "3be28ec5-348e-4d68-be89-5550a910e0e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "carvememodel.metabolites.get_by_id('lac__D_c').id='lac_D_c'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "2df2ad40-4f17-46b8-b815-3588f13b34a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>L_LACD4</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_L_LACD4</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd1720d90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>lac_L_c + nad_c --> h_c + nadh_c + pyr_c</p>\n",
       "                    <p style='text-align:right'>M_lac_L_c + M_nad_c --> M_h_c + M_nadh_c + M_pyr_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0639 or Orphan</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction L_LACD4 at 0x19cd1720d90>"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('L_LACD4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "a39b6955-7222-47a8-b76c-99d75d4c7902",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.4054101760153453</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.1354</td>\n",
       "      <td>6</td>\n",
       "      <td>1.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.102</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>0.2643</td>\n",
       "      <td>4</td>\n",
       "      <td>1.34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.00192</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.00192</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.03201</td>\n",
       "      <td>3</td>\n",
       "      <td>0.12%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.005943</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.2879</td>\n",
       "      <td>5</td>\n",
       "      <td>1.83%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>4.156</td>\n",
       "      <td>2</td>\n",
       "      <td>10.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>5.819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.041</td>\n",
       "      <td>6</td>\n",
       "      <td>0.31%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.1183</td>\n",
       "      <td>6</td>\n",
       "      <td>0.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.1916</td>\n",
       "      <td>6</td>\n",
       "      <td>1.46%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.1423</td>\n",
       "      <td>6</td>\n",
       "      <td>1.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.04157</td>\n",
       "      <td>5</td>\n",
       "      <td>0.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.003201</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.08069</td>\n",
       "      <td>9</td>\n",
       "      <td>0.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.2967</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.09991</td>\n",
       "      <td>5</td>\n",
       "      <td>0.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>76.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.001601</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.02795</td>\n",
       "      <td>11</td>\n",
       "      <td>0.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.05798</td>\n",
       "      <td>9</td>\n",
       "      <td>0.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.1854</td>\n",
       "      <td>5</td>\n",
       "      <td>1.18%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-0.0001111</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0001816</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-8.067E-07</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.4054</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0001808</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.007056</td>\n",
       "      <td>10</td>\n",
       "      <td>0.11%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.003137</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-19.75</td>\n",
       "      <td>2</td>\n",
       "      <td>63.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-22.63</td>\n",
       "      <td>1</td>\n",
       "      <td>36.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0002712</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-3.718</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-22.54</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cd175f2e0>"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#NGO0639 is a quinone depdendent L-lactate dehydrogenase. There are no known NAD/NADH dependent L-lactate dehydrogenases in Gc.\n",
    "LACD4 =model.reactions.get_by_id('L_LACD4')\n",
    "model.remove_reactions([LACD4])\n",
    "rpmi_with_lactate_only(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "1c8057d3-470c-407e-b301-14400dbb76b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>LDH_D</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_LDH_D</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd172d400</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>lac_D_c + nad_c <=> h_c + nadh_c + pyr_c</p>\n",
       "                    <p style='text-align:right'>M_lac_D_c + M_nad_c <=> M_h_c + M_nadh_c + M_pyr_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1336</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction LDH_D at 0x19cd172d400>"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#The reaction LDH_D is a the NAD dependent conversion of D_lac to pyruvate. It is completed by ldhA (NGO1336). \n",
    "#This reaction should be bidirectional. \n",
    "\n",
    "model.reactions.get_by_id('LDH_D')\n",
    "model.reactions.get_by_id('LDH_D').gene_reaction_rule='NGO1336'\n",
    "model.reactions.get_by_id('LDH_D').lower_bound =-1000.\n",
    "model.reactions.get_by_id('LDH_D')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "1c11dc22-14af-4e29-b9d1-1ee5f1578ea2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.40541016811719427</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.1354</td>\n",
       "      <td>6</td>\n",
       "      <td>1.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.102</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>0.2643</td>\n",
       "      <td>4</td>\n",
       "      <td>1.34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.00192</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.00192</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.03201</td>\n",
       "      <td>3</td>\n",
       "      <td>0.12%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.005943</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.2879</td>\n",
       "      <td>5</td>\n",
       "      <td>1.83%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>4.156</td>\n",
       "      <td>2</td>\n",
       "      <td>10.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>5.819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.041</td>\n",
       "      <td>6</td>\n",
       "      <td>0.31%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.1183</td>\n",
       "      <td>6</td>\n",
       "      <td>0.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.1916</td>\n",
       "      <td>6</td>\n",
       "      <td>1.46%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.1423</td>\n",
       "      <td>6</td>\n",
       "      <td>1.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.04157</td>\n",
       "      <td>5</td>\n",
       "      <td>0.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.003201</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.08069</td>\n",
       "      <td>9</td>\n",
       "      <td>0.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.2967</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.09991</td>\n",
       "      <td>5</td>\n",
       "      <td>0.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>76.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.001601</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.02795</td>\n",
       "      <td>11</td>\n",
       "      <td>0.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.05798</td>\n",
       "      <td>9</td>\n",
       "      <td>0.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.1854</td>\n",
       "      <td>5</td>\n",
       "      <td>1.18%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-0.0001111</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0001816</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-8.067E-07</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.4054</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0001808</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.007056</td>\n",
       "      <td>10</td>\n",
       "      <td>0.11%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.003137</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-19.75</td>\n",
       "      <td>2</td>\n",
       "      <td>63.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-22.63</td>\n",
       "      <td>1</td>\n",
       "      <td>36.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0002712</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-3.718</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-22.54</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cd7a09b20>"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpmi_with_lactate_only(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "1962f8c1-cce6-4d44-b2c0-2d501dfb6853",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>LDH_D2</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>D-lactate dehydrogenase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd8b4f5e0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>lac_D_c + q8_c --> pyr_c + q8h2_c</p>\n",
       "                    <p style='text-align:right'>M_lac_D_c + M_q8_c --> M_pyr_c + M_q8h2_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0890</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction LDH_D2 at 0x19cd8b4f5e0>"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#The reaction LDH_D2 is competed by ldhD (NGO0890), a Quinone dependent D-lactate dehydrogenase.\n",
    "LDH_D2=carvememodel.reactions.get_by_id('LDH_D2')\n",
    "carvememodel.reactions.get_by_id('LDH_D2').gene_reaction_rule='NGO0890'\n",
    "model.add_reactions([LDH_D2])\n",
    "model.reactions.get_by_id('LDH_D2')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "71d67d1e-e597-4eb9-aea5-8eea1ff17639",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.4054097879672639</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.1354</td>\n",
       "      <td>6</td>\n",
       "      <td>1.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.102</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>0.2643</td>\n",
       "      <td>4</td>\n",
       "      <td>1.34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.00192</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.00192</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.03201</td>\n",
       "      <td>3</td>\n",
       "      <td>0.12%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.005942</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.2879</td>\n",
       "      <td>5</td>\n",
       "      <td>1.83%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>4.156</td>\n",
       "      <td>2</td>\n",
       "      <td>10.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>5.819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.041</td>\n",
       "      <td>6</td>\n",
       "      <td>0.31%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.1183</td>\n",
       "      <td>6</td>\n",
       "      <td>0.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.1916</td>\n",
       "      <td>6</td>\n",
       "      <td>1.46%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.1423</td>\n",
       "      <td>6</td>\n",
       "      <td>1.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.04157</td>\n",
       "      <td>5</td>\n",
       "      <td>0.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.003201</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.08069</td>\n",
       "      <td>9</td>\n",
       "      <td>0.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.2967</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.09991</td>\n",
       "      <td>5</td>\n",
       "      <td>0.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>76.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.001601</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.02795</td>\n",
       "      <td>11</td>\n",
       "      <td>0.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.05798</td>\n",
       "      <td>9</td>\n",
       "      <td>0.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.1854</td>\n",
       "      <td>5</td>\n",
       "      <td>1.18%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-0.0001111</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0001816</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-8.067E-07</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.4054</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0001808</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.007056</td>\n",
       "      <td>10</td>\n",
       "      <td>0.11%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.003137</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-19.75</td>\n",
       "      <td>2</td>\n",
       "      <td>63.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-22.63</td>\n",
       "      <td>1</td>\n",
       "      <td>36.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0002712</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-3.718</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-22.54</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cd7fe1cd0>"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpmi_with_lactate_only(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "5a0f937e-b95e-4960-8e97-9fd6dc78bc4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>L_LACD2</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_L_LACD2</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd1720c70</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>lac_L_c + q8_c --> pyr_c + q8h2_c</p>\n",
       "                    <p style='text-align:right'>M_lac_L_c + M_q8_c --> M_pyr_c + M_q8h2_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0639 or (NGO0904 and NGO0905 and NGO0906)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction L_LACD2 at 0x19cd1720c70>"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#L_LACD2 is done by lldD (NGO0639), and l-lactate dehydrogenase that is quinone dependent. \n",
    "#IT was also found that lutCAB (NGO0904-NGO0906) are also L-lactate dehydrogenases, predicted to be quinone dependent.\n",
    "# These are added here. \n",
    "model.reactions.get_by_id('L_LACD2').gene_reaction_rule='NGO0639 or (NGO0904 and NGO0905 and NGO0906)'\n",
    "model.reactions.get_by_id('L_LACD2')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "fee11e6e-c0f6-4bd5-a38a-d2377ab714a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.4054097879672639</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.1354</td>\n",
       "      <td>6</td>\n",
       "      <td>1.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.102</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>0.2643</td>\n",
       "      <td>4</td>\n",
       "      <td>1.34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.00192</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.00192</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.03201</td>\n",
       "      <td>3</td>\n",
       "      <td>0.12%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.005942</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.2879</td>\n",
       "      <td>5</td>\n",
       "      <td>1.83%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>4.156</td>\n",
       "      <td>2</td>\n",
       "      <td>10.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>5.819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.041</td>\n",
       "      <td>6</td>\n",
       "      <td>0.31%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.1183</td>\n",
       "      <td>6</td>\n",
       "      <td>0.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.1916</td>\n",
       "      <td>6</td>\n",
       "      <td>1.46%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.1423</td>\n",
       "      <td>6</td>\n",
       "      <td>1.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.04157</td>\n",
       "      <td>5</td>\n",
       "      <td>0.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.003201</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.08069</td>\n",
       "      <td>9</td>\n",
       "      <td>0.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.2967</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.09991</td>\n",
       "      <td>5</td>\n",
       "      <td>0.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>76.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.001601</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.1172</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.02795</td>\n",
       "      <td>11</td>\n",
       "      <td>0.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.05798</td>\n",
       "      <td>9</td>\n",
       "      <td>0.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.1854</td>\n",
       "      <td>5</td>\n",
       "      <td>1.18%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.00128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-0.0001111</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0001816</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-8.067E-07</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.4054</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0001808</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.007056</td>\n",
       "      <td>10</td>\n",
       "      <td>0.11%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.003137</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-19.75</td>\n",
       "      <td>2</td>\n",
       "      <td>63.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-22.63</td>\n",
       "      <td>1</td>\n",
       "      <td>36.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0002712</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-3.718</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-22.54</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cd7fe1df0>"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpmi_with_lactate_only(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "f5b50a05-e78b-4f32-8813-b5468148cc1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Gene identifier</strong></td><td>NGO1449</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd14408b0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Functional</strong></td><td>True</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 3 reaction(s)</strong></td><td>\n",
       "                    GLYCLTt2rpp, LLACt3pp, DLACt2pp</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Gene NGO1449 at 0x19cd14408b0>"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.genes.get_by_id('NGO1449')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "dfba4bbd-8f0c-4412-a36c-da955a90b2f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DLACt2pp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_DLACt2pp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd15b9fd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h_p + lac_D_p <=> h_c + lac_D_c</p>\n",
       "                    <p style='text-align:right'>M_h_p + M_lac_D_p <=> M_h_c + M_lac_D_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1449</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-999999.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DLACt2pp at 0x19cd15b9fd0>"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#L-LActate transport is abolished in an lctP (NGO1449) mutant, therefore, removing NGO1361 from L-lactate transport.\n",
    "#NGO1361 is homologous to the L-lactate transporter lldP from E. coli, however it is much shorter in Gc, only 71 amino acids. It appears to be truncated/nonfunctioal. \n",
    "# So I am removing it also from DLACt2pp transport.\n",
    "\n",
    "model.reactions.get_by_id('LLACt3pp')\n",
    "model.reactions.get_by_id('LLACt3pp').gene_reaction_rule='NGO1449'\n",
    "model.reactions.get_by_id('LLACt3pp')\n",
    "\n",
    "model.reactions.get_by_id('DLACt2pp')\n",
    "model.reactions.get_by_id('DLACt2pp').gene_reaction_rule='NGO1449'\n",
    "model.reactions.get_by_id('DLACt2pp')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ec9070bc-4d14-4df1-bf12-376be6b773eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.7307883834294733</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.2441</td>\n",
       "      <td>6</td>\n",
       "      <td>0.57%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.1839</td>\n",
       "      <td>4</td>\n",
       "      <td>0.29%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>8.916</td>\n",
       "      <td>4</td>\n",
       "      <td>13.85%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.003462</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.003462</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.002308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.002308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>8.497</td>\n",
       "      <td>3</td>\n",
       "      <td>9.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.01071</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc_D_e</td>\n",
       "      <td>EX_glc_D_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>46.60%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.5189</td>\n",
       "      <td>5</td>\n",
       "      <td>1.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.07391</td>\n",
       "      <td>6</td>\n",
       "      <td>0.17%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.2132</td>\n",
       "      <td>6</td>\n",
       "      <td>0.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.1298</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.3454</td>\n",
       "      <td>6</td>\n",
       "      <td>0.80%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.2565</td>\n",
       "      <td>6</td>\n",
       "      <td>0.60%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.00577</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.002308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.002308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.1454</td>\n",
       "      <td>9</td>\n",
       "      <td>0.51%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.5349</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.1801</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>23.30%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.002885</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.2114</td>\n",
       "      <td>4</td>\n",
       "      <td>0.33%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.05039</td>\n",
       "      <td>11</td>\n",
       "      <td>0.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.1045</td>\n",
       "      <td>9</td>\n",
       "      <td>0.37%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.3342</td>\n",
       "      <td>5</td>\n",
       "      <td>0.65%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.002309</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-0.0001969</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0003274</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-1.454E-06</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.7308</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0003259</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.01272</td>\n",
       "      <td>10</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.005655</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-32.21</td>\n",
       "      <td>2</td>\n",
       "      <td>28.29%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-37.87</td>\n",
       "      <td>1</td>\n",
       "      <td>16.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0004889</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-10.45</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>-20.6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lac_D_e</td>\n",
       "      <td>EX_lac_D_e_</td>\n",
       "      <td>-27.81</td>\n",
       "      <td>3</td>\n",
       "      <td>36.65%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>-8.364</td>\n",
       "      <td>5</td>\n",
       "      <td>18.37%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-25.52</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cd951dcd0>"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpmi(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d3c0187e-64c7-464c-a36e-9e9b897a3126",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>UCYR</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_UCYR</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd1963fd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>2.0 focytC_p + q8h2_c <=> 2.0 ficytC_p + 2.0 h_p + q8_c</p>\n",
       "                    <p style='text-align:right'>2.0 M_focytC_p + M_q8h2_c <=> 2.0 M_ficytC_p + 2.0 M_h_p + M_q8_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO2031 and NGO2030 and NGO2029 and (NGO1080 or NGO0101 or Orphan)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-999999.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction UCYR at 0x19cd1963fd0>"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('UCYR')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "bcdf02a5-d512-4e99-9a2f-cd80ed8d238e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>q8_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>M_q8_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd13613d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C49H74O4</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>Cytosol</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 12 reaction(s)</strong></td><td>\n",
       "                    MDH2, NADHQNa, L_LACD2, FUMR, DHORD2, SUCD1, ASPO3, NADH16pp, NADH5, UCYR, LDH_D2, NADPHQR2</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite q8_c at 0x19cd13613d0>"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.get_by_id('q8_c')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "3b11e0dd-428c-42bc-8494-67aa5ca71abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "rpmi_req=set(['EX_apoACP_c_', 'EX_trdrd_c_','EX_cobalt2_e_','EX_cu2_e_','EX_fe3_e_','EX_mn2_e_','EX_mobd_e_','EX_zn2_e_','EX_ca2_e_','EX_no3_e_','EX_h2o_e_','EX_k_e_','EX_cl_e_','EX_mg2_e_','EX_so4_e_','EX_na1_e_','EX_pi_e_','EX_h_e_','EX_arg_L_e_','EX_asn_L_e_','EX_asp_L_e_','EX_cys_L_e_','EX_glu_L_e_','EX_gly_e_','EX_his_L_e_','EX_ile_L_e_','EX_leu_L_e_','EX_lys_L_e_','EX_met_L_e_','EX_phe_L_e_','EX_pro_L_e_','EX_ser_L_e_','EX_thr_L_e_','EX_trp_L_e_','EX_tyr_L_e_','EX_val_L_e_','EX_pnto_R_e_','EX_chol_e_','EX_inost_e_','EX_glc_D_e_','EX_gthrd_e_','EX_co2_e_'])\n",
    "\n",
    "def rpmi(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -0. \n",
    "\n",
    "\n",
    "\n",
    "mdm_req=set(['EX_apoACP_c_', 'EX_trdrd_c_','EX_co2_e_', 'EX_cobalt2_e_', 'EX_cu2_e_' , 'EX_h_e_', 'EX_h2o_e_','EX_mn2_e_','EX_mobd_e_','EX_tungs_e_', 'EX_zn2_e_','EX_na1_e_','EX_cl_e_','EX_k_e_','EX_so4_e_','EX_nh4_e_','EX_mg2_e_','EX_pi_e_','EX_ca2_e_','EX_fe3_e_','EX_no3_e_','EX_asp_L_e_','EX_glu_L_e_','EX_arg_L_e_','EX_gly_e_','EX_ser_L_e_','EX_leu_L_e_','EX_ile_L_e_','EX_val_L_e_','EX_tyr_L_e_','EX_cys_L_e_','EX_pro_L_e_','EX_trp_L_e_','EX_thr_L_e_','EX_phe_L_e_','EX_asn_L_e_','EX_gln_L_e_','EX_his_L_e_','EX_met_L_e_','EX_ala_L_e_','EX_lys_L_e_','EX_gthrd_e_','EX_thm_e_','EX_pnto_R_e_','EX_glc_D_e_','EX_hxan_e_','EX_ura_e_'])\n",
    "\n",
    "def mdm(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -0.        \n",
    "\n",
    "def mdm_with_lactate(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -10.              \n",
    "\n",
    "def mdm_with_lactate_only(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -0.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -10. \n",
    "            \n",
    "            \n",
    "def mdm_molarity_with_lactate(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -1000.\n",
    "        if 'EX_glc_D_e_' in reaction.id:\n",
    "            reaction.lower_bound = -27.8\n",
    "        if 'EX_ser_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.48\n",
    "        if 'EX_asp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -3.76\n",
    "        if 'EX_asn_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.17\n",
    "        if 'EX_glu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -8.84\n",
    "        if 'EX_gly_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.33\n",
    "        if 'EX_ala_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.12\n",
    "        if 'EX_val_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.51\n",
    "        if 'EX_leu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.69\n",
    "        if 'EX_gln_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.34\n",
    "        if 'EX_pro_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.43\n",
    "        if 'EX_phe_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_cl_e_' in reaction.id:\n",
    "            reaction.lower_bound = -107.94697\n",
    "        if 'EX_k_e_' in reaction.id:\n",
    "            reaction.lower_bound = -11.48\n",
    "        if 'EX_so4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -5.74\n",
    "        if 'EX_nh4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -4.11\n",
    "        if 'EX_mg2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.07\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_pi_e_' in reaction.id:\n",
    "            reaction.lower_bound = -40\n",
    "        if 'EX_ca2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.254\n",
    "        if 'EX_no3_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.051\n",
    "        if 'EX_arg_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.71\n",
    "        if 'EX_ile_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.23\n",
    "        if 'EX_tyr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_cys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.65\n",
    "        if 'EX_trp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_thr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.42\n",
    "        if 'EX_his_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.13\n",
    "        if 'EX_met_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1\n",
    "        if 'EX_gthrd_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_thm_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.00697\n",
    "        if 'EX_pnto_R_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.004\n",
    "        if 'EX_hxan_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.37\n",
    "        if 'EX_ura_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.45  \n",
    "        if 'EX_na1_e_' in reaction.id:\n",
    "            reaction.lower_bound = -105.018\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -20\n",
    "    \n",
    "\n",
    "def mdm_molarity_with_lactate_only(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -1000.\n",
    "        if 'EX_glc_D_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0\n",
    "        if 'EX_ser_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.48\n",
    "        if 'EX_asp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -3.76\n",
    "        if 'EX_asn_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.17\n",
    "        if 'EX_glu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -8.84\n",
    "        if 'EX_gly_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.33\n",
    "        if 'EX_ala_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.12\n",
    "        if 'EX_val_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.51\n",
    "        if 'EX_leu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.69\n",
    "        if 'EX_gln_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.34\n",
    "        if 'EX_pro_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.43\n",
    "        if 'EX_phe_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_cl_e_' in reaction.id:\n",
    "            reaction.lower_bound = -107.94697\n",
    "        if 'EX_k_e_' in reaction.id:\n",
    "            reaction.lower_bound = -11.48\n",
    "        if 'EX_so4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -5.74\n",
    "        if 'EX_nh4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -4.11\n",
    "        if 'EX_mg2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.07\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_pi_e_' in reaction.id:\n",
    "            reaction.lower_bound = -40\n",
    "        if 'EX_ca2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.254\n",
    "        if 'EX_no3_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.051\n",
    "        if 'EX_arg_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.71\n",
    "        if 'EX_ile_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.23\n",
    "        if 'EX_tyr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_cys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.65\n",
    "        if 'EX_trp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_thr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.42\n",
    "        if 'EX_his_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.13\n",
    "        if 'EX_met_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1\n",
    "        if 'EX_gthrd_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_thm_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.00697\n",
    "        if 'EX_pnto_R_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.004\n",
    "        if 'EX_hxan_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.37\n",
    "        if 'EX_ura_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.45  \n",
    "        if 'EX_na1_e_' in reaction.id:\n",
    "            reaction.lower_bound = -105.018\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -20\n",
    "               \n",
    "def mdm_molarity(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -1000.\n",
    "        if 'EX_glc_D_e_' in reaction.id:\n",
    "            reaction.lower_bound = -27.8\n",
    "        if 'EX_ser_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.48\n",
    "        if 'EX_asp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -3.76\n",
    "        if 'EX_asn_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.17\n",
    "        if 'EX_glu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -8.84\n",
    "        if 'EX_gly_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.33\n",
    "        if 'EX_ala_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.12\n",
    "        if 'EX_val_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.51\n",
    "        if 'EX_leu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.69\n",
    "        if 'EX_gln_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.34\n",
    "        if 'EX_pro_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.43\n",
    "        if 'EX_phe_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_cl_e_' in reaction.id:\n",
    "            reaction.lower_bound = -107.94697\n",
    "        if 'EX_k_e_' in reaction.id:\n",
    "            reaction.lower_bound = -11.48\n",
    "        if 'EX_so4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -5.74\n",
    "        if 'EX_nh4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -4.11\n",
    "        if 'EX_mg2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.07\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_pi_e_' in reaction.id:\n",
    "            reaction.lower_bound = -40\n",
    "        if 'EX_ca2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.254\n",
    "        if 'EX_no3_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.051\n",
    "        if 'EX_arg_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.71\n",
    "        if 'EX_ile_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.23\n",
    "        if 'EX_tyr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_cys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.65\n",
    "        if 'EX_trp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_thr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.42\n",
    "        if 'EX_his_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.13\n",
    "        if 'EX_met_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1\n",
    "        if 'EX_gthrd_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_thm_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.00697\n",
    "        if 'EX_pnto_R_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.004\n",
    "        if 'EX_hxan_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.37\n",
    "        if 'EX_ura_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.45  \n",
    "        if 'EX_na1_e_' in reaction.id:\n",
    "            reaction.lower_bound = -105.018\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -0\n",
    "                         \n",
    "        \n",
    "def complete(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=-1000.   \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "7ebb30c2-aaf7-4fe4-bc95-4481673a02b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for reaction in model.reactions:\n",
    "     if reaction.lower_bound < -4:\n",
    "            reaction.lower_bound=-1000.\n",
    "for reaction in model.reactions:\n",
    "     if reaction.upper_bound > 4:\n",
    "            reaction.upper_bound=1000."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "788d0c03-2ed0-4baa-909e-9893735e5f05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 3.0413126414123677</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ala_L_e</td>\n",
       "      <td>EX_ala_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>8.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>1.016</td>\n",
       "      <td>6</td>\n",
       "      <td>1.73%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.7652</td>\n",
       "      <td>4</td>\n",
       "      <td>0.87%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.01441</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.01441</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.009604</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.009604</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.2393</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.04458</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc_D_e</td>\n",
       "      <td>EX_glc_D_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>16.99%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gln_L_e</td>\n",
       "      <td>EX_gln_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>14.15%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>14.15%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>5.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.3076</td>\n",
       "      <td>6</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hxan_e</td>\n",
       "      <td>EX_hxan_e_</td>\n",
       "      <td>0.8593</td>\n",
       "      <td>5</td>\n",
       "      <td>1.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.8878</td>\n",
       "      <td>6</td>\n",
       "      <td>1.51%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.5401</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>1.437</td>\n",
       "      <td>6</td>\n",
       "      <td>2.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>1.068</td>\n",
       "      <td>6</td>\n",
       "      <td>1.81%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.3118</td>\n",
       "      <td>5</td>\n",
       "      <td>0.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.02401</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.009604</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.009604</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>154.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.6053</td>\n",
       "      <td>9</td>\n",
       "      <td>1.54%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>2.225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>14.15%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>8.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.01201</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thm_e</td>\n",
       "      <td>EX_thm_e_</td>\n",
       "      <td>0.0007709</td>\n",
       "      <td>12</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.879</td>\n",
       "      <td>4</td>\n",
       "      <td>1.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.2097</td>\n",
       "      <td>11</td>\n",
       "      <td>0.65%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.4341</td>\n",
       "      <td>9</td>\n",
       "      <td>1.11%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ura_e</td>\n",
       "      <td>EX_ura_e_</td>\n",
       "      <td>0.7955</td>\n",
       "      <td>4</td>\n",
       "      <td>0.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>1.391</td>\n",
       "      <td>5</td>\n",
       "      <td>1.97%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.009604</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.001362</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-6.051E-06</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-3.041</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.001356</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.05293</td>\n",
       "      <td>10</td>\n",
       "      <td>0.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.02353</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ade_e</td>\n",
       "      <td>EX_ade_e_</td>\n",
       "      <td>-0.0001514</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>-11.48</td>\n",
       "      <td>4</td>\n",
       "      <td>20.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-153.9</td>\n",
       "      <td>1</td>\n",
       "      <td>67.13%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-14.43</td>\n",
       "      <td>2</td>\n",
       "      <td>12.59%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-115.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-46.52</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cd94b3b80>"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdm(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "c960e06a-d15f-4fed-969a-73dfa121fc0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 3.3887632661248945</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ala_L_e</td>\n",
       "      <td>EX_ala_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>7.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>1.132</td>\n",
       "      <td>6</td>\n",
       "      <td>1.73%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.8526</td>\n",
       "      <td>4</td>\n",
       "      <td>0.87%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>0.6407</td>\n",
       "      <td>4</td>\n",
       "      <td>0.65%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.01605</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.01605</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.2666</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.04968</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc_D_e</td>\n",
       "      <td>EX_glc_D_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>15.27%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gln_L_e</td>\n",
       "      <td>EX_gln_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>12.72%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>12.72%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>5.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.3427</td>\n",
       "      <td>6</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hxan_e</td>\n",
       "      <td>EX_hxan_e_</td>\n",
       "      <td>0.9574</td>\n",
       "      <td>5</td>\n",
       "      <td>1.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.9892</td>\n",
       "      <td>6</td>\n",
       "      <td>1.51%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.6018</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lac_L_e</td>\n",
       "      <td>EX_lac_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>7.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>1.602</td>\n",
       "      <td>6</td>\n",
       "      <td>2.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>1.19</td>\n",
       "      <td>6</td>\n",
       "      <td>1.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.3475</td>\n",
       "      <td>5</td>\n",
       "      <td>0.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.02675</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>196</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.6744</td>\n",
       "      <td>9</td>\n",
       "      <td>1.54%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>2.479</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>12.72%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>7.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.01338</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thm_e</td>\n",
       "      <td>EX_thm_e_</td>\n",
       "      <td>0.0008886</td>\n",
       "      <td>12</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.9794</td>\n",
       "      <td>4</td>\n",
       "      <td>1.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.2336</td>\n",
       "      <td>11</td>\n",
       "      <td>0.65%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.4837</td>\n",
       "      <td>9</td>\n",
       "      <td>1.11%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ura_e</td>\n",
       "      <td>EX_ura_e_</td>\n",
       "      <td>0.8863</td>\n",
       "      <td>4</td>\n",
       "      <td>0.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>1.55</td>\n",
       "      <td>5</td>\n",
       "      <td>1.97%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.001518</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-6.743E-06</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-3.389</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.001511</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.05898</td>\n",
       "      <td>10</td>\n",
       "      <td>0.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.02622</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ade_e</td>\n",
       "      <td>EX_ade_e_</td>\n",
       "      <td>-0.0001513</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-202</td>\n",
       "      <td>1</td>\n",
       "      <td>79.28%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-26.08</td>\n",
       "      <td>2</td>\n",
       "      <td>20.47%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-144.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-57.26</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cd94b3e50>"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Note MDM with lactate, model doesn't actually consume any lactate. The TCA cycle also has zero flux.\n",
    "#This doesn't appear to be due to glucose, lactate, or o2 levels, which are equivalent in both situations.\n",
    "#Use of lactate requires asp_L, asn_L, ser_L, and gly to be limited. \n",
    "#Asn --> ASP -->fum --> mal-->nadh and pyruvate\n",
    "#glycine -->serine --> pyr\n",
    "\n",
    "\n",
    "#        if 'EX_asp_L_e_' in reaction.id:\n",
    "#            reaction.lower_bound = -3.76\n",
    "#        if 'EX_asn_L_e_' in reaction.id:\n",
    "#            reaction.lower_bound = -0.17\n",
    "#        if 'EX_ser_L_e_' in reaction.id:\n",
    "#            reaction.lower_bound = -0.48\n",
    "#        if 'EX_gly_e_' in reaction.id:\n",
    "#            reaction.lower_bound = -0.33\n",
    "\n",
    "mdm_with_lactate(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "7804318b-5fbb-4d6c-8011-1885f1314648",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 2.7799578010748527</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ala_L_e</td>\n",
       "      <td>EX_ala_L_e_</td>\n",
       "      <td>1.12</td>\n",
       "      <td>3</td>\n",
       "      <td>1.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.71</td>\n",
       "      <td>6</td>\n",
       "      <td>1.31%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.17</td>\n",
       "      <td>4</td>\n",
       "      <td>0.21%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>3.76</td>\n",
       "      <td>4</td>\n",
       "      <td>4.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.01317</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.01317</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.008779</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.008779</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.65</td>\n",
       "      <td>3</td>\n",
       "      <td>0.60%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.04075</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc_D_e</td>\n",
       "      <td>EX_glc_D_e_</td>\n",
       "      <td>27.8</td>\n",
       "      <td>6</td>\n",
       "      <td>51.32%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gln_L_e</td>\n",
       "      <td>EX_gln_L_e_</td>\n",
       "      <td>0.34</td>\n",
       "      <td>5</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>8.84</td>\n",
       "      <td>5</td>\n",
       "      <td>13.60%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>0.33</td>\n",
       "      <td>2</td>\n",
       "      <td>0.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.13</td>\n",
       "      <td>6</td>\n",
       "      <td>0.24%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hxan_e</td>\n",
       "      <td>EX_hxan_e_</td>\n",
       "      <td>0.37</td>\n",
       "      <td>5</td>\n",
       "      <td>0.57%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.23</td>\n",
       "      <td>6</td>\n",
       "      <td>0.42%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.4937</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lac_L_e</td>\n",
       "      <td>EX_lac_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>18.46%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.69</td>\n",
       "      <td>6</td>\n",
       "      <td>1.27%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.27</td>\n",
       "      <td>6</td>\n",
       "      <td>0.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.02195</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.008779</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.008779</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>4.11</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>110.6</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.15</td>\n",
       "      <td>9</td>\n",
       "      <td>0.42%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>2.034</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.43</td>\n",
       "      <td>5</td>\n",
       "      <td>0.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>0.48</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.01098</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thm_e</td>\n",
       "      <td>EX_thm_e_</td>\n",
       "      <td>0.0007116</td>\n",
       "      <td>12</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.42</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.1917</td>\n",
       "      <td>11</td>\n",
       "      <td>0.65%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.39</td>\n",
       "      <td>9</td>\n",
       "      <td>1.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ura_e</td>\n",
       "      <td>EX_ura_e_</td>\n",
       "      <td>0.45</td>\n",
       "      <td>4</td>\n",
       "      <td>0.55%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.51</td>\n",
       "      <td>5</td>\n",
       "      <td>0.78%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.008779</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.001245</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-5.531E-06</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-2.78</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.00124</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.04838</td>\n",
       "      <td>10</td>\n",
       "      <td>0.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.02151</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-39.52</td>\n",
       "      <td>2</td>\n",
       "      <td>37.35%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-118.3</td>\n",
       "      <td>1</td>\n",
       "      <td>55.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-6.5</td>\n",
       "      <td>2</td>\n",
       "      <td>6.14%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-145.7</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>-0.1462</td>\n",
       "      <td>5</td>\n",
       "      <td>0.35%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cd94b3280>"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#In MDM_molarity with lactate, model DOES consume lactate. \n",
    "mdm_molarity_with_lactate(model)\n",
    "model.summary()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "4e434a43-dff9-4a92-b0d2-cc997aef22d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 2.287793744474097</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ala_L_e</td>\n",
       "      <td>EX_ala_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>9.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.764</td>\n",
       "      <td>6</td>\n",
       "      <td>1.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.5756</td>\n",
       "      <td>4</td>\n",
       "      <td>0.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.01084</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.01084</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.18</td>\n",
       "      <td>3</td>\n",
       "      <td>0.18%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.03354</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gln_L_e</td>\n",
       "      <td>EX_gln_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>16.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>16.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>6.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.2314</td>\n",
       "      <td>6</td>\n",
       "      <td>0.45%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hxan_e</td>\n",
       "      <td>EX_hxan_e_</td>\n",
       "      <td>0.6464</td>\n",
       "      <td>5</td>\n",
       "      <td>1.05%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.6678</td>\n",
       "      <td>6</td>\n",
       "      <td>1.30%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.4063</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lac_L_e</td>\n",
       "      <td>EX_lac_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>9.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>1.081</td>\n",
       "      <td>6</td>\n",
       "      <td>2.11%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.8031</td>\n",
       "      <td>6</td>\n",
       "      <td>1.57%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.2346</td>\n",
       "      <td>5</td>\n",
       "      <td>0.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.01806</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.4553</td>\n",
       "      <td>9</td>\n",
       "      <td>1.33%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>1.674</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>16.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>9.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.009032</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thm_e</td>\n",
       "      <td>EX_thm_e_</td>\n",
       "      <td>0.000603</td>\n",
       "      <td>12</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.6612</td>\n",
       "      <td>4</td>\n",
       "      <td>0.86%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.1577</td>\n",
       "      <td>11</td>\n",
       "      <td>0.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.3265</td>\n",
       "      <td>9</td>\n",
       "      <td>0.96%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ura_e</td>\n",
       "      <td>EX_ura_e_</td>\n",
       "      <td>0.5984</td>\n",
       "      <td>4</td>\n",
       "      <td>0.78%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>1.046</td>\n",
       "      <td>5</td>\n",
       "      <td>1.70%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.001025</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-4.552E-06</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-2.288</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.00102</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.03982</td>\n",
       "      <td>10</td>\n",
       "      <td>0.19%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.0177</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ade_e</td>\n",
       "      <td>EX_ade_e_</td>\n",
       "      <td>-0.0001177</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>-6.065</td>\n",
       "      <td>4</td>\n",
       "      <td>11.32%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-147.9</td>\n",
       "      <td>1</td>\n",
       "      <td>69.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-20.85</td>\n",
       "      <td>2</td>\n",
       "      <td>19.47%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-89.43</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-54.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cdfb7e400>"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdm_with_lactate_only(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "ffec2132-8bd2-4dc9-836e-9f6ba6c25cce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.97565516447302</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ala_L_e</td>\n",
       "      <td>EX_ala_L_e_</td>\n",
       "      <td>1.12</td>\n",
       "      <td>3</td>\n",
       "      <td>4.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.3258</td>\n",
       "      <td>6</td>\n",
       "      <td>2.69%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.17</td>\n",
       "      <td>4</td>\n",
       "      <td>0.94%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.004622</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.004622</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.003081</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.003081</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.0768</td>\n",
       "      <td>3</td>\n",
       "      <td>0.32%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.0143</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gln_L_e</td>\n",
       "      <td>EX_gln_L_e_</td>\n",
       "      <td>0.34</td>\n",
       "      <td>5</td>\n",
       "      <td>2.34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>8.84</td>\n",
       "      <td>5</td>\n",
       "      <td>60.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>0.33</td>\n",
       "      <td>2</td>\n",
       "      <td>0.91%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>357.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.09867</td>\n",
       "      <td>6</td>\n",
       "      <td>0.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hxan_e</td>\n",
       "      <td>EX_hxan_e_</td>\n",
       "      <td>0.2756</td>\n",
       "      <td>5</td>\n",
       "      <td>1.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.23</td>\n",
       "      <td>6</td>\n",
       "      <td>1.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.1733</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.4611</td>\n",
       "      <td>6</td>\n",
       "      <td>3.81%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.27</td>\n",
       "      <td>6</td>\n",
       "      <td>2.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.69%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.007703</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.003081</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.003081</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>97.68</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.15</td>\n",
       "      <td>9</td>\n",
       "      <td>1.86%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.7138</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.43</td>\n",
       "      <td>5</td>\n",
       "      <td>2.96%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>0.48</td>\n",
       "      <td>3</td>\n",
       "      <td>1.98%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.003852</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thm_e</td>\n",
       "      <td>EX_thm_e_</td>\n",
       "      <td>0.0002199</td>\n",
       "      <td>12</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.3368</td>\n",
       "      <td>4</td>\n",
       "      <td>1.86%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.06727</td>\n",
       "      <td>11</td>\n",
       "      <td>1.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.1393</td>\n",
       "      <td>9</td>\n",
       "      <td>1.73%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ura_e</td>\n",
       "      <td>EX_ura_e_</td>\n",
       "      <td>0.2552</td>\n",
       "      <td>4</td>\n",
       "      <td>1.41%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.4461</td>\n",
       "      <td>5</td>\n",
       "      <td>3.07%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.003081</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0004371</td>\n",
       "      <td>5</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-1.941E-06</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.9757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0004351</td>\n",
       "      <td>5</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.01698</td>\n",
       "      <td>10</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.007549</td>\n",
       "      <td>1</td>\n",
       "      <td>0.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>-3.809</td>\n",
       "      <td>4</td>\n",
       "      <td>46.51%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-8.663</td>\n",
       "      <td>1</td>\n",
       "      <td>26.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-4.339</td>\n",
       "      <td>2</td>\n",
       "      <td>26.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-185.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-3.878</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cdd38aee0>"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def mdm_molarity_with_no_carbon_source(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -1000.\n",
    "        if 'EX_glc_D_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0\n",
    "        if 'EX_ser_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.48\n",
    "        if 'EX_asp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -3.76\n",
    "        if 'EX_asn_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.17\n",
    "        if 'EX_glu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -8.84\n",
    "        if 'EX_gly_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.33\n",
    "        if 'EX_ala_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.12\n",
    "        if 'EX_val_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.51\n",
    "        if 'EX_leu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.69\n",
    "        if 'EX_gln_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.34\n",
    "        if 'EX_pro_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.43\n",
    "        if 'EX_phe_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_cl_e_' in reaction.id:\n",
    "            reaction.lower_bound = -107.94697\n",
    "        if 'EX_k_e_' in reaction.id:\n",
    "            reaction.lower_bound = -11.48\n",
    "        if 'EX_so4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -5.74\n",
    "        if 'EX_nh4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -4.11\n",
    "        if 'EX_mg2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.07\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_pi_e_' in reaction.id:\n",
    "            reaction.lower_bound = -40\n",
    "        if 'EX_ca2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.254\n",
    "        if 'EX_no3_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.051\n",
    "        if 'EX_arg_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.71\n",
    "        if 'EX_ile_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.23\n",
    "        if 'EX_tyr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_cys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.65\n",
    "        if 'EX_trp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_thr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.42\n",
    "        if 'EX_his_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.13\n",
    "        if 'EX_met_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1\n",
    "        if 'EX_gthrd_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_thm_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.00697\n",
    "        if 'EX_pnto_R_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.004\n",
    "        if 'EX_hxan_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.37\n",
    "        if 'EX_ura_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.45  \n",
    "        if 'EX_na1_e_' in reaction.id:\n",
    "            reaction.lower_bound = -105.018\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -0\n",
    "        if reaction.id == 'EX_co2_e_':\n",
    "            reaction.lower_bound = -0\n",
    "\n",
    "mdm_molarity_with_no_carbon_source(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "a76e975c-cc60-4c5d-b7be-441d2f21fa57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7.887191036066456e-16\n",
      "no_p\n",
      "====\n",
      "Formula: NO\n",
      "\n",
      "Producing Reactions\n",
      "-------------------\n",
      "Percent  Flux Reaction                                             Definition\n",
      "100.00% 1.709     NO2R ficytC_p + h2o_p + no_p <=> focytC_p + 2.0 h_p + no2_p\n",
      "\n",
      "Consuming Reactions\n",
      "-------------------\n",
      "Percent   Flux Reaction                                                         Definition\n",
      "100.00% -1.709      NOR 2.0 ficytC_p + 2.0 h_c + 2.0 no_p <=> 2.0 focytC_p + h2o_p + n2o_p\n"
     ]
    }
   ],
   "source": [
    "#Check Gc model ability to grow anaerobically on nitrite. \n",
    "def anaerobic_rpmi(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -0.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_no2_e_':\n",
    "            reaction.lower_bound = -10. \n",
    "\n",
    "anaerobic_rpmi(model)\n",
    "print(model.slim_optimize())\n",
    "print(model.metabolites.no_p.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "0962ee2f-effa-4b9c-9737-06f73a4a4552",
   "metadata": {},
   "outputs": [],
   "source": [
    "#NO2R for nitrite reductase (aniA) is switched in terms of ficytC and focytC, which prevents use of \n",
    "#nitrite to support anaerobic growth. This also allows spontaneous creation of energy through ficytC when grwon on NO.\n",
    "#Correcting here. \n",
    "\n",
    "#r=model.reactions.get_by_id('NO2R')\n",
    "#ficytC_p=model.metabolites.get_by_id('ficytC_p')\n",
    "#focytC_p=model.metabolites.get_by_id('focytC_p')\n",
    "\n",
    "#r.add_metabolites({\n",
    "#    ficytC_p: 2.0,\n",
    "#    focytC_p: -2.0\n",
    "#  \n",
    "#})\n",
    "#r\n",
    "\n",
    "#This was actually okay as it was?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "2ea99ac5-122f-4f29-951b-82597302f5bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7.887191036066456e-16\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<h3>no_p</h3><p>NO</p><h4>Producing Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Percent</th>\n",
       "      <th>Flux</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>100.00%</td>\n",
       "      <td>1.709</td>\n",
       "      <td>NO2R</td>\n",
       "      <td>ficytC_p + h2o_p + no_p &lt;=&gt; focytC_p + 2.0 h_p + no2_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Consuming Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Percent</th>\n",
       "      <th>Flux</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>100.00%</td>\n",
       "      <td>-1.709</td>\n",
       "      <td>NOR</td>\n",
       "      <td>2.0 ficytC_p + 2.0 h_c + 2.0 no_p &lt;=&gt; 2.0 focytC_p + h2o_p + n2o_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.metabolite_summary.MetaboliteSummary at 0x19cdefffd30>"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Model now grows anaerobically on nitrite. \n",
    "anaerobic_rpmi(model)\n",
    "print(model.slim_optimize())\n",
    "model.metabolites.no_p.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "fadfcad0-5e94-46f1-8bef-0cc44ab34bce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14.96 minutes on MDM_molarity with glucose -20 and oxygen -1000\n",
      "43.56 minutes on MDM_molarity with glucose -20 and oxygen -20\n",
      "42.627 minutes on MDM_molarity with glucose 0 and oxygen -1000\n",
      "115.888 minutes on MDM_molarity with glucose 0 and oxygen -20\n"
     ]
    }
   ],
   "source": [
    "#The ability of Gc to grow on glucose is highly dependent on oxygen exchange.\n",
    "#IF oxygen is limited in MDM, the doubling time on glucose is increased,\n",
    "#but the doubling time without a dedicated carbon source (aka on amino acids) is completely unchanged. \n",
    "\n",
    "mdm_molarity(model)\n",
    "\n",
    "with model:\n",
    "    for reaction in model.reactions:\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "    print (doubling_time(model) + ' on MDM_molarity with glucose -20 and oxygen -1000')\n",
    "\n",
    "with model:\n",
    "    for reaction in model.reactions:\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "    print (doubling_time(model) + ' on MDM_molarity with glucose -20 and oxygen -20')   \n",
    "\n",
    "with model:\n",
    "    for reaction in model.reactions:\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -0.\n",
    "    print (doubling_time(model) + ' on MDM_molarity with glucose 0 and oxygen -1000')    \n",
    "    \n",
    "with model:\n",
    "    for reaction in model.reactions:\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -0.\n",
    "    print (doubling_time(model) + ' on MDM_molarity with glucose 0 and oxygen -20')\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fa821db-1c74-431c-a771-aa681a1223d3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b482dc5-7763-4eb3-9e48-5f8eb83aeae0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "07808461-691d-419b-b507-978498086c42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>CYSabcpp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_CYSabcpp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019cd1580f40</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>atp_c + cys_L_p + h2o_c --> adp_c + cys_L_c + h_c + pi_c</p>\n",
       "                    <p style='text-align:right'>M_atp_c + M_cys_L_p + M_h2o_c --> M_adp_c + M_cys_L_c + M_h_c + M_pi_c</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>(NGO0374 and NGO0373 and NGO0372) or (NGO0968 and NGO0269 and NGO1152)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction CYSabcpp at 0x19cd1580f40>"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('CYSabcpp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "fb13f499-0ac9-40a5-92c1-343eab8d39e0",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{}\n"
     ]
    }
   ],
   "source": [
    "#NGO2011, NGO2012, NGO2013, and NGO2014 are a cysteine (cys_L_) transporter. PMID: 22138345\n",
    "#NGO0372, NGO373, NGO0374 are a cystine (cystine_L_) transporter. PMID: 22138345. In the current model, this is annotated as a cysteine importer \n",
    "#and no reaction for cystine import exists, or for breakdown of cystine. \n",
    "#Adding that transport and breakdown reaction here and correcting the gene annotations.\n",
    "\n",
    "#NGO0968 and NGO0269 and NGO1152 is listed for lots of different transport reactions - its not very well annotated. Removing it from cystine and cysteine import. \n",
    "\n",
    "CYSabcpp = model.reactions.get_by_id('CYSabcpp')\n",
    "CYSabcpp.gene_reaction_rule='NGO2011 and NGO2012 and NGO2013 and NGO2014'\n",
    "CYSabcpp.notes = {'citations': 'PMID: 22138345', 'type': '', 'annotation': ''}\n",
    "\n",
    "\n",
    "from cobra import Reaction, Metabolite\n",
    "\n",
    "cystine_L_c = Metabolite(\n",
    "    'cystine_L_c',\n",
    "    formula='C6H12N2O4S2',\n",
    "    name='Cystine',\n",
    "    compartment='cytosol')\n",
    "\n",
    "cystine_L_e = Metabolite(\n",
    "    'cystine_L_e',\n",
    "    formula='C6H12N2O4S2',\n",
    "    name='Cystine',\n",
    "    compartment='extracellular')\n",
    "\n",
    "cystine_L_p = Metabolite(\n",
    "    'cystine_L_p',\n",
    "    formula='C6H12N2O4S2',\n",
    "    name='Cystine',\n",
    "    compartment='periplasm')\n",
    "\n",
    "\n",
    "h_c = model.metabolites.get_by_id('h_c')\n",
    "atp_c = model.metabolites.get_by_id('atp_c')\n",
    "h2o_c = model.metabolites.get_by_id('h2o_c')\n",
    "adp_c = model.metabolites.get_by_id('adp_c')\n",
    "pi_c = model.metabolites.get_by_id('pi_c')\n",
    "\n",
    "CYSTINEabcpp = cobra.Reaction('CYSTINEabcpp')\n",
    "CYSTINEabcpp.id = 'CYSTINEabcpp'\n",
    "CYSTINEabcpp.name = 'cystine import'\n",
    "CYSTINEabcpp.gene_reaction_rule = 'NGO0374 and NGO0373 and NGO0372'\n",
    "CYSTINEabcpp.lower_bound = -1000.\n",
    "CYSTINEabcpp.upper_bound = 1000.\n",
    "CYSTINEabcpp.add_metabolites({\n",
    "        atp_c: -1.0,\n",
    "        cystine_L_p: -1.,\n",
    "        h2o_c: -1. ,\n",
    "        adp_c:  1.,\n",
    "        cystine_L_c: 1. ,\n",
    "        h_c: 1.,\n",
    "        pi_c: 1.,\n",
    "        })\n",
    "\n",
    "model.add_reactions([CYSTINEabcpp])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "EX_cystine_L_ = cobra.Reaction('EX_cystine_L_')\n",
    "EX_cystine_L_.notes = {'citations': 'PMID: 22138345', 'type': '', 'annotation': ''}\n",
    "EX_cystine_L_.name = 'exchange of cystine'\n",
    "EX_cystine_L_.gene_reaction_rule = ''\n",
    "EX_cystine_L_.lower_bound = 0.\n",
    "EX_cystine_L_.upper_bound = 1000.\n",
    "EX_cystine_L_.add_metabolites({\n",
    "    cystine_L_e: -1.0,})\n",
    "\n",
    "\n",
    "model.add_reactions([EX_cystine_L_])\n",
    "\n",
    "\n",
    "h_c = model.metabolites.get_by_id('h_c')\n",
    "nadh_c = model.metabolites.get_by_id('nadh_c')\n",
    "nad_c = model.metabolites.get_by_id('nad_c')\n",
    "cys_L_c = model.metabolites.get_by_id('cys_L_c')\n",
    "\n",
    "CYSR = cobra.Reaction('CYSR')\n",
    "CYSR.notes = {'citations': 'PMCID: PMC2323225', 'type': '', 'annotation': ''}\n",
    "CYSR.name = 'cystine reductase'\n",
    "CYSR.gene_reaction_rule = 'Orphan'\n",
    "CYSR.lower_bound = -1000.\n",
    "CYSR.upper_bound = 1000.\n",
    "CYSR.add_metabolites({\n",
    "    cys_L_c: -2.0,\n",
    "    nad_c: -1.0, \n",
    "    cystine_L_c: 1.0,\n",
    "    h_c: 1.0,\n",
    "    nadh_c: 1.0,\n",
    "    })\n",
    "\n",
    "\n",
    "model.add_reactions([CYSR])\n",
    "print(CYSR.check_mass_balance())\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "8ddebaab-c3e9-4195-95e8-e3359e3f33ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Aimee\\anaconda3\\lib\\site-packages\\cobra\\core\\model.py:727: UserWarning: need to pass in a list\n",
      "  warn(\"need to pass in a list\")\n"
     ]
    }
   ],
   "source": [
    "#Neisseria, like many other proteobacteria, lack the permease portion of the phosphotransferase. \n",
    "#And therefore do not use it for carbohydrate transport.\n",
    "#The remaining portions of the system are thought to perform regulatory roles, related to virulence. \n",
    "#PMID: 26858137\n",
    "\n",
    "model.remove_reactions(model.reactions.GLCptspp)\n",
    "model.remove_reactions(model.reactions.FRUpts2pp)\n",
    "model.remove_reactions(model.reactions.ASCBptspp)\n",
    "model.remove_reactions(model.reactions.MALTptspp)\n",
    "model.remove_reactions(model.reactions.MANptspp)\n",
    "model.remove_reactions(model.reactions.SUCptspp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "3b61b79e-ee11-4524-883f-265fe123ce93",
   "metadata": {},
   "outputs": [],
   "source": [
    "#NADPH does not catalyze the FabI reaction, only NADH. Therefore all of the NADPH dependent reactions were removed. \n",
    "#PMID: 26567338   \n",
    "model.remove_reactions(model.reactions.EAR120y)\n",
    "model.remove_reactions(model.reactions.EAR100y)\n",
    "model.remove_reactions(model.reactions.EAR121y)\n",
    "model.remove_reactions(model.reactions.EAR140y)\n",
    "model.remove_reactions(model.reactions.EAR141y)\n",
    "model.remove_reactions(model.reactions.EAR160y)\n",
    "model.remove_reactions(model.reactions.EAR161y)\n",
    "model.remove_reactions(model.reactions.EAR180y)\n",
    "model.remove_reactions(model.reactions.EAR181y)\n",
    "model.remove_reactions(model.reactions.EAR40y)\n",
    "model.remove_reactions(model.reactions.EAR60y)\n",
    "model.remove_reactions(model.reactions.EAR80y)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "72f0d8d1-d2af-4865-a4c7-98111a62a6e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Gc can only grow on glucose, lactate, and pyruvate. It cannot grow on gluconate. \n",
    "#Despite similarities to gluconate transporters and gluconate kinase, Nm also can't grow on gluconate so this should be removed. \n",
    "#PMID: 27454890\n",
    "model.remove_reactions(model.reactions.GLCNtex)\n",
    "model.remove_reactions(model.reactions.EX_glcn_e_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "7d4fb26e-0ff2-4156-92c8-8748fe3bad17",
   "metadata": {},
   "outputs": [],
   "source": [
    "#To remove an infeasible loop, where gln is exchanged infinitely between the two Na+ and h+ dependent transporters, limited these to import only.\n",
    "\n",
    "model.reactions.GLNt2pp.lower_bound =0\n",
    "model.reactions.GLNt1pp.lower_bound =0  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "357e0690-d3cb-454b-80d7-db5f6311ea89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{}\n",
      "{}\n"
     ]
    }
   ],
   "source": [
    "#D_ala import is predicted to occur through an orphan reaction and spontaneously\n",
    "#Linking D_ala import to L_ala import genetically and na+ and h+ according to the transporters as appropriate\n",
    "\n",
    "\n",
    "ALADt2pp =model.reactions.ALADt\n",
    "\n",
    "ALADt2pp.id = 'ALADt2pp'\n",
    "ALADt2pp.gene_reaction_rule = 'NGO1807 or NGO1787 or NGO1290 or NGO1957'\n",
    "\n",
    "h_p = model.metabolites.get_by_id('h_p')\n",
    "h_c = model.metabolites.get_by_id('h_c')\n",
    "na1_c = model.metabolites.get_by_id('na1_c')\n",
    "na1_p = model.metabolites.get_by_id('na1_p')\n",
    "ala_D_p = model.metabolites.get_by_id('ala_D_p')\n",
    "ala_D_c = model.metabolites.get_by_id('ala_D_c')\n",
    "\n",
    "ALADt2pp.add_metabolites({\n",
    "        na1_p: -1.0,\n",
    "        na1_c: 1.0\n",
    "        })\n",
    "\n",
    "print(ALADt2pp.check_mass_balance())\n",
    "\n",
    "ALADt1pp = cobra.Reaction('ALADt1pp')\n",
    "ALADt1pp.name = 'D_ala import'\n",
    "ALADt1pp.gene_reaction_rule = 'NGO1290 or NGO1787'\n",
    "ALADt1pp.lower_bound = -1000.\n",
    "ALADt1pp.upper_bound = 1000.\n",
    "ALADt1pp.add_metabolites({\n",
    "    ala_D_p: -1.0,\n",
    "    h_p: -1.0,\n",
    "    ala_D_c: 1.0,\n",
    "    h_c: 1.0\n",
    "    })\n",
    "\n",
    "\n",
    "print(ALADt1pp.check_mass_balance())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "1f377ace-d5d4-4277-b2b8-b8f4e8693496",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove looping reactions\n",
    "model.reactions.ACKr.upper_bound=0\n",
    "model.reactions.ALAt1pp.lower_bound = 0\n",
    "model.reactions.ALAt2pp.lower_bound = 0\n",
    "model.reactions.ARGt1pp.lower_bound = 0\n",
    "model.reactions.ARGt2pp.lower_bound = 0\n",
    "model.reactions.ASNt1pp.lower_bound = 0\n",
    "model.reactions.ASNt2pp.lower_bound = 0\n",
    "model.reactions.ASPt1pp.lower_bound = 0\n",
    "model.reactions.ASPt2pp.lower_bound = 0\n",
    "model.reactions.CYSt1pp.lower_bound = 0\n",
    "model.reactions.CYSt2pp.lower_bound = 0\n",
    "model.reactions.GLUt1pp.lower_bound = 0\n",
    "model.reactions.GLUt2pp.lower_bound = 0\n",
    "model.reactions.HISt1pp.lower_bound = 0\n",
    "model.reactions.HISt2pp.lower_bound = 0\n",
    "model.reactions.ILEt1pp.lower_bound = 0\n",
    "model.reactions.ILEt3pp.lower_bound = 0\n",
    "model.reactions.LEUt1pp.lower_bound = 0\n",
    "model.reactions.LEUt3pp.lower_bound = 0\n",
    "model.reactions.LYSt1pp.lower_bound = 0\n",
    "model.reactions.LYSt2pp.lower_bound = 0\n",
    "model.reactions.METt1pp.lower_bound = 0\n",
    "model.reactions.METt2pp.lower_bound = 0\n",
    "model.reactions.PHEt1pp.lower_bound = 0\n",
    "model.reactions.PHEt2pp.lower_bound = 0\n",
    "model.reactions.PROt1pp.lower_bound = 0\n",
    "model.reactions.PROt2rpp.lower_bound = 0\n",
    "model.reactions.SERt1rpp.lower_bound = 0\n",
    "model.reactions.SERt2pp.lower_bound = 0\n",
    "model.reactions.THRt1rpp.lower_bound = 0\n",
    "model.reactions.THRt2pp.lower_bound = 0\n",
    "model.reactions.VALt1pp.lower_bound = 0\n",
    "model.reactions.VALt3pp.lower_bound = 0\n",
    "model.reactions.TRPS1.lower_bound=0\n",
    "\n",
    "model.reactions.ADK1.lower_bound=0\n",
    "model.reactions.NDPK1.lower_bound=0\n",
    "model.reactions.VALTA.upper_bound=0\n",
    "model.reactions.VPAMT.upper_bound=0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "45597d95-f2e8-4bd6-b0a3-ecee60e1be81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>AE004969.1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019ccfff56a0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1337</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1504</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*Nm_Ess_biomass - 1.0*Nm_Ess_biomass_reverse_46af5</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Cytosol, C_c, C_p, c, periplasm, cytosol, extracellular</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model AE004969.1 at 0x19ccfff56a0>"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "a2c07e2c-ac72-4313-917b-b5a8f07decde",
   "metadata": {},
   "outputs": [],
   "source": [
    "for metabolite in model.metabolites:\n",
    "    if '_c' in metabolite.id:\n",
    "        metabolite.compartment = 'cytosol'\n",
    "    if '_p' in metabolite.id:\n",
    "        metabolite.compartment = 'periplasm'\n",
    "    if '_e' in metabolite.id:\n",
    "        metabolite.compartment = 'extracellular'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "546633ae-7b34-458e-9f64-88853503baae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>AE004969.1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019ccfff56a0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1337</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1504</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*Nm_Ess_biomass - 1.0*Nm_Ess_biomass_reverse_46af5</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>periplasm, cytosol, extracellular, Cytosol</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model AE004969.1 at 0x19ccfff56a0>"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "f3b385ca-1155-4a98-b7ea-38d479740fa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identify any compartments that were miss added\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.compartment == 'C_c':\n",
    "        print (metabolite.id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "ab2e32a6-43f1-4272-b5f4-39ead016c4ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identify any compartments that were miss added\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.compartment == 'C_p':\n",
    "        print (metabolite.id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "2bb29feb-0f20-4124-bb81-b289b1145da8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identify any compartments that were miss added\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.compartment == 'c':\n",
    "        print (metabolite.id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "05988cec-ce0a-407c-b497-6f414bf9441f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "essbiomass\n",
      "protein\n",
      "capminusbiomass\n",
      "biomass\n"
     ]
    }
   ],
   "source": [
    "#Identify any compartments that were miss added\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.compartment == 'Cytosol':\n",
    "        print (metabolite.id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "82eb2677-3010-4cd2-bbca-b101b4f6386b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Correct any compartments that were miss added\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.compartment == 'C_c':\n",
    "        metabolite.compartment= 'cytosol'\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.compartment == 'c':\n",
    "        metabolite.compartment= 'cytosol'\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.compartment == 'C_p':\n",
    "        metabolite.compartment= 'periplasm'\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.compartment == 'Cytosol':\n",
    "        metabolite.compartment= 'cytosol'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "34378ae7-9956-4da3-a17d-3aa846bcf3d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>AE004969.1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x019ccfff56a0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1337</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1504</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*Nm_Ess_biomass - 1.0*Nm_Ess_biomass_reverse_46af5</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>periplasm, cytosol, extracellular</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model AE004969.1 at 0x19ccfff56a0>"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "714d59a2-f076-4fdf-96a9-b5fb1fc29033",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "c68383d2-c4f2-47d4-949d-f5eade1e8ba7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "559"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.genes)\n",
    "#Subtract 2 for \"Sgene, and orphan\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "ec1df743-c896-47f0-87d4-49d9acdb34b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.id='iNGO_557'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "060a19bc-fe2c-4a2e-9f7b-e005a2cb7cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "cobra.io.write_sbml_model(model, 'C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/curated_models/carvemeupdatedGCmodel.xml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "9d39270e-3346-4975-9327-6970b28b7e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "cobra.io.save_json_model(model, \"C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/curated_models/carvemeupdatedGCmodel.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "258228fc-1dc9-4042-b95e-d817cf6920d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Send model to modelPolisher\n",
    "\n",
    "#Cd ModelPolisher\n",
    "#docker-compose run -v C:\\Users\\Aimee\\models:/models/ polisher java -jar /ModelPolisher-2.1-beta.jar --input=/models/carvemeupdatedGCmodel.sbml --output=/models/output/carvemeupdatedGCmodel.sbml --annotate-with-bigg=true --add-adb-annotations=true --output-combine=true\n",
    "\n",
    "#Extract from Zip file (Users-->Aimee-->models)\n",
    "#Save model to \"C://Users/Aimee/Documents/UVA/Metabolic_Modeling/organized/models/curated_models/carvemeupdatedGCmodelpolished.sbml\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "0a4e6c2f-a9fa-4bb1-9dcc-ce3eeee72d8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 2.9362979549360113</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ala_L_e</td>\n",
       "      <td>EX_ala_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>8.69%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.9806</td>\n",
       "      <td>6</td>\n",
       "      <td>1.71%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.01391</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.01391</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.009273</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.009273</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.2311</td>\n",
       "      <td>3</td>\n",
       "      <td>0.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.04304</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc_D_e</td>\n",
       "      <td>EX_glc_D_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>17.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gln_L_e</td>\n",
       "      <td>EX_gln_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>14.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>14.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>5.80%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.297</td>\n",
       "      <td>6</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hxan_e</td>\n",
       "      <td>EX_hxan_e_</td>\n",
       "      <td>0.8295</td>\n",
       "      <td>5</td>\n",
       "      <td>1.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.8571</td>\n",
       "      <td>6</td>\n",
       "      <td>1.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.5215</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>1.388</td>\n",
       "      <td>6</td>\n",
       "      <td>2.41%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>1.031</td>\n",
       "      <td>6</td>\n",
       "      <td>1.79%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.301</td>\n",
       "      <td>5</td>\n",
       "      <td>0.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.02318</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.009273</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.009273</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>171.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.5844</td>\n",
       "      <td>9</td>\n",
       "      <td>1.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>2.148</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>14.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>8.69%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.01159</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thm_e</td>\n",
       "      <td>EX_thm_e_</td>\n",
       "      <td>0.0007686</td>\n",
       "      <td>12</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.8486</td>\n",
       "      <td>4</td>\n",
       "      <td>0.98%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.2025</td>\n",
       "      <td>11</td>\n",
       "      <td>0.65%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.4191</td>\n",
       "      <td>9</td>\n",
       "      <td>1.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>1.343</td>\n",
       "      <td>5</td>\n",
       "      <td>1.95%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.009273</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.001315</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-5.843E-06</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-2.936</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.00131</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.05111</td>\n",
       "      <td>10</td>\n",
       "      <td>0.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.02272</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-175.6</td>\n",
       "      <td>1</td>\n",
       "      <td>77.97%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-24.54</td>\n",
       "      <td>2</td>\n",
       "      <td>21.79%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-126</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-55.39</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19ce14fef40>"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdm(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3e6f30e-4735-4865-9fe4-45fc97ea1bd5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
