{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "49c4429d-c3cb-4354-9ae1-36987fe42c19",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Import libraries - REQUIRES pip version 9.0.3\n",
    "import pandas\n",
    "import pandas as pd\n",
    "import os\n",
    "from os.path import join\n",
    "import sys\n",
    "import scipy.stats\n",
    "import numpy\n",
    "import math\n",
    "import pickle\n",
    "import copy\n",
    "import time\n",
    "import warnings\n",
    "import gc\n",
    "\n",
    "#Simplify reading/writing files\n",
    "cwd=os.path.realpath(os.path.join(os.path.dirname(os.getcwd()),\"..\",\"..\"))\n",
    "\n",
    "# Using Cobrapy 0.13.0\n",
    "import cobra\n",
    "import cobra.test\n",
    "import cobra.flux_analysis.gapfilling\n",
    "from cobra import Reaction, Metabolite\n",
    "from cobra.io import write_sbml_model\n",
    "from cobra.flux_analysis import  flux_variability_analysis\n",
    "from cobra.flux_analysis.reaction import assess_component\n",
    "from cobra.manipulation.delete import *\n",
    "from cobra.flux_analysis.parsimonious import add_pfba\n",
    "from cobra.medium import find_boundary_types\n",
    "from cobra.util import solver as sutil\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5a3962c7-590a-4628-a5c7-6ab1f9a8800e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "nlog2 = math.log(2)\n",
    "\n",
    "#Calculate doubing time from objective value\n",
    "def doubling_time(model):\n",
    "    with model as m:\n",
    "        ov = m.slim_optimize()\n",
    "        if ov < 1e-6:\n",
    "            growth = 'No growth'\n",
    "        else:  \n",
    "            growth = nlog2*60/ov\n",
    "            growth = str(round(growth, 3)) + ' minutes'\n",
    "        \n",
    "    return growth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e6c6d365-ad81-4970-8b54-9b6a36a24587",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Quicker way to read in models\n",
    "import pickle\n",
    "def read_model(fileName, obj='none'):\n",
    "    \n",
    "    fileType = fileName.split('.')[-1]\n",
    "    \n",
    "    if fileType == 'sbml' or fileType == 'xml':\n",
    "        model = cobra.io.read_sbml_model(fileName)\n",
    "    elif fileType == 'json':\n",
    "        model = cobra.io.load_json_model(fileName)\n",
    "    elif fileType == 'yaml':\n",
    "        model = cobra.io.load_yaml_model(fileName)\n",
    "    elif fileType == 'mat':\n",
    "        model = cobra.io.load_matlab_model(fileName)\n",
    "    elif fileType == 'pkl':\n",
    "        model = pickle.load(open(fileName, 'rb'))\n",
    "    else:\n",
    "        raise TypeError('Unrecognized file extension')\n",
    "    \n",
    "    if obj != 'none': model.objective = obj\n",
    "    for rxn in model.boundary: rxn.bounds = (-1000., 1000.)\n",
    "        \n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8d93ab32-fc79-4bed-875d-a4db2ed22a45",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set medias\n",
    "rpmi_req=set(['EX_apoACP_c_', 'EX_trdrd_c_','EX_cobalt2_e_','EX_cu2_e_','EX_fe3_e_','EX_mn2_e_','EX_mobd_e_','EX_zn2_e_','EX_ca2_e_','EX_no3_e_','EX_h2o_e_','EX_k_e_','EX_cl_e_','EX_mg2_e_','EX_so4_e_','EX_na1_e_','EX_pi_e_','EX_h_e_','EX_arg_L_e_','EX_asn_L_e_','EX_asp_L_e_','EX_cys_L_e_','EX_glu_L_e_','EX_gly_e_','EX_his_L_e_','EX_ile_L_e_','EX_leu_L_e_','EX_lys_L_e_','EX_met_L_e_','EX_phe_L_e_','EX_pro_L_e_','EX_ser_L_e_','EX_thr_L_e_','EX_trp_L_e_','EX_tyr_L_e_','EX_val_L_e_','EX_pnto_R_e_','EX_chol_e_','EX_inost_e_','EX_glc_D_e_','EX_gthrd_e_','EX_co2_e_'])\n",
    "mdm_req=set(['EX_apoACP_c_', 'EX_trdrd_c_','EX_co2_e_', 'EX_cobalt2_e_', 'EX_cu2_e_' , 'EX_h_e_', 'EX_h2o_e_','EX_mn2_e_','EX_mobd_e_','EX_tungs_e_', 'EX_zn2_e_','EX_na1_e_','EX_cl_e_','EX_k_e_','EX_so4_e_','EX_nh4_e_','EX_mg2_e_','EX_pi_e_','EX_ca2_e_','EX_fe3_e_','EX_no3_e_','EX_asp_L_e_','EX_glu_L_e_','EX_arg_L_e_','EX_gly_e_','EX_ser_L_e_','EX_leu_L_e_','EX_ile_L_e_','EX_val_L_e_','EX_tyr_L_e_','EX_cys_L_e_','EX_pro_L_e_','EX_trp_L_e_','EX_thr_L_e_','EX_phe_L_e_','EX_asn_L_e_','EX_gln_L_e_','EX_his_L_e_','EX_met_L_e_','EX_ala_L_e_','EX_lys_L_e_','EX_gthrd_e_','EX_thm_e_','EX_pnto_R_e_','EX_glc_D_e_','EX_hxan_e_','EX_ura_e_'])\n",
    "\n",
    "def rpmi(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "\n",
    "def complete(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=-20.\n",
    "\n",
    "def mdm(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -20.\n",
    "        elif reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "        elif reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "\n",
    "def rpmi_molarity(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -20. \n",
    "        if 'EX_glc_D_e_' in reaction.id:\n",
    "            reaction.lower_bound = -11.1012\n",
    "        if 'EX_ser_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.2855\n",
    "        if 'EX_asp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1503\n",
    "        if 'EX_asn_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.3784\n",
    "        if 'EX_glu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1359\n",
    "        if 'EX_gly_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1332\n",
    "        if 'EX_val_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1707\n",
    "        if 'EX_leu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.3812\n",
    "        if 'EX_pro_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1737\n",
    "        if 'EX_phe_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.0908\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.219\n",
    "        if 'EX_ca2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.424\n",
    "        if 'EX_no3_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.847\n",
    "        if 'EX_k_e_' in reaction.id:\n",
    "            reaction.lower_bound = -5.3655\n",
    "        if 'EX_cl_e_' in reaction.id:\n",
    "            reaction.lower_bound = -108.6993\n",
    "        if 'EX_mg2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.4056\n",
    "        if 'EX_so4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.4056\n",
    "        if 'EX_na1_e_' in reaction.id:\n",
    "            reaction.lower_bound = -137.9818\n",
    "        if 'EX_pi_e_' in reaction.id:\n",
    "            reaction.lower_bound = -6.354\n",
    "        if 'EX_arg_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.1481\n",
    "        if 'EX_cys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -10.416\n",
    "        if 'EX_his_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.0967\n",
    "        if 'EX_ile_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.3812\n",
    "        if 'EX_met_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1005\n",
    "        if 'EX_thr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1679\n",
    "        if 'EX_trp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.0245\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.219\n",
    "        if 'EX_tyr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1104\n",
    "        if 'EX_pnto_R_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.001\n",
    "        if 'EX_chol_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.0215\n",
    "        if 'EX_inost_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1943\n",
    "        if 'EX_gthrd_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.0033\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -20."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b5e8ec30-61c6-4927-8136-efa58e30a42d",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Read in  model\n",
    "import cobra\n",
    "model=read_model(cwd+'/Gc_GENRE_2022/Curate_Gc_Model/In_progress_curation/NGO_513.json')\n",
    "\n",
    "# Read in  other models to assist with annotation\n",
    "NM_model = read_model(cwd+'/Gc_GENRE_2022/Models/Nmb_iTM560.json')\n",
    "\n",
    "#Ecoli model NM model is based on - to be used for metabolite charges\n",
    "ecoli = read_model(cwd+'/Gc_GENRE_2022/Models/ecoli.json')\n",
    "\n",
    "#Universal model for adding reactions\n",
    "universalmodel= read_model(cwd+'/Gc_GENRE_2022/Models/universal_model_cobrapy.json')\n",
    "\n",
    "#Downloaded carvememodel for Neisseria gonorrhoeae FA1090 from PMID: 30192979 to compare annotations and add Gc specific reactions\n",
    "carveme_model= cobra.io.read_sbml_model(cwd+'/Gc_GENRE_2022/Models/Neisseria_gonorrhoeae_FA_1090.xml.gz')\n",
    "carvememodel=carveme_model\n",
    "cobra.io.save_json_model(carveme_model, cwd+'/Gc_GENRE_2022/Models/Neisseria_gonorrhoeae_FA_1090.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "87ee6c25-73f3-4925-934b-d21f16bec6a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "rpmi_req=set(['EX_apoACP_c_', 'EX_trdrd_c_','EX_cobalt2_e_','EX_cu2_e_','EX_fe3_e_','EX_mn2_e_','EX_mobd_e_','EX_zn2_e_','EX_ca2_e_','EX_no3_e_','EX_h2o_e_','EX_k_e_','EX_cl_e_','EX_mg2_e_','EX_so4_e_','EX_na1_e_','EX_pi_e_','EX_h_e_','EX_arg_L_e_','EX_asn_L_e_','EX_asp_L_e_','EX_cys_L_e_','EX_glu_L_e_','EX_gly_e_','EX_his_L_e_','EX_ile_L_e_','EX_leu_L_e_','EX_lys_L_e_','EX_met_L_e_','EX_phe_L_e_','EX_pro_L_e_','EX_ser_L_e_','EX_thr_L_e_','EX_trp_L_e_','EX_tyr_L_e_','EX_val_L_e_','EX_pnto_R_e_','EX_chol_e_','EX_inost_e_','EX_glc_D_e_','EX_gthrd_e_','EX_co2_e_'])\n",
    "\n",
    "def rpmi_with_lactate_only(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -0.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "            \n",
    "def rpmi(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -20.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "95818e08-3e1c-426a-8874-75101a21f092",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 55.985</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12DGR120tipp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR140tipp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR141tipp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR160tipp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12DGR161tipp</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zn2tex</th>\n",
       "      <td>0.176801</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_apoACP_c_</th>\n",
       "      <td>-2.057390</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLCTR2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.031651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GALT3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.031651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_trdrd_c_</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1413 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 55.985 at 0x1fcb3f9bc70>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def complete(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=-1000.\n",
    "            \n",
    "complete(model)\n",
    "model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b73ae5b0-4464-4515-b9aa-098e08d507f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NGO0057', 'NGO0059', 'NGO0063', 'NGO0084', 'NGO0085', 'NGO0114', 'NGO0176', 'NGO0177', 'NGO0186', 'NGO0188', 'NGO0189', 'NGO0199', 'NGO0231', 'NGO0294', 'NGO0351', 'NGO0382', 'NGO0387', 'NGO0388', 'NGO0541', 'NGO0567', 'NGO0571', 'NGO0604', 'NGO0608', 'NGO0632', 'NGO0633', 'NGO0652', 'NGO0653', 'NGO0659', 'NGO0676', 'NGO0718', 'NGO0763', 'NGO0829', 'NGO0830', 'NGO0842', 'NGO0858', 'NGO0886', 'NGO0902', 'NGO0904', 'NGO0905', 'NGO0906', 'NGO0934', 'NGO0941', 'NGO0974', 'NGO1032', 'NGO1033', 'NGO1047', 'NGO1052', 'NGO1055', 'NGO1154', 'NGO1204', 'NGO1207', 'NGO1226', 'NGO1227', 'NGO1243', 'NGO1298', 'NGO1323', 'NGO1325', 'NGO1334', 'NGO1364', 'NGO1365', 'NGO1381', 'NGO1404', 'NGO1432', 'NGO1433', 'NGO1434', 'NGO1454', 'NGO1530', 'NGO1559', 'NGO1577', 'NGO1684', 'NGO1686', 'NGO1692', 'NGO1722', 'NGO1732', 'NGO1733', 'NGO1753', 'NGO1769', 'NGO1858', 'NGO1866', 'NGO1922', 'NGO1930', 'NGO1963', 'NGO1977', 'NGO1990', 'NGO2011', 'NGO2012', 'NGO2013', 'NGO2014', 'NGO2015', 'NGO2027', 'NGO2033', 'NGO2040', 'NGO2055', 'NGO2100', 'NGO2113', 'NGO2135', 'NGO2171', 'NGO2172']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-15-1829916bf6eb>:18: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.\n",
      "  annotation_table['ProteinID'] = annotation_table['ProteinID'].str.replace('.', '_')\n"
     ]
    }
   ],
   "source": [
    "#Identify genes/reactions to add to the model from carveme\n",
    "model_genes  = []\n",
    "for gene in model.genes:\n",
    "    model_genes.append(gene.id)\n",
    "\n",
    "#Get list of genes from Carveme model for FA1090 in NGO format\n",
    "carveme_model_genes= list()\n",
    "for gene in carveme_model.genes:\n",
    "    carveme_model_genes.append(gene.id)\n",
    "carveme_model_genes\n",
    "\n",
    "#Use look_back table of annotations from https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5399841/\n",
    "annotation_table = pd.read_csv (cwd+'/Gc_GENRE_2022/Curate_Gc_Model/Annotation_resources/look_back_table.csv', usecols= ['Locus Tag ID','Protein.ID'])\n",
    "annotation_table=annotation_table.rename(columns={'Protein.ID': 'ProteinID', 'Locus Tag ID': \"NGO_ID\"})\n",
    "annotation_table\n",
    "                        \n",
    "#Match id format between kegg and model\n",
    "annotation_table['ProteinID'] = annotation_table['ProteinID'].str.replace('.', '_')\n",
    "\n",
    "annotatedgenes= annotation_table[annotation_table.ProteinID.isin(carveme_model_genes)]\n",
    "converted_carveme_model_genes= list()\n",
    "for tag in annotatedgenes['NGO_ID']:\n",
    "        converted_carveme_model_genes.append(tag)\n",
    "        \n",
    "unique_genes_carveme = set(converted_carveme_model_genes) - set(model_genes) # correct elements, but not yet in sorted order\n",
    "print(sorted(unique_genes_carveme)) # sort and print\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "012c0bba-df82-4c2f-ae0d-982919a3b0bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add NGOIDs to the carveme model as an annotation\n",
    "for gene in annotatedgenes['ProteinID']:\n",
    "    if gene in carveme_model.genes:\n",
    "        row=annotatedgenes.index[annotatedgenes['ProteinID']==gene]\n",
    "        id=annotatedgenes['NGO_ID']\n",
    "        annotation =id.loc[row].item()\n",
    "        carveme_model.genes.get_by_id(gene).annotation['NGO_ID']= annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7e4162b5-728d-4a88-9ac6-31ab5d9d21af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>carvemegenes</th>\n",
       "      <th>NGO_IDs</th>\n",
       "      <th>carvemereactions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YP_208072_1</td>\n",
       "      <td>{'NGO_ID': 'NGO0974'}</td>\n",
       "      <td>[2AGPGAT160: 2agpg160_c + atp_c + hdca_c --&gt; a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YP_208834_1</td>\n",
       "      <td>{'NGO_ID': 'NGO1804'}</td>\n",
       "      <td>[3HAD40: 3haACP_c --&gt; but2eACP_c + h2o_c, 3HAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YP_209169_1</td>\n",
       "      <td>{'NGO_ID': 'NGO2163'}</td>\n",
       "      <td>[3OAR121: 3ocddec5eACP_c + h_c + nadph_c --&gt; 3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YP_208796_1</td>\n",
       "      <td>{'NGO_ID': 'NGO1763'}</td>\n",
       "      <td>[KAS15: accoa_c + h_c + malACP_c --&gt; actACP_c ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YP_209174_1</td>\n",
       "      <td>{'NGO_ID': 'NGO2168'}</td>\n",
       "      <td>[3OAS180: h_c + malACP_c + palmACP_c --&gt; 3ooct...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>YP_207518_1</td>\n",
       "      <td>{'NGO_ID': 'NGO0362'}</td>\n",
       "      <td>[UPPDC1: 4.0 h_c + uppg3_c --&gt; 4.0 co2_c + cpp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>YP_207509_1</td>\n",
       "      <td>{'NGO_ID': 'NGO0353'}</td>\n",
       "      <td>[UPPRT: prpp_c + ura_c --&gt; ppi_c + ump_c]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>507</th>\n",
       "      <td>YP_207732_1</td>\n",
       "      <td>{'NGO_ID': 'NGO0589'}</td>\n",
       "      <td>[URAt: ura_e &lt;=&gt; ura_c, URAt2: h_e + ura_e --&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>YP_208502_1</td>\n",
       "      <td>{'NGO_ID': 'NGO1448'}</td>\n",
       "      <td>[USHD: h2o_c + u23ga_c --&gt; 2.0 h_c + lipidX_c ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>509</th>\n",
       "      <td>YP_207255_1</td>\n",
       "      <td>{'NGO_ID': 'NGO0082'}</td>\n",
       "      <td>[VPAMTr: 3mob_c + ala__L_c --&gt; pyr_c + val__L_c]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>510 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    carvemegenes                NGO_IDs  \\\n",
       "0    YP_208072_1  {'NGO_ID': 'NGO0974'}   \n",
       "1    YP_208834_1  {'NGO_ID': 'NGO1804'}   \n",
       "2    YP_209169_1  {'NGO_ID': 'NGO2163'}   \n",
       "3    YP_208796_1  {'NGO_ID': 'NGO1763'}   \n",
       "4    YP_209174_1  {'NGO_ID': 'NGO2168'}   \n",
       "..           ...                    ...   \n",
       "505  YP_207518_1  {'NGO_ID': 'NGO0362'}   \n",
       "506  YP_207509_1  {'NGO_ID': 'NGO0353'}   \n",
       "507  YP_207732_1  {'NGO_ID': 'NGO0589'}   \n",
       "508  YP_208502_1  {'NGO_ID': 'NGO1448'}   \n",
       "509  YP_207255_1  {'NGO_ID': 'NGO0082'}   \n",
       "\n",
       "                                      carvemereactions  \n",
       "0    [2AGPGAT160: 2agpg160_c + atp_c + hdca_c --> a...  \n",
       "1    [3HAD40: 3haACP_c --> but2eACP_c + h2o_c, 3HAD...  \n",
       "2    [3OAR121: 3ocddec5eACP_c + h_c + nadph_c --> 3...  \n",
       "3    [KAS15: accoa_c + h_c + malACP_c --> actACP_c ...  \n",
       "4    [3OAS180: h_c + malACP_c + palmACP_c --> 3ooct...  \n",
       "..                                                 ...  \n",
       "505  [UPPDC1: 4.0 h_c + uppg3_c --> 4.0 co2_c + cpp...  \n",
       "506          [UPPRT: prpp_c + ura_c --> ppi_c + ump_c]  \n",
       "507  [URAt: ura_e <=> ura_c, URAt2: h_e + ura_e -->...  \n",
       "508  [USHD: h2o_c + u23ga_c --> 2.0 h_c + lipidX_c ...  \n",
       "509   [VPAMTr: 3mob_c + ala__L_c --> pyr_c + val__L_c]  \n",
       "\n",
       "[510 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "carvemegenes = []\n",
    "carvemengoids=[]\n",
    "carvemereactions = []\n",
    "\n",
    "\n",
    "for gene in carveme_model.genes:\n",
    "    carvemegenes.append(gene.id)\n",
    "    carvemengoids.append(gene.annotation)\n",
    "    sets=list(gene.reactions)\n",
    "    rxnlist=[]\n",
    "    for x in sets:\n",
    "        rxnlist.append(x)\n",
    "    carvemereactions.append(rxnlist)\n",
    "\n",
    "dict = {'carvemegenes': carvemegenes, 'NGO_IDs':carvemengoids,'carvemereactions':carvemereactions}\n",
    "carvemegenes_reactions_DF = pandas.DataFrame(dict)\n",
    "carvemegenes_reactions_DF.to_csv(cwd+'/Gc_GENRE_2022/Curate_Gc_Model/Annotation_resources/carveme_genes_reactions.csv')\n",
    "carvemegenes_reactions_DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0a01f816-49a9-4542-b975-b996170ecd39",
   "metadata": {},
   "outputs": [],
   "source": [
    "carveme_model.metabolites.ala__L_c.id ='ala_L_c'\n",
    "carveme_model.metabolites.glu__L_c.id ='glu_L_c'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "06da6b2d-9302-4029-b213-a8bc941b90af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>ALATA_L</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>L-alanine transaminase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcbe948d90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>akg_c + ala_L_c <=> glu_L_c + pyr_c</p>\n",
       "                    <p style='text-align:right'>2-Oxoglutarate + L-Alanine <=> L-Glutamate + Pyruvate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1047</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction ALATA_L at 0x1fcbe948d90>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALATA_L =carvememodel.reactions.get_by_id('ALATA_L')\n",
    "carvememodel.reactions.get_by_id('ALATA_L').gene_reaction_rule='NGO1047'\n",
    "model.add_reactions([ALATA_L])\n",
    "model.reactions.get_by_id('ALATA_L')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "943cac6e-5eff-4760-8764-bc9e0cc36eb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>AP4AH</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Ap4A hydrolase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcb2e027f0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>ap4a_c + h2o_c --> 2.0 adp_c + 2.0 h_c</p>\n",
       "                    <p style='text-align:right'>P1,P4-Bis(5'-adenosyl) tetraphosphate + H2O --> 2.0 ADP + 2.0 H</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0231</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction AP4AH at 0x1fcb2e027f0>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AP4AH =carvememodel.reactions.get_by_id('AP4AH')\n",
    "carvememodel.reactions.get_by_id('AP4AH').gene_reaction_rule='NGO0231'\n",
    "model.add_reactions([AP4AH])\n",
    "model.reactions.get_by_id('AP4AH')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d02078a8-ba82-4ad6-82f5-a61410932f0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NGO0872\n",
      "NGO0872 and NGO1204\n"
     ]
    }
   ],
   "source": [
    "#ANS already exists in annotated Gc model, the model GPR should require NGO9872 and NGO1204\n",
    "print (model.reactions.get_by_id('ANS').gene_reaction_rule)\n",
    "model.reactions.get_by_id('ANS').gene_reaction_rule='NGO0872 and NGO1204'\n",
    "print (model.reactions.get_by_id('ANS').gene_reaction_rule)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6e2500b5-f019-47f2-ac5d-3aa296f1b0c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>BTDD_RR</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R R butanediol dehydrogenase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcb4899d60</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>btd_RR_c + nad_c <=> actn__R_c + h_c + nadh_c</p>\n",
       "                    <p style='text-align:right'>R R 2 3 Butanediol C4H10O2 + Nicotinamide-adenine-dinucleotide <=> R Acetoin C4H8O2 + H + Nicotinamide-adenine-dinucleotide--reduced</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0186</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction BTDD_RR at 0x1fcb4899d60>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BTDD_RR =carvememodel.reactions.get_by_id('BTDD_RR')\n",
    "carvememodel.reactions.get_by_id('BTDD_RR').gene_reaction_rule='NGO0186'\n",
    "model.add_reactions([BTDD_RR])\n",
    "model.reactions.get_by_id('BTDD_RR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "483952ee-5b3d-49b5-9de5-7866e9eae446",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>CYTOM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Cytosine 5 methyltransferase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcb4bdac70</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>amet_c + csn_c <=> 5mcsn_c + ahcys_c + h_c</p>\n",
       "                    <p style='text-align:right'>S-Adenosyl-L-methionine + Cytosine <=> 5-Methylcytosine + S-Adenosyl-L-homocysteine + H</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0676</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction CYTOM at 0x1fcb4bdac70>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CYTOM =carvememodel.reactions.get_by_id('CYTOM')\n",
    "carvememodel.reactions.get_by_id('CYTOM').gene_reaction_rule='NGO0676'\n",
    "model.add_reactions([CYTOM])\n",
    "model.reactions.get_by_id('CYTOM')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "83096db6-9909-49cf-8766-21977986debf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DNTPPA</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_DNTPPA</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac579af0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>ahdt_c + h2o_c --> dhpmp_c + h_c + ppi_c</p>\n",
       "                    <p style='text-align:right'>2-Amino-4-hydroxy-6-erythro-1-2-3-trihydroxypropyl-dihydropteridine-triphosphate + H2O --> Dihydroneopterin-monophosphate + H + Diphosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0224</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DNTPPA at 0x1fcac579af0>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#The gene currently indicated for this reaction is NGO1506 - however KEGG annotates this gene as a 8-oxo-dGTP diphosphatase (MutT), KEGG reaction \t R09832\n",
    "#CARVEME indicates this reactions is produced by NGO1334. However this gene is annotated as a putative RNA pyrophosphohydrolase. The function isn't clear. \n",
    "# Replacing with the KEGG annotated gene for this reaction NGO0224\n",
    "\n",
    "DNTPPA =carvememodel.reactions.get_by_id('DNTPPA')\n",
    "model.reactions.get_by_id('DNTPPA').gene_reaction_rule='NGO0224'\n",
    "model.reactions.get_by_id('DNTPPA')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "20388902-2684-4781-8da8-4585b660567a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DNMPPA</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_DNMPPA</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac579580</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>dhpmp_c + h2o_c --> dhnpt_c + pi_c</p>\n",
       "                    <p style='text-align:right'>Dihydroneopterin-monophosphate + H2O --> Dihydroneopterin + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1334</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DNMPPA at 0x1fcac579580>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#DNMPPA is not annotated in KEGG, but is predicted to use an enzyme with E.C. number 3.6.1-, and is listed as an orphan gene in the current model. \n",
    "#I am tentatively de-orphaning the gene by adding the annotation from CARVEME for NGO1334\n",
    "\n",
    "DNMPPA =carvememodel.reactions.get_by_id('DNMPPA')\n",
    "model.reactions.get_by_id('DNMPPA').gene_reaction_rule='NGO1334'\n",
    "model.reactions.get_by_id('DNMPPA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d0e01a89-6219-44de-8e05-3727c50aace3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>FADRx</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>FAD reductase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcb2e634f0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>fad_c + h_c + nadh_c --> fadh2_c + nad_c</p>\n",
       "                    <p style='text-align:right'>Flavin-adenine-dinucleotide-oxidized + H + Nicotinamide-adenine-dinucleotide--reduced --> Flavin-adenine-dinucleotide-reduced + Nicotinamide-adenine-dinucleotide</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0059</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction FADRx at 0x1fcb2e634f0>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME indicates FADRx uses NGO0059 or NGO0659. However NGO0659 has different annotations in KEGG that do not match this function. Removed NGO0659. \n",
    "FADRx =carvememodel.reactions.get_by_id('FADRx')\n",
    "carvememodel.reactions.get_by_id('FADRx').gene_reaction_rule='NGO0059'\n",
    "model.add_reactions([FADRx])\n",
    "model.reactions.get_by_id('FADRx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fdd8de44-f8c7-47c3-9475-6dce142a1bef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>HPA3MO</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>4 hydroxyphenylacetate 3 monooxygenase nadh</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcb30f5100</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>4hphac_c + h_c + nadh_c + o2_c --> 34dhpha_c + h2o_c + nad_c</p>\n",
       "                    <p style='text-align:right'>4-Hydroxyphenylacetate + H + Nicotinamide-adenine-dinucleotide--reduced + O2 --> 3-4-Dihydroxyphenylacetate + H2O + Nicotinamide-adenine-dinucleotide</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0059</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction HPA3MO at 0x1fcb30f5100>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME and KEGG indicate presence of this reaction\n",
    "HPA3MO =carvememodel.reactions.get_by_id('HPA3MO')\n",
    "carvememodel.reactions.get_by_id('HPA3MO').gene_reaction_rule='NGO0059'\n",
    "model.add_reactions([HPA3MO])\n",
    "model.reactions.get_by_id('HPA3MO')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1bcac71b-08c9-45f6-8b94-461db3681cd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GLYCL</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_GLYCL</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac671ca0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>gly_c + nad_c + thf_c <=> co2_c + mlthf_c + nadh_c + nh4_c</p>\n",
       "                    <p style='text-align:right'>Glycine + Nicotinamide-adenine-dinucleotide + 5-6-7-8-Tetrahydrofolate <=> CO2 + 5-10-Methylenetetrahydrofolate + Nicotinamide-adenine-dinucleotide--reduced + Ammonium</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0562 and NGO0915 and NGO0925 and NGO1325 and NGO1404 and NGO1406</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-999999.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GLYCL at 0x1fcac671ca0>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Model has NGO1406 annotated for GLYCL. CARVEME has NGO1325 annotated. \n",
    "#The Glycine cleavage system is a functional system composed of 4 different components *H, P, T, and L. https://www.biocyc.org/GCF_000006845/NEW-IMAGE?type=PATHWAY&object=GLYCLEAV-PWY\n",
    "#All 4 components should be required for function of this system. As such, I am listing ALL 6 genes, indicated in KEGG, for this reaction. \n",
    "\n",
    "GLYCL =carvememodel.reactions.get_by_id('GLYCL')\n",
    "model.reactions.get_by_id('GLYCL').gene_reaction_rule='NGO0562 and NGO0915 and NGO0925 and NGO1325 and NGO1404 and NGO1406'\n",
    "model.reactions.get_by_id('GLYCL')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "638d74b9-48cc-4650-b51b-5493e4b9f79b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GLYOX</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Hydroxyacylglutathione hydrolase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcb3073d90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h2o_c + lgt__S_c --> gthrd_c + h_c + lac__D_c</p>\n",
       "                    <p style='text-align:right'>H2O + (R)-S-Lactoylglutathione --> Reduced glutathione + H+ + D-Lactate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YP_208259_1 or YP_209053_1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GLYOX at 0x1fcb3073d90>"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME indicated NGO2033\n",
    "#KEGG indicates NGO1187 AND NGO2033 participate in this reaction https://www.genome.jp/entry/ngo:NGO_1187+ngo:NGO_2033\n",
    "\n",
    "\n",
    "GLYOX =carvememodel.reactions.get_by_id('GLYOX')\n",
    "carvememodel.reactions.get_by_id('GLYOX')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "bd931187-eaa4-45bd-af9d-349b52873405",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GLYOX</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_GLYOX</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac67d460</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h2o_c + lgt_S_c <=> gthrd_c + h_c + lac_D_c</p>\n",
       "                    <p style='text-align:right'>H2O + R-S-Lactoylglutathione <=> Reduced-glutathione + H + D-Lactate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1187</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-999999.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GLYOX at 0x1fcac67d460>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('HPA3MO').gene_reaction_rule='NGO1187 and NGO2033'\n",
    "model.reactions.get_by_id('GLYOX')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "84e0b48f-e4ca-44d7-b4e3-a2702f54378c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>GTPCI</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_GTPCI</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac68a910</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>gtp_c + h2o_c --> ahdt_c + for_c + h_c</p>\n",
       "                    <p style='text-align:right'>GTP + H2O --> 2-Amino-4-hydroxy-6-erythro-1-2-3-trihydroxypropyl-dihydropteridine-triphosphate + Formate + H</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0387</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction GTPCI at 0x1fcac68a910>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME indicates NGO0387 as performing this reaction, and so does KEGG. https://www.genome.jp/entry/ngo:NGO_0387\n",
    "GTPCI =carvememodel.reactions.get_by_id('GTPCI')\n",
    "model.reactions.get_by_id('GTPCI').gene_reaction_rule='NGO0387'\n",
    "model.reactions.get_by_id('GTPCI')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a7014173-e6d6-4228-8d8f-fc8afe1e6b82",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring reaction 'G3PAT120' since it already exists.\n",
      "Ignoring reaction 'G3PAT180' since it already exists.\n",
      "Ignoring reaction 'G3PAT181' since it already exists.\n",
      "Ignoring reaction 'G3PAT140' since it already exists.\n",
      "Ignoring reaction 'G3PAT160' since it already exists.\n"
     ]
    }
   ],
   "source": [
    "#Adding reactions according to PMCID: PMC4697154/PMID: 26567338 which are conducted by plsX (Ngo2151)\n",
    "x = ['ACPPAT161' , 'G3PAT120' , 'ACPPAT181' , 'G3PAT180' , 'ACPPAT120' , 'G3PAT181' , 'ACPPAT141' , 'G3PAT140' , 'ACPPAT160' , 'G3PAT161' , 'G3PAT160' , 'ACPPAT140' , 'ACPPAT180']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO2171'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "402507da-306c-40da-bbf7-b3b31ce3c724",
   "metadata": {},
   "outputs": [],
   "source": [
    "#providing gene for orphan reactions according to PMCID: PMC4697154/PMID: 26567338 which are conducted by plsX (Ngo2151)\n",
    "x = ['G3PAT120' , 'G3PAT180' , 'G3PAT181' , 'G3PAT140' , 'G3PAT161' , 'G3PAT160' ]\n",
    "for r in x:\n",
    "    reaction =model.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO2171'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c0a63216-dd74-4d74-911a-7177bb5e88a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Adding reactions according to PMCID: PMC4697154/PMID: 26567338 which are conducted by plsY (Ngo2151)\n",
    "x = ['APG3PAT181' , 'APG3PAT160' , 'APG3PAT180' , 'APG3PAT120' , 'APG3PAT141' , 'APG3PAT161' , 'APG3PAT140']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0858'\n",
    "    model.add_reactions([reaction])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "458492ae-c934-4ef1-aac6-737e5b13a6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Adding peptidase reaction\n",
    "x = ['AMPTASEPG']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO1227'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "03324b10-47cb-4ed7-950f-100de086fe6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cardiolipin has been in cultures of Gc PMCID: PMC235958/ PMID: 810478\n",
    "x = ['CLPNS181pp' , 'CLPNS141pp' , 'CLPNS140pp' , 'CLPNS160pp' , 'CLPNS120pp' , 'CLPNS161pp' , 'CLPNS180pp']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0902'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b9cc3884-c9d9-4a03-9778-80afac12e73b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CYPOX: h2o2_c + 2.0 h_c --> 2.0 h2o_c\n",
      "{}\n"
     ]
    }
   ],
   "source": [
    "#reduction of H2O2 to water facilitated by cytochrome c peroxidase, with Laz as an electron donor. Laz is encoded by NGO0994\n",
    "reaction = cobra.Reaction('CYPOX')\n",
    "reaction.name = 'Cytochrome C Peroxidase'\n",
    "reaction.lower_bound = 0.  \n",
    "reaction.upper_bound = 1000.  \n",
    "reaction.gene_reaction_rule='NGO1769 and NGO0994'\n",
    "\n",
    "\n",
    "\n",
    "h_c=model.metabolites.get_by_id('h_c')\n",
    "h2o2= model.metabolites.get_by_id('h2o2_c')\n",
    "h2o_c=model.metabolites.get_by_id('h2o_c')\n",
    "\n",
    "reaction.add_metabolites({\n",
    "    h2o2: -1.0,\n",
    "    h_c: -2.0,\n",
    "    h2o_c: 2.0,\n",
    "})\n",
    "\n",
    "model.add_reactions([reaction])\n",
    "print(model.reactions.CYPOX)\n",
    "print(model.reactions.CYPOX.check_mass_balance())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "06484aa7-b0d7-4ee8-bed5-b274738022a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ['FLVR', 'FLVRx']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0388'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "79a86004-0a87-4d63-966b-5c0a9ab846fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This reaction is indicated as a bidirectional reaction in Bigg. However Kegg indicates it is \n",
    "#unidirection with NGO1977 performing the forward reaction and NGO1406 the reverse reaction (https://www.kegg.jp/pathway/ngo00670+NGO_1977)\n",
    "x = ['FOMETRi']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO1977 and NGO1406'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9bb3facf-bdeb-4dc8-b336-9fa37f7d929b",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.GLYCLTDx.gene_reaction_rule='NGO2043 or NGO1243'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "083509b7-a646-4b0f-81eb-f255dddff745",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ['GRXR']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0114 or NGO1381 or NGO0351'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "7d18ac12-7ebb-4485-9de3-ec263fdc9ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Carveme did not correctly annotate NGO1684, (its Not GTPCI) but it does perform a reaction just downstream. \n",
    "#Involved in folate biosynthesis. Corrected and added the full pathway.\n",
    "#To include NGO0387 (GTPCI), NGO0131, NGO00720, NGO0129, NGO1684 (https://www.genome.jp/pathway/ngo00790+NGO_1684)\n",
    "x = ['PTHPS']\n",
    "for r in x:\n",
    "    reaction =universalmodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0131'\n",
    "    model.add_reactions([reaction])\n",
    "x = ['CPH4S']\n",
    "for r in x:\n",
    "    reaction =universalmodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0131'\n",
    "    model.add_reactions([reaction])\n",
    "x = ['CDGS']\n",
    "for r in x:\n",
    "    reaction =universalmodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO00720'\n",
    "    model.add_reactions([reaction])\n",
    "x = ['CDGR']\n",
    "for r in x:\n",
    "    reaction =universalmodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO1684'\n",
    "    model.add_reactions([reaction])\n",
    "x = ['CCGS']\n",
    "for r in x:\n",
    "    reaction =universalmodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0129'\n",
    "    model.add_reactions([reaction])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1644ac28-4779-4daa-8b9b-868787c19c6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.Kt2pp.gene_reaction_rule='NGO1774 or NGO1154'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "7ffa49c7-e8ac-4224-9e23-dafe2e1f740c",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ['MLDEP2pp']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0571 or NGO1686'\n",
    "    model.add_reactions([reaction])\n",
    "#ltgD (0626) and ltgC (2048) ltgB (1033) and ltgE(0608) are all murein lytic transcglycosylases. PMCID: PMC3412582/PMID: 22432703\n",
    "#Adding all of them as potential genes \n",
    "x = ['MLTGY3pp']\n",
    "for r in x:\n",
    "    reaction =carvememodel.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0608 or NGO1033 or NGO0626 or NGO2048'\n",
    "    model.add_reactions([reaction])\n",
    "\n",
    "x = ['MLTGY1pp' ,'MLTGY4pp' , 'MLTGY2pp']\n",
    "for r in x:\n",
    "    reaction =model.reactions.get_by_id(r)\n",
    "    reaction.gene_reaction_rule='NGO0608 or NGO1033 or NGO0626 or NGO2048'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "eb4794c8-f72b-4b97-b740-97ea4a768dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Deorphan PPPGO\n",
    "model.reactions.PPPGO.gene_reaction_rule= 'NGO1692'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "d5c19908-4d75-4144-a64b-32c9d21004bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add additional thioredoxin proteins\n",
    "model.reactions.THIORDXi.gene_reaction_rule=  'NGO0328 or NGO1923 or NGO0652 or NGO0057'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "747cb964-d0c2-4d40-8367-9ce5b2904d28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>PGLYCP</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_PGLYCP</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac7bffd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>2pglyc_c + h2o_c --> glyclt_c + pi_c</p>\n",
       "                    <p style='text-align:right'>2-Phosphoglycolate + H2O --> Glycolate + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0073 or NGO1052</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction PGLYCP at 0x1fcac7bffd0>"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME indicates NGO1052\n",
    "#KEGG indicates NGO0073 and NGO1052\n",
    "#NGO0203 has a different EC number - EC 3.1.3.105 . I am replacing with the KEGG/CARVEME gene annotations\n",
    "\n",
    "PGLYCP =carvememodel.reactions.get_by_id('PGLYCP')\n",
    "\n",
    "model.reactions.get_by_id('PGLYCP').gene_reaction_rule='NGO0073 or NGO1052'\n",
    "model.reactions.get_by_id('PGLYCP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "73d8c37a-6c45-4c63-a77e-693c4f6409ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>SPMS</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_SPMS</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac877c40</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>ametam_c + ptrc_c --> 5mta_c + h_c + spmd_c</p>\n",
       "                    <p style='text-align:right'>S-Adenosylmethioninamine + Putrescine --> 5-Methylthioadenosine + H + Spermidine</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0435 or NGO1752</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction SPMS at 0x1fcac877c40>"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CARVEME indicates NGO0435 or NGO1752. Model has mislabeled NGO number with an underscore. Replacing.\n",
    "SPMS =carvememodel.reactions.get_by_id('SPMS')\n",
    "model.reactions.get_by_id('SPMS').gene_reaction_rule='NGO0435 or NGO1752'\n",
    "model.reactions.get_by_id('SPMS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "bda925b0-d6e5-423d-9534-51a4c3f08082",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Model contains ethanol production - however Gc does not produce ethanol. Removing this reaction.\n",
    "ALCD2x =model.reactions.get_by_id('ALCD2x')\n",
    "model.remove_reactions([ALCD2x])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0a9834e8-fc0f-4910-9c92-dd01e8ad9e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Similarly, the mening model had added acetylaldehyde production, without genetic evidence, to support the production of ethanol observed in mening. \n",
    "#Because Gc does not produce ethanol or acetylaldehyde, removing these associated reaction.\n",
    "ALDD2y =model.reactions.get_by_id('ALDD2y')\n",
    "model.remove_reactions([ALDD2y])\n",
    "\n",
    "ACALDtpp =model.reactions.get_by_id('ACALDtpp')\n",
    "model.remove_reactions([ACALDtpp])\n",
    "\n",
    "EX_acald_e_ =model.reactions.get_by_id('EX_acald_e_')\n",
    "model.remove_reactions([EX_acald_e_])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "6e1d2180-1a74-434f-b94a-48b2f5425ff5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.6264521997332229</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>apoACP_c</td>\n",
       "      <td>EX_apoACP_c_</td>\n",
       "      <td>0.02302</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.2092</td>\n",
       "      <td>6</td>\n",
       "      <td>0.33%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.5166</td>\n",
       "      <td>4</td>\n",
       "      <td>0.54%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>21.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.002968</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.002968</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>15.79%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.009183</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc_D_e</td>\n",
       "      <td>EX_glc_D_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>31.58%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.4448</td>\n",
       "      <td>5</td>\n",
       "      <td>0.59%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>5.808</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.06335</td>\n",
       "      <td>6</td>\n",
       "      <td>0.10%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.1113</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.2961</td>\n",
       "      <td>6</td>\n",
       "      <td>0.47%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.2199</td>\n",
       "      <td>6</td>\n",
       "      <td>0.35%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.004946</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.1247</td>\n",
       "      <td>9</td>\n",
       "      <td>0.30%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.4868</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.1544</td>\n",
       "      <td>5</td>\n",
       "      <td>0.20%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>15.79%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>11.56</td>\n",
       "      <td>4</td>\n",
       "      <td>12.17%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.04319</td>\n",
       "      <td>11</td>\n",
       "      <td>0.13%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.08955</td>\n",
       "      <td>9</td>\n",
       "      <td>0.21%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.2986</td>\n",
       "      <td>5</td>\n",
       "      <td>0.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.001978</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-0.0001397</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0002794</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.6265</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0002794</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.02302</td>\n",
       "      <td>10</td>\n",
       "      <td>0.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.004847</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-55.03</td>\n",
       "      <td>2</td>\n",
       "      <td>31.06%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-77.55</td>\n",
       "      <td>1</td>\n",
       "      <td>21.88%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0004191</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-25.58</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>-11.2</td>\n",
       "      <td>6</td>\n",
       "      <td>18.95%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>-19.87</td>\n",
       "      <td>5</td>\n",
       "      <td>28.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-37.75</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x1fcac6c8cd0>"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpmi(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "5d62aae3-d34a-44d4-946e-f4e43ba7d53b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CARVEME revealed that NGO0904-NGO0906 is not present in the Gc model - which was recently discovered as an Gc\n",
    "#lactate dehydrogenase. Additionally other aspects of lactate metabolism are corrected here. \n",
    "\n",
    "#PMCID: PMC4215069\n",
    "#PMCID: PMC7530546"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "00fe0604-2c52-4c32-83f1-9b6f7998984d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>lac_D_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>D-Lactate</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcabf36910</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C3H5O3</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>cytosol</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 3 reaction(s)</strong></td><td>\n",
       "                    LDH_D, GLYOX, DLACt2pp</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite lac_D_c at 0x1fcabf36910>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.get_by_id('lac_D_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "3be28ec5-348e-4d68-be89-5550a910e0e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "carvememodel.metabolites.get_by_id('lac__D_c').id='lac_D_c'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2df2ad40-4f17-46b8-b815-3588f13b34a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>L_LACD4</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_L_LACD4</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac6d4bb0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>lac_L_c + nad_c --> h_c + nadh_c + pyr_c</p>\n",
       "                    <p style='text-align:right'>L-Lactate + Nicotinamide-adenine-dinucleotide --> H + Nicotinamide-adenine-dinucleotide--reduced + Pyruvate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0639 or Orphan</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction L_LACD4 at 0x1fcac6d4bb0>"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('L_LACD4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "a39b6955-7222-47a8-b76c-99d75d4c7902",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.40026472029671456</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>apoACP_c</td>\n",
       "      <td>EX_apoACP_c_</td>\n",
       "      <td>0.01471</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.1337</td>\n",
       "      <td>6</td>\n",
       "      <td>1.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.1007</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>0.2687</td>\n",
       "      <td>4</td>\n",
       "      <td>1.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.001896</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.001896</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.03932</td>\n",
       "      <td>3</td>\n",
       "      <td>0.15%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.005867</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.2842</td>\n",
       "      <td>5</td>\n",
       "      <td>1.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>3.826</td>\n",
       "      <td>2</td>\n",
       "      <td>9.80%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>4.802</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.04048</td>\n",
       "      <td>6</td>\n",
       "      <td>0.31%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.1168</td>\n",
       "      <td>6</td>\n",
       "      <td>0.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.07109</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.1892</td>\n",
       "      <td>6</td>\n",
       "      <td>1.45%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.1405</td>\n",
       "      <td>6</td>\n",
       "      <td>1.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.04104</td>\n",
       "      <td>5</td>\n",
       "      <td>0.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.00316</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.07966</td>\n",
       "      <td>9</td>\n",
       "      <td>0.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.311</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.09865</td>\n",
       "      <td>5</td>\n",
       "      <td>0.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>76.88%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.00158</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.1157</td>\n",
       "      <td>4</td>\n",
       "      <td>0.59%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.0276</td>\n",
       "      <td>11</td>\n",
       "      <td>0.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.05722</td>\n",
       "      <td>9</td>\n",
       "      <td>0.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.1908</td>\n",
       "      <td>5</td>\n",
       "      <td>1.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-8.926E-05</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0001785</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.4003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0001785</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.01471</td>\n",
       "      <td>10</td>\n",
       "      <td>0.24%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.003097</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-19.64</td>\n",
       "      <td>2</td>\n",
       "      <td>63.62%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-22.3</td>\n",
       "      <td>1</td>\n",
       "      <td>36.13%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0002678</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-3.757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-22.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x1fcac6c8d60>"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#NGO0639 is a quinone depdendent L-lactate dehydrogenase. There are no known NAD/NADH dependent L-lactate dehydrogenases in Gc.\n",
    "LACD4 =model.reactions.get_by_id('L_LACD4')\n",
    "model.remove_reactions([LACD4])\n",
    "rpmi_with_lactate_only(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "1c8057d3-470c-407e-b301-14400dbb76b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>LDH_D</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_LDH_D</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac6d4fd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>lac_D_c + nad_c <=> h_c + nadh_c + pyr_c</p>\n",
       "                    <p style='text-align:right'>D-Lactate + Nicotinamide-adenine-dinucleotide <=> H + Nicotinamide-adenine-dinucleotide--reduced + Pyruvate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1336</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction LDH_D at 0x1fcac6d4fd0>"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#The reaction LDH_D is a the NAD dependent conversion of D_lac to pyruvate. It is completed by ldhA (NGO1336). \n",
    "#This reaction should be bidirectional. \n",
    "\n",
    "model.reactions.get_by_id('LDH_D')\n",
    "model.reactions.get_by_id('LDH_D').gene_reaction_rule='NGO1336'\n",
    "model.reactions.get_by_id('LDH_D').lower_bound =-1000.\n",
    "model.reactions.get_by_id('LDH_D')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1c11dc22-14af-4e29-b9d1-1ee5f1578ea2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.40026472029670784</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>apoACP_c</td>\n",
       "      <td>EX_apoACP_c_</td>\n",
       "      <td>0.01471</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.1337</td>\n",
       "      <td>6</td>\n",
       "      <td>1.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.1007</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>0.2687</td>\n",
       "      <td>4</td>\n",
       "      <td>1.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.001896</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.001896</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.03932</td>\n",
       "      <td>3</td>\n",
       "      <td>0.15%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.005867</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.2842</td>\n",
       "      <td>5</td>\n",
       "      <td>1.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>3.826</td>\n",
       "      <td>2</td>\n",
       "      <td>9.80%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>4.802</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.04048</td>\n",
       "      <td>6</td>\n",
       "      <td>0.31%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.1168</td>\n",
       "      <td>6</td>\n",
       "      <td>0.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.07109</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.1892</td>\n",
       "      <td>6</td>\n",
       "      <td>1.45%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.1405</td>\n",
       "      <td>6</td>\n",
       "      <td>1.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.04104</td>\n",
       "      <td>5</td>\n",
       "      <td>0.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.00316</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.07966</td>\n",
       "      <td>9</td>\n",
       "      <td>0.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.311</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.09865</td>\n",
       "      <td>5</td>\n",
       "      <td>0.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>76.88%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.00158</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.1157</td>\n",
       "      <td>4</td>\n",
       "      <td>0.59%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.0276</td>\n",
       "      <td>11</td>\n",
       "      <td>0.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.05722</td>\n",
       "      <td>9</td>\n",
       "      <td>0.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.1908</td>\n",
       "      <td>5</td>\n",
       "      <td>1.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-8.926E-05</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0001785</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.4003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0001785</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.01471</td>\n",
       "      <td>10</td>\n",
       "      <td>0.24%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.003097</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-19.64</td>\n",
       "      <td>2</td>\n",
       "      <td>63.62%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-22.3</td>\n",
       "      <td>1</td>\n",
       "      <td>36.13%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0002678</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-3.757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-22.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x1fcb37cbfa0>"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpmi_with_lactate_only(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "1962f8c1-cce6-4d44-b2c0-2d501dfb6853",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>LDH_D2</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>D-lactate dehydrogenase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcb3583df0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>lac_D_c + q8_c --> pyr_c + q8h2_c</p>\n",
       "                    <p style='text-align:right'>D-Lactate + Ubiquinone-8 --> Pyruvate + Ubiquinol-8</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0890</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction LDH_D2 at 0x1fcb3583df0>"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#The reaction LDH_D2 is competed by ldhD (NGO0890), a Quinone dependent D-lactate dehydrogenase.\n",
    "LDH_D2=carvememodel.reactions.get_by_id('LDH_D2')\n",
    "carvememodel.reactions.get_by_id('LDH_D2').gene_reaction_rule='NGO0890'\n",
    "model.add_reactions([LDH_D2])\n",
    "model.reactions.get_by_id('LDH_D2')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "71d67d1e-e597-4eb9-aea5-8eea1ff17639",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.4002647202967156</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>apoACP_c</td>\n",
       "      <td>EX_apoACP_c_</td>\n",
       "      <td>0.01471</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.1337</td>\n",
       "      <td>6</td>\n",
       "      <td>1.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.1007</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>0.2687</td>\n",
       "      <td>4</td>\n",
       "      <td>1.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.001896</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.001896</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.03932</td>\n",
       "      <td>3</td>\n",
       "      <td>0.15%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.005867</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.2842</td>\n",
       "      <td>5</td>\n",
       "      <td>1.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>3.826</td>\n",
       "      <td>2</td>\n",
       "      <td>9.80%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>4.802</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.04048</td>\n",
       "      <td>6</td>\n",
       "      <td>0.31%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.1168</td>\n",
       "      <td>6</td>\n",
       "      <td>0.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.07109</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.1892</td>\n",
       "      <td>6</td>\n",
       "      <td>1.45%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.1405</td>\n",
       "      <td>6</td>\n",
       "      <td>1.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.04104</td>\n",
       "      <td>5</td>\n",
       "      <td>0.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.00316</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.07966</td>\n",
       "      <td>9</td>\n",
       "      <td>0.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.311</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.09865</td>\n",
       "      <td>5</td>\n",
       "      <td>0.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>76.88%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.00158</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.1157</td>\n",
       "      <td>4</td>\n",
       "      <td>0.59%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.0276</td>\n",
       "      <td>11</td>\n",
       "      <td>0.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.05722</td>\n",
       "      <td>9</td>\n",
       "      <td>0.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.1908</td>\n",
       "      <td>5</td>\n",
       "      <td>1.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-8.926E-05</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0001785</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.4003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0001785</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.01471</td>\n",
       "      <td>10</td>\n",
       "      <td>0.24%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.003097</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-19.64</td>\n",
       "      <td>2</td>\n",
       "      <td>63.62%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-22.3</td>\n",
       "      <td>1</td>\n",
       "      <td>36.13%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0002678</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-3.757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-22.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x1fcb5c36d60>"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpmi_with_lactate_only(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "5a0f937e-b95e-4960-8e97-9fd6dc78bc4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>L_LACD2</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_L_LACD2</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac6d4a90</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>lac_L_c + q8_c --> pyr_c + q8h2_c</p>\n",
       "                    <p style='text-align:right'>L-Lactate + Ubiquinone-8 --> Pyruvate + Ubiquinol-8</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO0639 or (NGO0904 and NGO0905 and NGO0906)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction L_LACD2 at 0x1fcac6d4a90>"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#L_LACD2 is done by lldD (NGO0639), and l-lactate dehydrogenase that is quinone dependent. \n",
    "#IT was also found that lutCAB (NGO0904-NGO0906) are also L-lactate dehydrogenases, predicted to be quinone dependent.\n",
    "# These are added here. \n",
    "model.reactions.get_by_id('L_LACD2').gene_reaction_rule='NGO0639 or (NGO0904 and NGO0905 and NGO0906)'\n",
    "model.reactions.get_by_id('L_LACD2')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "fee11e6e-c0f6-4bd5-a38a-d2377ab714a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.4002647202967156</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>apoACP_c</td>\n",
       "      <td>EX_apoACP_c_</td>\n",
       "      <td>0.01471</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.1337</td>\n",
       "      <td>6</td>\n",
       "      <td>1.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.1007</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>0.2687</td>\n",
       "      <td>4</td>\n",
       "      <td>1.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.001896</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.001896</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.03932</td>\n",
       "      <td>3</td>\n",
       "      <td>0.15%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.005867</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.2842</td>\n",
       "      <td>5</td>\n",
       "      <td>1.82%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>3.826</td>\n",
       "      <td>2</td>\n",
       "      <td>9.80%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>4.802</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.04048</td>\n",
       "      <td>6</td>\n",
       "      <td>0.31%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.1168</td>\n",
       "      <td>6</td>\n",
       "      <td>0.90%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.07109</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.1892</td>\n",
       "      <td>6</td>\n",
       "      <td>1.45%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.1405</td>\n",
       "      <td>6</td>\n",
       "      <td>1.08%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.04104</td>\n",
       "      <td>5</td>\n",
       "      <td>0.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.00316</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.07966</td>\n",
       "      <td>9</td>\n",
       "      <td>0.92%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.311</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.09865</td>\n",
       "      <td>5</td>\n",
       "      <td>0.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>76.88%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.00158</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.1157</td>\n",
       "      <td>4</td>\n",
       "      <td>0.59%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.0276</td>\n",
       "      <td>11</td>\n",
       "      <td>0.39%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.05722</td>\n",
       "      <td>9</td>\n",
       "      <td>0.66%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.1908</td>\n",
       "      <td>5</td>\n",
       "      <td>1.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-8.926E-05</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0001785</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.4003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0001785</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.01471</td>\n",
       "      <td>10</td>\n",
       "      <td>0.24%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.003097</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-19.64</td>\n",
       "      <td>2</td>\n",
       "      <td>63.62%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-22.3</td>\n",
       "      <td>1</td>\n",
       "      <td>36.13%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0002678</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-3.757</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-22.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x1fcb3aa8bb0>"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpmi_with_lactate_only(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f5b50a05-e78b-4f32-8813-b5468148cc1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Gene identifier</strong></td><td>NGO1449</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac406430</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Functional</strong></td><td>True</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 3 reaction(s)</strong></td><td>\n",
       "                    LLACt3pp, GLYCLTt2rpp, DLACt2pp</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Gene NGO1449 at 0x1fcac406430>"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.genes.get_by_id('NGO1449')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "dfba4bbd-8f0c-4412-a36c-da955a90b2f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DLACt2pp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_DLACt2pp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac56cf70</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>h_p + lac_D_p <=> h_c + lac_D_c</p>\n",
       "                    <p style='text-align:right'>H + D-Lactate <=> H + D-Lactate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1449</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-999999.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DLACt2pp at 0x1fcac56cf70>"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#L-LActate transport is abolished in an lctP (NGO1449) mutant, therefore, removing NGO1361 from L-lactate transport.\n",
    "#NGO1361 is homologous to the L-lactate transporter lldP from E. coli, however it is much shorter in Gc, only 71 amino acids. It appears to be truncated/nonfunctioal. \n",
    "# So I am removing it also from DLACt2pp transport.\n",
    "\n",
    "model.reactions.get_by_id('LLACt3pp')\n",
    "model.reactions.get_by_id('LLACt3pp').gene_reaction_rule='NGO1449'\n",
    "model.reactions.get_by_id('LLACt3pp')\n",
    "\n",
    "model.reactions.get_by_id('DLACt2pp')\n",
    "model.reactions.get_by_id('DLACt2pp').gene_reaction_rule='NGO1449'\n",
    "model.reactions.get_by_id('DLACt2pp')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "ec9070bc-4d14-4df1-bf12-376be6b773eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 0.7212808737444482</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>apoACP_c</td>\n",
       "      <td>EX_apoACP_c_</td>\n",
       "      <td>0.02651</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.2409</td>\n",
       "      <td>6</td>\n",
       "      <td>0.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.1815</td>\n",
       "      <td>4</td>\n",
       "      <td>0.28%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>9.068</td>\n",
       "      <td>4</td>\n",
       "      <td>14.04%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.003417</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.003417</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.002278</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.002278</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>8.655</td>\n",
       "      <td>3</td>\n",
       "      <td>10.05%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.01057</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc_D_e</td>\n",
       "      <td>EX_glc_D_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>46.43%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>0.5122</td>\n",
       "      <td>5</td>\n",
       "      <td>0.99%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.07294</td>\n",
       "      <td>6</td>\n",
       "      <td>0.17%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.2105</td>\n",
       "      <td>6</td>\n",
       "      <td>0.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.1281</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>0.3409</td>\n",
       "      <td>6</td>\n",
       "      <td>0.79%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.2532</td>\n",
       "      <td>6</td>\n",
       "      <td>0.59%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.005695</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.002278</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.002278</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.1436</td>\n",
       "      <td>9</td>\n",
       "      <td>0.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>0.5605</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>0.1778</td>\n",
       "      <td>5</td>\n",
       "      <td>0.34%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>20</td>\n",
       "      <td>3</td>\n",
       "      <td>23.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.002848</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.2085</td>\n",
       "      <td>4</td>\n",
       "      <td>0.32%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.04973</td>\n",
       "      <td>11</td>\n",
       "      <td>0.21%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.1031</td>\n",
       "      <td>9</td>\n",
       "      <td>0.36%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>0.3438</td>\n",
       "      <td>5</td>\n",
       "      <td>0.67%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.002278</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4hba_c</td>\n",
       "      <td>DM_4HBA</td>\n",
       "      <td>-0.0001608</td>\n",
       "      <td>7</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.0003217</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-0.7213</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.0003217</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.02651</td>\n",
       "      <td>10</td>\n",
       "      <td>0.12%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.005581</td>\n",
       "      <td>1</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ac_e</td>\n",
       "      <td>EX_ac_e_</td>\n",
       "      <td>-32.22</td>\n",
       "      <td>2</td>\n",
       "      <td>28.14%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-37.87</td>\n",
       "      <td>1</td>\n",
       "      <td>16.53%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.0004825</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-10.22</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>-21.92</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lac_D_e</td>\n",
       "      <td>EX_lac_D_e_</td>\n",
       "      <td>-27.97</td>\n",
       "      <td>3</td>\n",
       "      <td>36.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>-8.51</td>\n",
       "      <td>5</td>\n",
       "      <td>18.58%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-25.65</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x1fcb3aa8fa0>"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpmi(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "d3c0187e-64c7-464c-a36e-9e9b897a3126",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>UCYR</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_UCYR</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac8c2430</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>2.0 focytC_p + q8h2_c <=> 2.0 ficytC_p + 2.0 h_p + q8_c</p>\n",
       "                    <p style='text-align:right'>2.0 ferrocytochrome, cytc + Ubiquinol-8 <=> 2.0 ferricytochrome, cytc + 2.0 H + Ubiquinone-8</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO2031 and NGO2030 and NGO2029 and (NGO1080 or NGO0101 or Orphan)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-999999.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>999999.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction UCYR at 0x1fcac8c2430>"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('UCYR')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "bcdf02a5-d512-4e99-9a2f-cd80ed8d238e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>q8_c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Ubiquinone-8</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac31ad60</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C49H74O4</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>cytosol</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 12 reaction(s)</strong></td><td>\n",
       "                    FUMR, NADH16pp, UCYR, L_LACD2, SUCD1, NADHQNa, NADPHQR2, NADH5, ASPO3, DHORD2, MDH2, LDH_D2</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite q8_c at 0x1fcac31ad60>"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.get_by_id('q8_c')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "3b11e0dd-428c-42bc-8494-67aa5ca71abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "rpmi_req=set(['EX_apoACP_c_', 'EX_trdrd_c_','EX_cobalt2_e_','EX_cu2_e_','EX_fe3_e_','EX_mn2_e_','EX_mobd_e_','EX_zn2_e_','EX_ca2_e_','EX_no3_e_','EX_h2o_e_','EX_k_e_','EX_cl_e_','EX_mg2_e_','EX_so4_e_','EX_na1_e_','EX_pi_e_','EX_h_e_','EX_arg_L_e_','EX_asn_L_e_','EX_asp_L_e_','EX_cys_L_e_','EX_glu_L_e_','EX_gly_e_','EX_his_L_e_','EX_ile_L_e_','EX_leu_L_e_','EX_lys_L_e_','EX_met_L_e_','EX_phe_L_e_','EX_pro_L_e_','EX_ser_L_e_','EX_thr_L_e_','EX_trp_L_e_','EX_tyr_L_e_','EX_val_L_e_','EX_pnto_R_e_','EX_chol_e_','EX_inost_e_','EX_glc_D_e_','EX_gthrd_e_','EX_co2_e_'])\n",
    "\n",
    "def rpmi(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -20.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -0. \n",
    "\n",
    "\n",
    "\n",
    "mdm_req=set(['EX_apoACP_c_', 'EX_trdrd_c_','EX_co2_e_', 'EX_cobalt2_e_', 'EX_cu2_e_' , 'EX_h_e_', 'EX_h2o_e_','EX_mn2_e_','EX_mobd_e_','EX_tungs_e_', 'EX_zn2_e_','EX_na1_e_','EX_cl_e_','EX_k_e_','EX_so4_e_','EX_nh4_e_','EX_mg2_e_','EX_pi_e_','EX_ca2_e_','EX_fe3_e_','EX_no3_e_','EX_asp_L_e_','EX_glu_L_e_','EX_arg_L_e_','EX_gly_e_','EX_ser_L_e_','EX_leu_L_e_','EX_ile_L_e_','EX_val_L_e_','EX_tyr_L_e_','EX_cys_L_e_','EX_pro_L_e_','EX_trp_L_e_','EX_thr_L_e_','EX_phe_L_e_','EX_asn_L_e_','EX_gln_L_e_','EX_his_L_e_','EX_met_L_e_','EX_ala_L_e_','EX_lys_L_e_','EX_gthrd_e_','EX_thm_e_','EX_pnto_R_e_','EX_glc_D_e_','EX_hxan_e_','EX_ura_e_'])\n",
    "\n",
    "def mdm(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -0.        \n",
    "\n",
    "def mdm_with_lactate(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -10.              \n",
    "\n",
    "def mdm_with_lactate_only(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -0.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -10. \n",
    "            \n",
    "            \n",
    "def mdm_molarity_with_lactate(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -1000.\n",
    "        if 'EX_glc_D_e_' in reaction.id:\n",
    "            reaction.lower_bound = -27.8\n",
    "        if 'EX_ser_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.48\n",
    "        if 'EX_asp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -3.76\n",
    "        if 'EX_asn_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.17\n",
    "        if 'EX_glu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -8.84\n",
    "        if 'EX_gly_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.33\n",
    "        if 'EX_ala_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.12\n",
    "        if 'EX_val_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.51\n",
    "        if 'EX_leu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.69\n",
    "        if 'EX_gln_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.34\n",
    "        if 'EX_pro_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.43\n",
    "        if 'EX_phe_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_cl_e_' in reaction.id:\n",
    "            reaction.lower_bound = -107.94697\n",
    "        if 'EX_k_e_' in reaction.id:\n",
    "            reaction.lower_bound = -11.48\n",
    "        if 'EX_so4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -5.74\n",
    "        if 'EX_nh4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -4.11\n",
    "        if 'EX_mg2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.07\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_pi_e_' in reaction.id:\n",
    "            reaction.lower_bound = -40\n",
    "        if 'EX_ca2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.254\n",
    "        if 'EX_no3_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.051\n",
    "        if 'EX_arg_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.71\n",
    "        if 'EX_ile_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.23\n",
    "        if 'EX_tyr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_cys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.65\n",
    "        if 'EX_trp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_thr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.42\n",
    "        if 'EX_his_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.13\n",
    "        if 'EX_met_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1\n",
    "        if 'EX_gthrd_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_thm_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.00697\n",
    "        if 'EX_pnto_R_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.004\n",
    "        if 'EX_hxan_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.37\n",
    "        if 'EX_ura_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.45  \n",
    "        if 'EX_na1_e_' in reaction.id:\n",
    "            reaction.lower_bound = -105.018\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -20\n",
    "    \n",
    "\n",
    "def mdm_molarity_with_lactate_only(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -1000.\n",
    "        if 'EX_glc_D_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0\n",
    "        if 'EX_ser_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.48\n",
    "        if 'EX_asp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -3.76\n",
    "        if 'EX_asn_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.17\n",
    "        if 'EX_glu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -8.84\n",
    "        if 'EX_gly_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.33\n",
    "        if 'EX_ala_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.12\n",
    "        if 'EX_val_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.51\n",
    "        if 'EX_leu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.69\n",
    "        if 'EX_gln_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.34\n",
    "        if 'EX_pro_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.43\n",
    "        if 'EX_phe_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_cl_e_' in reaction.id:\n",
    "            reaction.lower_bound = -107.94697\n",
    "        if 'EX_k_e_' in reaction.id:\n",
    "            reaction.lower_bound = -11.48\n",
    "        if 'EX_so4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -5.74\n",
    "        if 'EX_nh4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -4.11\n",
    "        if 'EX_mg2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.07\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_pi_e_' in reaction.id:\n",
    "            reaction.lower_bound = -40\n",
    "        if 'EX_ca2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.254\n",
    "        if 'EX_no3_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.051\n",
    "        if 'EX_arg_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.71\n",
    "        if 'EX_ile_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.23\n",
    "        if 'EX_tyr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_cys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.65\n",
    "        if 'EX_trp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_thr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.42\n",
    "        if 'EX_his_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.13\n",
    "        if 'EX_met_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1\n",
    "        if 'EX_gthrd_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_thm_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.00697\n",
    "        if 'EX_pnto_R_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.004\n",
    "        if 'EX_hxan_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.37\n",
    "        if 'EX_ura_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.45  \n",
    "        if 'EX_na1_e_' in reaction.id:\n",
    "            reaction.lower_bound = -105.018\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -20\n",
    "               \n",
    "def mdm_molarity(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in mdm_req:\n",
    "            reaction.lower_bound = -1000.\n",
    "        if 'EX_glc_D_e_' in reaction.id:\n",
    "            reaction.lower_bound = -27.8\n",
    "        if 'EX_ser_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.48\n",
    "        if 'EX_asp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -3.76\n",
    "        if 'EX_asn_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.17\n",
    "        if 'EX_glu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -8.84\n",
    "        if 'EX_gly_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.33\n",
    "        if 'EX_ala_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.12\n",
    "        if 'EX_val_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.51\n",
    "        if 'EX_leu_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.69\n",
    "        if 'EX_gln_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.34\n",
    "        if 'EX_pro_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.43\n",
    "        if 'EX_phe_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_cl_e_' in reaction.id:\n",
    "            reaction.lower_bound = -107.94697\n",
    "        if 'EX_k_e_' in reaction.id:\n",
    "            reaction.lower_bound = -11.48\n",
    "        if 'EX_so4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -5.74\n",
    "        if 'EX_nh4_e_' in reaction.id:\n",
    "            reaction.lower_bound = -4.11\n",
    "        if 'EX_mg2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -1.07\n",
    "        if 'EX_lys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.27\n",
    "        if 'EX_pi_e_' in reaction.id:\n",
    "            reaction.lower_bound = -40\n",
    "        if 'EX_ca2_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.254\n",
    "        if 'EX_no3_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.051\n",
    "        if 'EX_arg_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.71\n",
    "        if 'EX_ile_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.23\n",
    "        if 'EX_tyr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_cys_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.65\n",
    "        if 'EX_trp_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.39\n",
    "        if 'EX_thr_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.42\n",
    "        if 'EX_his_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.13\n",
    "        if 'EX_met_L_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.1\n",
    "        if 'EX_gthrd_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.15\n",
    "        if 'EX_thm_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.00697\n",
    "        if 'EX_pnto_R_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.004\n",
    "        if 'EX_hxan_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.37\n",
    "        if 'EX_ura_e_' in reaction.id:\n",
    "            reaction.lower_bound = -0.45  \n",
    "        if 'EX_na1_e_' in reaction.id:\n",
    "            reaction.lower_bound = -105.018\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -1000.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -0\n",
    "                         \n",
    "        \n",
    "def complete(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=-1000.   \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "7ebb30c2-aaf7-4fe4-bc95-4481673a02b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for reaction in model.reactions:\n",
    "     if reaction.lower_bound < -4:\n",
    "            reaction.lower_bound=-1000.\n",
    "for reaction in model.reactions:\n",
    "     if reaction.upper_bound > 4:\n",
    "            reaction.upper_bound=1000."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "788d0c03-2ed0-4baa-909e-9893735e5f05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 2.8627986056411365</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ala_L_e</td>\n",
       "      <td>EX_ala_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>8.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>apoACP_c</td>\n",
       "      <td>EX_apoACP_c_</td>\n",
       "      <td>0.1052</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.9561</td>\n",
       "      <td>6</td>\n",
       "      <td>1.67%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.7203</td>\n",
       "      <td>4</td>\n",
       "      <td>0.84%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.01356</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.01356</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.009041</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.009041</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.2806</td>\n",
       "      <td>3</td>\n",
       "      <td>0.25%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.04196</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glc_D_e</td>\n",
       "      <td>EX_glc_D_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>17.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gln_L_e</td>\n",
       "      <td>EX_gln_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>14.58%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>14.58%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>5.83%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.2895</td>\n",
       "      <td>6</td>\n",
       "      <td>0.51%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.8357</td>\n",
       "      <td>6</td>\n",
       "      <td>1.46%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.5084</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>1.353</td>\n",
       "      <td>6</td>\n",
       "      <td>2.37%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>1.005</td>\n",
       "      <td>6</td>\n",
       "      <td>1.76%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.2935</td>\n",
       "      <td>5</td>\n",
       "      <td>0.43%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.0226</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.009041</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.009041</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>129.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.5698</td>\n",
       "      <td>9</td>\n",
       "      <td>1.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>2.225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>14.58%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>8.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.0113</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thm_e</td>\n",
       "      <td>EX_thm_e_</td>\n",
       "      <td>0.0006384</td>\n",
       "      <td>12</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.8274</td>\n",
       "      <td>4</td>\n",
       "      <td>0.97%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.1974</td>\n",
       "      <td>11</td>\n",
       "      <td>0.63%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.4086</td>\n",
       "      <td>9</td>\n",
       "      <td>1.07%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>1.364</td>\n",
       "      <td>5</td>\n",
       "      <td>1.99%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.009041</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.001277</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-2.863</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.001277</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.1052</td>\n",
       "      <td>10</td>\n",
       "      <td>0.46%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.02215</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ade_e</td>\n",
       "      <td>EX_ade_e_</td>\n",
       "      <td>-0.8939</td>\n",
       "      <td>5</td>\n",
       "      <td>1.98%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>-25.34</td>\n",
       "      <td>4</td>\n",
       "      <td>44.80%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-119.3</td>\n",
       "      <td>1</td>\n",
       "      <td>52.74%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-0.001915</td>\n",
       "      <td>2</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-120.7</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-23.88</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x1fcb3aa87c0>"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdm(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "4e434a43-dff9-4a92-b0d2-cc997aef22d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Objective</h3><p>1.0 Nm_Ess_biomass = 2.287793744474097</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ala_L_e</td>\n",
       "      <td>EX_ala_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>9.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>arg_L_e</td>\n",
       "      <td>EX_arg_L_e_</td>\n",
       "      <td>0.764</td>\n",
       "      <td>6</td>\n",
       "      <td>1.49%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asn_L_e</td>\n",
       "      <td>EX_asn_L_e_</td>\n",
       "      <td>0.5756</td>\n",
       "      <td>4</td>\n",
       "      <td>0.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ca2_e</td>\n",
       "      <td>EX_ca2_e_</td>\n",
       "      <td>0.01084</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cl_e</td>\n",
       "      <td>EX_cl_e_</td>\n",
       "      <td>0.01084</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cobalt2_e</td>\n",
       "      <td>EX_cobalt2_e_</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cu2_e</td>\n",
       "      <td>EX_cu2_e_</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>cys_L_e</td>\n",
       "      <td>EX_cys_L_e_</td>\n",
       "      <td>0.18</td>\n",
       "      <td>3</td>\n",
       "      <td>0.18%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>fe3_e</td>\n",
       "      <td>EX_fe3_e_</td>\n",
       "      <td>0.03354</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gln_L_e</td>\n",
       "      <td>EX_gln_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>16.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glu_L_e</td>\n",
       "      <td>EX_glu_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>16.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gly_e</td>\n",
       "      <td>EX_gly_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>6.50%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h_e</td>\n",
       "      <td>EX_h_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>his_L_e</td>\n",
       "      <td>EX_his_L_e_</td>\n",
       "      <td>0.2314</td>\n",
       "      <td>6</td>\n",
       "      <td>0.45%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hxan_e</td>\n",
       "      <td>EX_hxan_e_</td>\n",
       "      <td>0.6464</td>\n",
       "      <td>5</td>\n",
       "      <td>1.05%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ile_L_e</td>\n",
       "      <td>EX_ile_L_e_</td>\n",
       "      <td>0.6678</td>\n",
       "      <td>6</td>\n",
       "      <td>1.30%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>k_e</td>\n",
       "      <td>EX_k_e_</td>\n",
       "      <td>0.4063</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lac_L_e</td>\n",
       "      <td>EX_lac_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>9.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>leu_L_e</td>\n",
       "      <td>EX_leu_L_e_</td>\n",
       "      <td>1.081</td>\n",
       "      <td>6</td>\n",
       "      <td>2.11%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>lys_L_e</td>\n",
       "      <td>EX_lys_L_e_</td>\n",
       "      <td>0.8031</td>\n",
       "      <td>6</td>\n",
       "      <td>1.57%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>met_L_e</td>\n",
       "      <td>EX_met_L_e_</td>\n",
       "      <td>0.2346</td>\n",
       "      <td>5</td>\n",
       "      <td>0.38%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mg2_e</td>\n",
       "      <td>EX_mg2_e_</td>\n",
       "      <td>0.01806</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mn2_e</td>\n",
       "      <td>EX_mn2_e_</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>mobd_e</td>\n",
       "      <td>EX_mobd_e_</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>o2_e</td>\n",
       "      <td>EX_o2_e_</td>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>phe_L_e</td>\n",
       "      <td>EX_phe_L_e_</td>\n",
       "      <td>0.4553</td>\n",
       "      <td>9</td>\n",
       "      <td>1.33%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pi_e</td>\n",
       "      <td>EX_pi_e_</td>\n",
       "      <td>1.674</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pro_L_e</td>\n",
       "      <td>EX_pro_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>16.26%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ser_L_e</td>\n",
       "      <td>EX_ser_L_e_</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>9.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>so4_e</td>\n",
       "      <td>EX_so4_e_</td>\n",
       "      <td>0.009032</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thm_e</td>\n",
       "      <td>EX_thm_e_</td>\n",
       "      <td>0.000603</td>\n",
       "      <td>12</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>thr_L_e</td>\n",
       "      <td>EX_thr_L_e_</td>\n",
       "      <td>0.6612</td>\n",
       "      <td>4</td>\n",
       "      <td>0.86%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>trp_L_e</td>\n",
       "      <td>EX_trp_L_e_</td>\n",
       "      <td>0.1577</td>\n",
       "      <td>11</td>\n",
       "      <td>0.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>tyr_L_e</td>\n",
       "      <td>EX_tyr_L_e_</td>\n",
       "      <td>0.3265</td>\n",
       "      <td>9</td>\n",
       "      <td>0.96%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ura_e</td>\n",
       "      <td>EX_ura_e_</td>\n",
       "      <td>0.5984</td>\n",
       "      <td>4</td>\n",
       "      <td>0.78%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>val_L_e</td>\n",
       "      <td>EX_val_L_e_</td>\n",
       "      <td>1.046</td>\n",
       "      <td>5</td>\n",
       "      <td>1.70%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>zn2_e</td>\n",
       "      <td>EX_zn2_e_</td>\n",
       "      <td>0.007225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Metabolite</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Flux</th>\n",
       "      <th>C-Number</th>\n",
       "      <th>C-Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>5drib_c</td>\n",
       "      <td>DM_5DRIB</td>\n",
       "      <td>-0.001025</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>amob_c</td>\n",
       "      <td>DM_AMOB</td>\n",
       "      <td>-4.552E-06</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>essbiomass</td>\n",
       "      <td>DM_ESSBIOMASS</td>\n",
       "      <td>-2.288</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>hmfurn_c</td>\n",
       "      <td>DM_HMFURN</td>\n",
       "      <td>-0.00102</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>pap_c</td>\n",
       "      <td>DM_PAP</td>\n",
       "      <td>-0.03982</td>\n",
       "      <td>10</td>\n",
       "      <td>0.19%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>urea_c</td>\n",
       "      <td>DM_UREA</td>\n",
       "      <td>-0.0177</td>\n",
       "      <td>1</td>\n",
       "      <td>0.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ade_e</td>\n",
       "      <td>EX_ade_e_</td>\n",
       "      <td>-0.0001177</td>\n",
       "      <td>5</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>asp_L_e</td>\n",
       "      <td>EX_asp_L_e_</td>\n",
       "      <td>-6.065</td>\n",
       "      <td>4</td>\n",
       "      <td>11.32%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>co2_e</td>\n",
       "      <td>EX_co2_e_</td>\n",
       "      <td>-147.9</td>\n",
       "      <td>1</td>\n",
       "      <td>69.01%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>glyclt_e</td>\n",
       "      <td>EX_glyclt_e_</td>\n",
       "      <td>-20.85</td>\n",
       "      <td>2</td>\n",
       "      <td>19.47%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>h2o_e</td>\n",
       "      <td>EX_h2o_e_</td>\n",
       "      <td>-89.43</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>nh4_e</td>\n",
       "      <td>EX_nh4_e_</td>\n",
       "      <td>-54.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.model_summary.ModelSummary at 0x19cdfb7e400>"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdm_with_lactate_only(model)\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "a76e975c-cc60-4c5d-b7be-441d2f21fa57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-2.095119304419103e-31\n",
      "no_p\n",
      "====\n",
      "Formula: NO\n",
      "\n",
      "Producing Reactions\n",
      "-------------------\n",
      "Percent  Flux Reaction                                             Definition\n",
      "100.00% 1.709     NO2R ficytC_p + h2o_p + no_p <=> focytC_p + 2.0 h_p + no2_p\n",
      "\n",
      "Consuming Reactions\n",
      "-------------------\n",
      "Percent   Flux Reaction                                                         Definition\n",
      "100.00% -1.709      NOR 2.0 ficytC_p + 2.0 h_c + 2.0 no_p <=> 2.0 focytC_p + h2o_p + n2o_p\n"
     ]
    }
   ],
   "source": [
    "#Check Gc model ability to grow anaerobically on nitrite. \n",
    "def anaerobic_rpmi(model):\n",
    "    for reaction in model.reactions:\n",
    "        if 'EX_' in  reaction.id:\n",
    "            reaction.lower_bound=0\n",
    "        if reaction.id in rpmi_req:\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_o2_e_':\n",
    "            reaction.lower_bound = -0.\n",
    "        if reaction.id == 'EX_glc_D_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_lac_L_e_':\n",
    "            reaction.lower_bound = -10.\n",
    "        if reaction.id == 'EX_no2_e_':\n",
    "            reaction.lower_bound = -10. \n",
    "\n",
    "anaerobic_rpmi(model)\n",
    "print(model.slim_optimize())\n",
    "print(model.metabolites.no_p.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "0962ee2f-effa-4b9c-9737-06f73a4a4552",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>NO2R</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_NO2R</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac769790</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>focytC_p + h2o_p + no_p <=> ficytC_p + 2.0 h_p + no2_p</p>\n",
       "                    <p style='text-align:right'>ferrocytochrome, cytc + H2O + Nitric-oxide <=> ferricytochrome, cytc + 2.0 H + Nitrite</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>NGO1276 and NGO1328</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction NO2R at 0x1fcac769790>"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#NO2R for nitrite reductase (aniA) is switched in terms of ficytC and focytC, which prevents use of \n",
    "#nitrite to support anaerobic growth. This also allows spontaneous creation of energy through ficytC when grwon on NO.\n",
    "#Correcting here. \n",
    "\n",
    "r=model.reactions.get_by_id('NO2R')\n",
    "ficytC_p=model.metabolites.get_by_id('ficytC_p')\n",
    "focytC_p=model.metabolites.get_by_id('focytC_p')\n",
    "\n",
    "r.add_metabolites({\n",
    "    ficytC_p: 2.0,\n",
    "    focytC_p: -2.0\n",
    "  \n",
    "})\n",
    "r\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "2ea99ac5-122f-4f29-951b-82597302f5bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.31597267853073796\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<h3>no_p</h3><p>NO</p><h4>Producing Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Percent</th>\n",
       "      <th>Flux</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>100.00%</td>\n",
       "      <td>10</td>\n",
       "      <td>NO2R</td>\n",
       "      <td>focytC_p + h2o_p + no_p &lt;=&gt; ficytC_p + 2.0 h_p + no2_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><h4>Consuming Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Percent</th>\n",
       "      <th>Flux</th>\n",
       "      <th>Reaction</th>\n",
       "      <th>Definition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>100.00%</td>\n",
       "      <td>-10</td>\n",
       "      <td>NOR</td>\n",
       "      <td>2.0 ficytC_p + 2.0 h_c + 2.0 no_p &lt;=&gt; 2.0 focytC_p + h2o_p + n2o_p</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<cobra.summary.metabolite_summary.MetaboliteSummary at 0x1fcb3f9bc40>"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Model now grows anaerobically on nitrite. \n",
    "anaerobic_rpmi(model)\n",
    "print(model.slim_optimize())\n",
    "model.metabolites.no_p.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "07808461-691d-419b-b507-978498086c42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>CYSabcpp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>R_CYSabcpp</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fcac53d6d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>atp_c + cys_L_p + h2o_c --> adp_c + cys_L_c + h_c + pi_c</p>\n",
       "                    <p style='text-align:right'>ATP + L-Cysteine + H2O --> ADP + L-Cysteine + H + Phosphate</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>(NGO0374 and NGO0373 and NGO0372) or (NGO0968 and NGO0269 and NGO1152)</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction CYSabcpp at 0x1fcac53d6d0>"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('CYSabcpp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "fb13f499-0ac9-40a5-92c1-343eab8d39e0",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{}\n"
     ]
    }
   ],
   "source": [
    "#NGO2011, NGO2012, NGO2013, and NGO2014 are a cysteine (cys_L_) transporter. PMID: 22138345\n",
    "#NGO0372, NGO373, NGO0374 are a cystine (cystine_L_) transporter. PMID: 22138345. In the current model, this is annotated as a cysteine importer \n",
    "#and no reaction for cystine import exists, or for breakdown of cystine. \n",
    "#Adding that transport and breakdown reaction here and correcting the gene annotations.\n",
    "\n",
    "#NGO0968 and NGO0269 and NGO1152 is listed for lots of different transport reactions - its not very well annotated. Removing it from cystine and cysteine import. \n",
    "\n",
    "CYSabcpp = model.reactions.get_by_id('CYSabcpp')\n",
    "CYSabcpp.gene_reaction_rule='NGO2011 and NGO2012 and NGO2013 and NGO2014'\n",
    "CYSabcpp.notes = {'citations': 'PMID: 22138345', 'type': '', 'annotation': ''}\n",
    "\n",
    "\n",
    "from cobra import Reaction, Metabolite\n",
    "\n",
    "cystine_L_c = Metabolite(\n",
    "    'cystine_L_c',\n",
    "    formula='C6H12N2O4S2',\n",
    "    name='Cystine',\n",
    "    compartment='cytosol')\n",
    "\n",
    "cystine_L_e = Metabolite(\n",
    "    'cystine_L_e',\n",
    "    formula='C6H12N2O4S2',\n",
    "    name='Cystine',\n",
    "    compartment='extracellular')\n",
    "\n",
    "cystine_L_p = Metabolite(\n",
    "    'cystine_L_p',\n",
    "    formula='C6H12N2O4S2',\n",
    "    name='Cystine',\n",
    "    compartment='periplasm')\n",
    "\n",
    "\n",
    "h_c = model.metabolites.get_by_id('h_c')\n",
    "atp_c = model.metabolites.get_by_id('atp_c')\n",
    "h2o_c = model.metabolites.get_by_id('h2o_c')\n",
    "adp_c = model.metabolites.get_by_id('adp_c')\n",
    "pi_c = model.metabolites.get_by_id('pi_c')\n",
    "\n",
    "CYSTINEabcpp = cobra.Reaction('CYSTINEabcpp')\n",
    "CYSTINEabcpp.id = 'CYSTINEabcpp'\n",
    "CYSTINEabcpp.name = 'cystine import'\n",
    "CYSTINEabcpp.gene_reaction_rule = 'NGO0374 and NGO0373 and NGO0372'\n",
    "CYSTINEabcpp.lower_bound = -1000.\n",
    "CYSTINEabcpp.upper_bound = 1000.\n",
    "CYSTINEabcpp.add_metabolites({\n",
    "        atp_c: -1.0,\n",
    "        cystine_L_p: -1.,\n",
    "        h2o_c: -1. ,\n",
    "        adp_c:  1.,\n",
    "        cystine_L_c: 1. ,\n",
    "        h_c: 1.,\n",
    "        pi_c: 1.,\n",
    "        })\n",
    "\n",
    "model.add_reactions([CYSTINEabcpp])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "EX_cystine_L_ = cobra.Reaction('EX_cystine_L_')\n",
    "EX_cystine_L_.notes = {'citations': 'PMID: 22138345', 'type': '', 'annotation': ''}\n",
    "EX_cystine_L_.name = 'exchange of cystine'\n",
    "EX_cystine_L_.gene_reaction_rule = ''\n",
    "EX_cystine_L_.lower_bound = 0.\n",
    "EX_cystine_L_.upper_bound = 1000.\n",
    "EX_cystine_L_.add_metabolites({\n",
    "    cystine_L_e: -1.0,})\n",
    "\n",
    "\n",
    "model.add_reactions([EX_cystine_L_])\n",
    "\n",
    "\n",
    "h_c = model.metabolites.get_by_id('h_c')\n",
    "nadh_c = model.metabolites.get_by_id('nadh_c')\n",
    "nad_c = model.metabolites.get_by_id('nad_c')\n",
    "cys_L_c = model.metabolites.get_by_id('cys_L_c')\n",
    "\n",
    "CYSR = cobra.Reaction('CYSR')\n",
    "CYSR.notes = {'citations': 'PMCID: PMC2323225', 'type': '', 'annotation': ''}\n",
    "CYSR.name = 'cystine reductase'\n",
    "CYSR.gene_reaction_rule = 'Orphan'\n",
    "CYSR.lower_bound = -1000.\n",
    "CYSR.upper_bound = 1000.\n",
    "CYSR.add_metabolites({\n",
    "    cys_L_c: -2.0,\n",
    "    nad_c: -1.0, \n",
    "    cystine_L_c: 1.0,\n",
    "    h_c: 1.0,\n",
    "    nadh_c: 1.0,\n",
    "    })\n",
    "\n",
    "\n",
    "model.add_reactions([CYSR])\n",
    "print(CYSR.check_mass_balance())\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "8ddebaab-c3e9-4195-95e8-e3359e3f33ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Aimee\\anaconda3\\lib\\site-packages\\cobra\\core\\model.py:727: UserWarning: need to pass in a list\n",
      "  warn(\"need to pass in a list\")\n"
     ]
    }
   ],
   "source": [
    "#Neisseria, like many other proteobacteria, lack the permease portion of the phosphotransferase. \n",
    "#And therefore do not use it for carbohydrate transport.\n",
    "#The remaining portions of the system are thought to perform regulatory roles, related to virulence. \n",
    "#PMID: 26858137\n",
    "\n",
    "model.remove_reactions(model.reactions.GLCptspp)\n",
    "model.remove_reactions(model.reactions.FRUpts2pp)\n",
    "model.remove_reactions(model.reactions.ASCBptspp)\n",
    "model.remove_reactions(model.reactions.MALTptspp)\n",
    "model.remove_reactions(model.reactions.MANptspp)\n",
    "model.remove_reactions(model.reactions.SUCptspp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "3b61b79e-ee11-4524-883f-265fe123ce93",
   "metadata": {},
   "outputs": [],
   "source": [
    "#NADPH does not catalyze the FabI reaction, only NADH. Therefore all of the NADPH dependent reactions were removed. \n",
    "#PMID: 26567338   \n",
    "model.remove_reactions(model.reactions.EAR120y)\n",
    "model.remove_reactions(model.reactions.EAR100y)\n",
    "model.remove_reactions(model.reactions.EAR121y)\n",
    "model.remove_reactions(model.reactions.EAR140y)\n",
    "model.remove_reactions(model.reactions.EAR141y)\n",
    "model.remove_reactions(model.reactions.EAR160y)\n",
    "model.remove_reactions(model.reactions.EAR161y)\n",
    "model.remove_reactions(model.reactions.EAR180y)\n",
    "model.remove_reactions(model.reactions.EAR181y)\n",
    "model.remove_reactions(model.reactions.EAR40y)\n",
    "model.remove_reactions(model.reactions.EAR60y)\n",
    "model.remove_reactions(model.reactions.EAR80y)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "72f0d8d1-d2af-4865-a4c7-98111a62a6e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Gc can only grow on glucose, lactate, and pyruvate. It cannot grow on gluconate. \n",
    "#Despite similarities to gluconate transporters and gluconate kinase, Nm also can't grow on gluconate so this should be removed. \n",
    "#PMID: 27454890\n",
    "model.remove_reactions(model.reactions.GLCNtex)\n",
    "model.remove_reactions(model.reactions.EX_glcn_e_)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "7d4fb26e-0ff2-4156-92c8-8748fe3bad17",
   "metadata": {},
   "outputs": [],
   "source": [
    "#To remove an infeasible loop, where gln is exchanged infinitely between the two Na+ and h+ dependent transporters, limited these to import only.\n",
    "\n",
    "model.reactions.GLNt2pp.lower_bound =0\n",
    "model.reactions.GLNt1pp.lower_bound =0  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "357e0690-d3cb-454b-80d7-db5f6311ea89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{}\n",
      "{}\n"
     ]
    }
   ],
   "source": [
    "#D_ala import is predicted to occur through an orphan reaction and spontaneously\n",
    "#Linking D_ala import to L_ala import genetically and na+ and h+ according to the transporters as appropriate\n",
    "\n",
    "\n",
    "ALADt2pp =model.reactions.ALADt\n",
    "\n",
    "ALADt2pp.id = 'ALADt2pp'\n",
    "ALADt2pp.gene_reaction_rule = 'NGO1807 or NGO1787 or NGO1290 or NGO1957'\n",
    "\n",
    "h_p = model.metabolites.get_by_id('h_p')\n",
    "h_c = model.metabolites.get_by_id('h_c')\n",
    "na1_c = model.metabolites.get_by_id('na1_c')\n",
    "na1_p = model.metabolites.get_by_id('na1_p')\n",
    "ala_D_p = model.metabolites.get_by_id('ala_D_p')\n",
    "ala_D_c = model.metabolites.get_by_id('ala_D_c')\n",
    "\n",
    "ALADt2pp.add_metabolites({\n",
    "        na1_p: -1.0,\n",
    "        na1_c: 1.0\n",
    "        })\n",
    "\n",
    "print(ALADt2pp.check_mass_balance())\n",
    "\n",
    "ALADt1pp = cobra.Reaction('ALADt1pp')\n",
    "ALADt1pp.name = 'D_ala import'\n",
    "ALADt1pp.gene_reaction_rule = 'NGO1290 or NGO1787'\n",
    "ALADt1pp.lower_bound = -1000.\n",
    "ALADt1pp.upper_bound = 1000.\n",
    "ALADt1pp.add_metabolites({\n",
    "    ala_D_p: -1.0,\n",
    "    h_p: -1.0,\n",
    "    ala_D_c: 1.0,\n",
    "    h_c: 1.0\n",
    "    })\n",
    "\n",
    "\n",
    "print(ALADt1pp.check_mass_balance())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "1f377ace-d5d4-4277-b2b8-b8f4e8693496",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove looping reactions\n",
    "model.reactions.ALAt1pp.lower_bound = 0\n",
    "model.reactions.ALAt2pp.lower_bound = 0\n",
    "model.reactions.ARGt1pp.lower_bound = 0\n",
    "model.reactions.ARGt2pp.lower_bound = 0\n",
    "model.reactions.ASNt1pp.lower_bound = 0\n",
    "model.reactions.ASNt2pp.lower_bound = 0\n",
    "model.reactions.ASPt1pp.lower_bound = 0\n",
    "model.reactions.ASPt2pp.lower_bound = 0\n",
    "model.reactions.CYSt1pp.lower_bound = 0\n",
    "model.reactions.CYSt2pp.lower_bound = 0\n",
    "model.reactions.GLUt1pp.lower_bound = 0\n",
    "model.reactions.GLUt2pp.lower_bound = 0\n",
    "model.reactions.HISt1pp.lower_bound = 0\n",
    "model.reactions.HISt2pp.lower_bound = 0\n",
    "model.reactions.ILEt1pp.lower_bound = 0\n",
    "model.reactions.ILEt3pp.lower_bound = 0\n",
    "model.reactions.LEUt1pp.lower_bound = 0\n",
    "model.reactions.LEUt3pp.lower_bound = 0\n",
    "model.reactions.LYSt1pp.lower_bound = 0\n",
    "model.reactions.LYSt2pp.lower_bound = 0\n",
    "model.reactions.METt1pp.lower_bound = 0\n",
    "model.reactions.METt2pp.lower_bound = 0\n",
    "model.reactions.PHEt1pp.lower_bound = 0\n",
    "model.reactions.PHEt2pp.lower_bound = 0\n",
    "model.reactions.PROt1pp.lower_bound = 0\n",
    "model.reactions.PROt2rpp.lower_bound = 0\n",
    "model.reactions.SERt1rpp.lower_bound = 0\n",
    "model.reactions.SERt2pp.lower_bound = 0\n",
    "model.reactions.THRt1rpp.lower_bound = 0\n",
    "model.reactions.THRt2pp.lower_bound = 0\n",
    "model.reactions.VALt1pp.lower_bound = 0\n",
    "model.reactions.VALt3pp.lower_bound = 0\n",
    "model.reactions.TRPS1.lower_bound=0\n",
    "model.reactions.ADK1.lower_bound=0\n",
    "model.reactions.NDPK1.lower_bound=0\n",
    "model.reactions.VALTA.upper_bound=0\n",
    "model.reactions.VPAMT.upper_bound=0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "45597d95-f2e8-4bd6-b0a3-ecee60e1be81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>NGO_513</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fca4858e20</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1341</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1433</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*Nm_Ess_biomass - 1.0*Nm_Ess_biomass_reverse_46af5</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>periplasm, cytosol, extracellular, C_c, C_p, c</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model NGO_513 at 0x1fca4858e20>"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "a2c07e2c-ac72-4313-917b-b5a8f07decde",
   "metadata": {},
   "outputs": [],
   "source": [
    "for metabolite in model.metabolites:\n",
    "    if '_c' in metabolite.id:\n",
    "        metabolite.compartment = 'cytosol'\n",
    "    if '_p' in metabolite.id:\n",
    "        metabolite.compartment = 'periplasm'\n",
    "    if '_e' in metabolite.id:\n",
    "        metabolite.compartment = 'extracellular'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "546633ae-7b34-458e-9f64-88853503baae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>NGO_513</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x01fca4858e20</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1341</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1433</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*Nm_Ess_biomass - 1.0*Nm_Ess_biomass_reverse_46af5</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>periplasm, cytosol, extracellular</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model NGO_513 at 0x1fca4858e20>"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "c68383d2-c4f2-47d4-949d-f5eade1e8ba7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "559"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.genes)\n",
    "#Subtract 2 for \"Sgene, and orphan\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "ec1df743-c896-47f0-87d4-49d9acdb34b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.id='iNGO_557'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "060a19bc-fe2c-4a2e-9f7b-e005a2cb7cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "cobra.io.write_sbml_model(model, cwd+'/Gc_GENRE_2022/Curate_Gc_Model/In_progress_curation/NGO_557.sbml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "9d39270e-3346-4975-9327-6970b28b7e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "cobra.io.save_json_model(model, cwd+'/Gc_GENRE_2022/Curate_Gc_Model/In_progress_curation/NGO_557.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "258228fc-1dc9-4042-b95e-d817cf6920d3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
